// tamin-framework/esm5/lib/services/overlay/overlay.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var OverlayService = /** @class */ (function () {
    function OverlayService() {
    }
    /**
     * @param {?=} element
     * @return {?}
     */
    OverlayService.prototype.show = /**
     * @param {?=} element
     * @return {?}
     */
    function (element) {
        if (element === void 0) { element = null; }
        /** @type {?} */
        var overlay = new PlainOverlay(element, { blur: 1, style: { zIndex: 9000 } });
        return overlay.show();
    };
    /**
     * @param {?} message
     * @param {?=} element
     * @return {?}
     */
    OverlayService.prototype.showMessage = /**
     * @param {?} message
     * @param {?=} element
     * @return {?}
     */
    function (message, element) {
        if (element === void 0) { element = null; }
        /** @type {?} */
        var html = '<div></div><div class="tamin-loader-loader4"></div>' + '<hr>' + message;
        /** @type {?} */
        var messageElement = document.createElement('div');
        messageElement.classList.add('overlay-message');
        // messageElement.classList.add('tamin-loader-loader3');
        messageElement.innerHTML = html;
        /** @type {?} */
        var overlay = new PlainOverlay(element, { blur: 1, face: messageElement });
        return overlay.show();
    };
    /**
     * @param {?} overlay
     * @return {?}
     */
    OverlayService.prototype.isShown = /**
     * @param {?} overlay
     * @return {?}
     */
    function (overlay) {
        return (overlay && (overlay.state === PlainOverlay.STATE_SHOWING || overlay.state === PlainOverlay.STATE_SHOWN));
    };
    /**
     * @param {?} overlay
     * @return {?}
     */
    OverlayService.prototype.hide = /**
     * @param {?} overlay
     * @return {?}
     */
    function (overlay) {
        setTimeout((/**
         * @return {?}
         */
        function () {
            overlay.hide();
        }), 0);
    };
    OverlayService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    OverlayService.ctorParameters = function () { return []; };
    /** @nocollapse */ OverlayService.ngInjectableDef = i0.defineInjectable({ factory: function OverlayService_Factory() { return new OverlayService(); }, token: OverlayService, providedIn: "root" });
    return OverlayService;
}());
export { OverlayService };
