// tamin-framework/esm5/lib/components/fields/tamin-field-validator/tamin-field-validator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldValidatorComponent = /** @class */ (function () {
    function TaminFieldValidatorComponent(controlContainer) {
        this.controlContainer = controlContainer;
    }
    Object.defineProperty(TaminFieldValidatorComponent.prototype, "errorMessages", {
        set: /**
         * @param {?} errorMessages
         * @return {?}
         */
        function (errorMessages) {
            // errorMessages.forEach((item) => {
            //   const temp = (<ValidationMessageModel[]>this._errorMessages).find(c => c.name === item.name);
            //   if (!temp) {
            //     this._errorMessages.push(item);
            //   } else {
            //     temp.value = item.value;
            //   }
            // });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.configureErrorMessages();
        this._control = this.controlContainer.control.get(this.formControlName);
    };
    Object.defineProperty(TaminFieldValidatorComponent.prototype, "control", {
        get: /**
         * @return {?}
         */
        function () {
            return this._control;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaminFieldValidatorComponent.prototype, "errorMessage", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this._control.valid) {
                var _loop_1 = function (propertyName) {
                    if (this_1._control.errors.hasOwnProperty(propertyName)) {
                        return { value: ((/** @type {?} */ (this_1._errorMessages))).find((/**
                             * @param {?} p
                             * @return {?}
                             */
                            function (p) { return p.name === propertyName; })).value };
                    }
                };
                var this_1 = this;
                for (var propertyName in this._control.errors) {
                    var state_1 = _loop_1(propertyName);
                    if (typeof state_1 === "object")
                        return state_1.value;
                }
            }
            return '';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.configureErrorMessages = /**
     * @return {?}
     */
    function () {
        this._errorMessages = [
            { name: 'required', value: 'مقدار فیلد الزامی می باشد' },
            { name: 'pattern', value: 'الگوی مقدار صحیح نمی باشد' },
            { name: 'email', value: 'ایمیل نادرست است' },
            { name: 'minlength', value: 'مقدار فیلد از حد مجاز کمتر است' },
            { name: 'maxlength', value: 'مقدار فیلد از حد مجاز بیشتر است' },
            { name: 'min', value: 'حداقل مقدار رعایت نشده است' },
            { name: 'max', value: 'حداکثر مقدار رعایت نشده است' },
            { name: 'requiredTrue', value: '' },
            { name: 'nullValidator', value: '' },
            { name: 'compose', value: '' },
            { name: 'composeAsync', value: '' },
            { name: 'requiredNumber', value: 'مقدار فیلد الزامی می باشد' },
            { name: 'persianDate', value: 'تاریخ نادرست است' },
            { name: 'nationalId', value: 'کد ملی نادرست است' }
        ];
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaminFieldValidatorComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
    };
    TaminFieldValidatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-validator',
                    template: "<span *ngIf=\"control?.touched && !control?.valid\" class=\"tamin-form-control-error\">\n  <span style=\"color: #b00020;\">{{errorMessage}}</span>\n</span>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldValidatorComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminFieldValidatorComponent.ctorParameters = function () { return [
        { type: ControlContainer }
    ]; };
    TaminFieldValidatorComponent.propDecorators = {
        formControlName: [{ type: Input }],
        errorMessages: [{ type: Input }]
    };
    return TaminFieldValidatorComponent;
}());
export { TaminFieldValidatorComponent };
if (false) {
    /** @type {?} */
    TaminFieldValidatorComponent.prototype._errorMessages;
    /** @type {?} */
    TaminFieldValidatorComponent.prototype.formControlName;
    /** @type {?} */
    TaminFieldValidatorComponent.prototype._control;
    /**
     * @type {?}
     * @private
     */
    TaminFieldValidatorComponent.prototype.controlContainer;
}
