// tamin-framework/esm2015/lib/components/containers/tamin-tab/tamin-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class TaminTabComponent {
    constructor() {
        this.visible = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
TaminTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tab',
                template: "<div class=\"tabcontent\" [style.display]=\"present ? 'block': 'none'\">\r\n  <ng-content *ngIf=\"present\"></ng-content>\r\n</div>\r\n",
                styles: [".tabcontent{display:none;padding:6px 12px;border:1px solid #ccc;border-top:none;min-height:100px;margin:0 3px 3px}"]
            }] }
];
/** @nocollapse */
TaminTabComponent.ctorParameters = () => [];
TaminTabComponent.propDecorators = {
    title: [{ type: Input }],
    visible: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminTabComponent.prototype.title;
    /** @type {?} */
    TaminTabComponent.prototype.visible;
    /** @type {?} */
    TaminTabComponent.prototype.present;
}
