// tamin-framework/esm5/lib/components/tamin-map/tamin-map.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { TaminLazyLoadService } from '../../services/tamin-lazy-load/tamin-lazy-load.service';
var TaminMapComponent = /** @class */ (function () {
    function TaminMapComponent(taminLazyLoadService) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.js = 'http://www.openlayers.org/api/OpenLayers.js';
        this.width = '100%';
        this.height = '100%';
    }
    /**
     * @return {?}
     */
    TaminMapComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.id = (new Date()).getTime().toString();
    };
    /**
     * @return {?}
     */
    TaminMapComponent.prototype.load = /**
     * @return {?}
     */
    function () {
        return this.taminLazyLoadService.loadJs(this.js).toPromise();
    };
    /**
     * @param {?} latitude
     * @param {?} longitude
     * @param {?} zoom
     * @return {?}
     */
    TaminMapComponent.prototype.setLocation = /**
     * @param {?} latitude
     * @param {?} longitude
     * @param {?} zoom
     * @return {?}
     */
    function (latitude, longitude, zoom) {
        /** @type {?} */
        var map = new OpenLayers.Map(this.id);
        map.addLayer(new OpenLayers.Layer.OSM());
        /** @type {?} */
        var lonLat = new OpenLayers.LonLat(longitude, latitude)
            .transform(new OpenLayers.Projection('EPSG:4326'), map.getProjectionObject());
        /** @type {?} */
        var markers = new OpenLayers.Layer.Markers('Markers');
        markers.addMarker(new OpenLayers.Marker(lonLat));
        map.setCenter(lonLat, zoom);
    };
    TaminMapComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-map',
                    template: "<div [attr.id]=\"id\" [style.width]=\"width\" [style.height]=\"height\"></div>\n\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminMapComponent.ctorParameters = function () { return [
        { type: TaminLazyLoadService }
    ]; };
    TaminMapComponent.propDecorators = {
        js: [{ type: Input }],
        css: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return TaminMapComponent;
}());
export { TaminMapComponent };
if (false) {
    /** @type {?} */
    TaminMapComponent.prototype.js;
    /** @type {?} */
    TaminMapComponent.prototype.css;
    /** @type {?} */
    TaminMapComponent.prototype.width;
    /** @type {?} */
    TaminMapComponent.prototype.height;
    /** @type {?} */
    TaminMapComponent.prototype.id;
    /**
     * @type {?}
     * @private
     */
    TaminMapComponent.prototype.taminLazyLoadService;
}
