// tamin-framework/esm2015/lib/components/fields/tamin-field-radio-button/tamin-field-radio-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
export class TaminFieldRadioButtonComponent {
    constructor() {
        this.change = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get name() {
        return this._name;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set name(value) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this._name = value;
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    setSelected() {
        this.inputElement.nativeElement.checked = true;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onChangeRadio(event) {
        if (this.inputElement.nativeElement.checked) {
            this.change.emit(this.value);
        }
    }
}
TaminFieldRadioButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-radio-button',
                template: "<input #inputElement\n       [attr.id]=\"name + '__' + value\"\n       type=\"radio\"\n       [attr.name]=\"name\"\n       [attr.value]=\"value\" (change)=\"onChangeRadio($event)\">\n<label [attr.for]=\"name + '__' + value\" [ngClass]=\"{'radio-seleted': inputElement.checked}\">\n  <i style=\"font-size: 1.3em\" [ngClass]=\"{'icon-circle-empty': !inputElement.checked, 'icon-dot-circled': inputElement.checked}\"></i>\n  <ng-content></ng-content>\n</label>\n\n",
                styles: [".radio-button{cursor:pointer}input[type=radio]{display:none}.radio-seleted{color:#00f;font-weight:700}label{cursor:pointer;color:#000}"]
            }] }
];
TaminFieldRadioButtonComponent.propDecorators = {
    inputElement: [{ type: ViewChild, args: ['inputElement',] }],
    value: [{ type: Input }],
    change: [{ type: Output }],
    name: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldRadioButtonComponent.prototype.inputElement;
    /** @type {?} */
    TaminFieldRadioButtonComponent.prototype.value;
    /** @type {?} */
    TaminFieldRadioButtonComponent.prototype.change;
    /**
     * @type {?}
     * @private
     */
    TaminFieldRadioButtonComponent.prototype._name;
}
