// tamin-framework/esm5/lib/components/image-editor/image-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, ViewChild } from '@angular/core';
import { TaminLazyLoadService } from '../../services/tamin-lazy-load/tamin-lazy-load.service';
import { Subscription } from 'rxjs';
import { TaminModalComponent } from '../tamin-modal/tamin-modal.component';
var ImageEditorComponent = /** @class */ (function () {
    function ImageEditorComponent(taminLazyLoadService, changeDetectorRef) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.changeDetectorRef = changeDetectorRef;
        this._subscription = new Subscription();
    }
    /**
     * @return {?}
     */
    ImageEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ImageEditorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ImageEditorComponent.prototype.getImage = /**
     * @return {?}
     */
    function () {
        this.cropper.getCroppedCanvas().toBlob((/**
         * @param {?} blob
         * @return {?}
         */
        function (blob) {
            /** @type {?} */
            var size = Math.ceil(blob.size / 1024).toString() + ' KB';
            alert(size);
        }));
    };
    /**
     * @param {?} image
     * @return {?}
     */
    ImageEditorComponent.prototype.open = /**
     * @param {?} image
     * @return {?}
     */
    function (image) {
        var _this = this;
        /** @type {?} */
        var me = this;
        this._subscription.add(this.taminLazyLoadService.loadJs('assets/cropperjs/cropper.js').subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this._subscription.add(_this.taminLazyLoadService.loadCss('assets/cropperjs/cropper.css').subscribe((/**
             * @param {?} value1
             * @return {?}
             */
            function (value1) {
                _this.thModal.show();
                if (_this.cropper) {
                    _this.cropper.clear();
                    _this.cropper.replace(image);
                }
                else {
                    _this.image.nativeElement.src = image;
                    _this.cropper = new Cropper(_this.image.nativeElement, {
                        aspectRatio: 16 / 9,
                        preview: me.imagePreview.nativeElement,
                        crop: (/**
                         * @param {?} e
                         * @return {?}
                         */
                        function (e) {
                        }),
                    });
                }
            })));
        })));
    };
    /**
     * @return {?}
     */
    ImageEditorComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._subscription.unsubscribe();
    };
    ImageEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-image-editor',
                    template: "<tamin-modal #thModal>\r\n  <div class=\"row\" style=\"direction: ltr\">\r\n    <div class=\"col-3\">\r\n      <div #imagePreview class=\"image-preview\"></div>\r\n      <table class=\"table-button\" cellpadding=\"0\" cellspacing=\"0\">\r\n        <tr>\r\n          <td (click)=\"cropper.move(-10,0)\"><i class=\"icon-left-big\"></i></td>\r\n          <td (click)=\"cropper.move(10,0)\"><i class=\"icon-right-big\"></i></td>\r\n          <td (click)=\"cropper.move(0,-10)\"><i class=\"icon-up-big\"></i></td>\r\n          <td (click)=\"cropper.move(0,10)\"><i class=\"icon-down-big\"></i></td>\r\n        </tr>\r\n      </table>\r\n      <table class=\"table-button\" cellpadding=\"0\" cellspacing=\"0\">\r\n        <tr>\r\n          <td (click)=\"cropper.rotate(-90)\"><i class=\"icon-cw\"></i></td>\r\n          <td (click)=\"cropper.rotate(90)\"><i class=\"icon-ccw\"></i></td>\r\n        </tr>\r\n      </table>\r\n      <table class=\"table-button\" cellpadding=\"0\" cellspacing=\"0\">\r\n        <tr>\r\n          <td (click)=\"cropper.zoom(-0.1)\"><i class=\"icon-zoom-out\"></i></td>\r\n          <td (click)=\"cropper.zoom(0.1)\"><i class=\"icon-zoom-in\"></i></td>\r\n        </tr>\r\n      </table>\r\n      <table class=\"table-button\" cellpadding=\"0\" cellspacing=\"0\">\r\n        <tr>\r\n          <td (click)=\"cropper.reset()\"><i class=\"icon-arrows-cw\"></i></td>\r\n        </tr>\r\n      </table>\r\n      <button (click)=\"getImage()\">Get Image</button>\r\n    </div>\r\n    <div class=\"col-9\" style=\"direction: ltr\">\r\n      <div style=\"width:100%; height: 500px\">\r\n        <img #image style=\"border: solid 1px #cacaca;\">\r\n      </div>\r\n    </div>\r\n  </div>\r\n</tamin-modal>\r\n",
                    styles: [".col-1,.col-10,.col-11,.col-12,.col-2,.col-3,.col-4,.col-5,.col-6,.col-7,.col-8,.col-9{padding:5px}.image-preview{overflow:hidden;width:256px;height:144px;border:1px solid #000}.image-preview>img{max-width:100%}.table-button{margin:5px;border-collapse:collapse}.table-button td{border:1px solid #007bff;background-color:#007bff;color:#fff;padding:5px;cursor:pointer}.table-button td:hover{border:1px solid #0062cc;background-color:#0062cc}"]
                }] }
    ];
    /** @nocollapse */
    ImageEditorComponent.ctorParameters = function () { return [
        { type: TaminLazyLoadService },
        { type: ChangeDetectorRef }
    ]; };
    ImageEditorComponent.propDecorators = {
        image: [{ type: ViewChild, args: ['image',] }],
        thModal: [{ type: ViewChild, args: ['thModal',] }],
        imagePreview: [{ type: ViewChild, args: ['imagePreview',] }]
    };
    return ImageEditorComponent;
}());
export { ImageEditorComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImageEditorComponent.prototype._subscription;
    /** @type {?} */
    ImageEditorComponent.prototype.image;
    /** @type {?} */
    ImageEditorComponent.prototype.thModal;
    /** @type {?} */
    ImageEditorComponent.prototype.imagePreview;
    /** @type {?} */
    ImageEditorComponent.prototype.cropper;
    /**
     * @type {?}
     * @private
     */
    ImageEditorComponent.prototype.taminLazyLoadService;
    /**
     * @type {?}
     * @private
     */
    ImageEditorComponent.prototype.changeDetectorRef;
}
