// tamin-framework/esm5/lib/components/tamin-tab/tab-label/tab-label.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, TemplateRef, ViewChild } from '@angular/core';
var TaminTabLabelComponent = /** @class */ (function () {
    function TaminTabLabelComponent() {
    }
    /**
     * @return {?}
     */
    TaminTabLabelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    TaminTabLabelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tab-label',
                    template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminTabLabelComponent.ctorParameters = function () { return []; };
    TaminTabLabelComponent.propDecorators = {
        labelContent: [{ type: ViewChild, args: [TemplateRef,] }]
    };
    return TaminTabLabelComponent;
}());
export { TaminTabLabelComponent };
if (false) {
    /** @type {?} */
    TaminTabLabelComponent.prototype.labelContent;
}
