// tamin-framework/esm2015/lib/components/containers/picker-container/picker-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostListener, Input, Output, ViewChild } from '@angular/core';
import PopperJs from 'popper.js';
export class PickerContainerComponent {
    /**
     * @param {?} changeDetectorRef
     */
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.adjustWidth = true;
        this.placement = 'bottom-end';
        this.padding = '5px';
        this.berforeHide = new EventEmitter();
        this.afterHide = new EventEmitter();
        this.berforeShow = new EventEmitter();
        this.afterShow = new EventEmitter();
    }
    /**
     * @return {?}
     */
    toggle() {
        if (this.shown) {
            this.hide();
        }
        else {
            this.show();
        }
    }
    /**
     * @return {?}
     */
    show() {
        this.berforeShow.emit();
        this.shown = true;
        this.changeDetectorRef.detectChanges();
        if (this.adjustWidth) {
            this.picker.nativeElement.style.width = this.parent.nativeElement.getBoundingClientRect().width - 3 + 'px';
        }
        if (this._thePopper) {
            this._thePopper = null;
        }
        this._thePopper = new PopperJs(this.parent.nativeElement, this.picker.nativeElement, {
            placement: (/** @type {?} */ (this.placement)),
        });
        this.afterShow.emit();
    }
    /**
     * @return {?}
     */
    hide() {
        this.berforeHide.emit();
        this.shown = false;
        this.afterHide.emit();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    update() {
        if (this.shown) {
            this._thePopper.scheduleUpdate();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    documentEventHandler(event) {
        if (this.shown) {
            if (event.code === 'Tab' || event.code === 'Escape') {
                this.toggle();
            }
            /** @type {?} */
            let shouldHidePicker = !this.picker.nativeElement.contains(event.target);
            shouldHidePicker = shouldHidePicker && !this.parent.nativeElement.contains(event.target);
            if (this.button) {
                shouldHidePicker = shouldHidePicker && !this.button.nativeElement.contains(event.target);
            }
            if (shouldHidePicker) {
                this.hide();
            }
        }
    }
}
PickerContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-picker-container',
                template: "<div class=\"picker-container\" *ngIf=\"shown\" #picker>\n  <ng-content></ng-content>\n</div>\n",
                styles: [".picker-container{max-height:60vh;overflow:auto;z-index:10010;background-color:#fff!important;box-shadow:0 0 18px -1px rgba(0,0,0,.62);margin:5px 0 5px 5px}"]
            }] }
];
/** @nocollapse */
PickerContainerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PickerContainerComponent.propDecorators = {
    shown: [{ type: Input }],
    parent: [{ type: Input }],
    button: [{ type: Input }],
    adjustWidth: [{ type: Input }],
    placement: [{ type: Input }],
    padding: [{ type: Input }],
    berforeHide: [{ type: Output }],
    afterHide: [{ type: Output }],
    berforeShow: [{ type: Output }],
    afterShow: [{ type: Output }],
    picker: [{ type: ViewChild, args: ['picker',] }],
    documentEventHandler: [{ type: HostListener, args: ['document:click', ['$event'],] }, { type: HostListener, args: ['document:keyup', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    PickerContainerComponent.prototype.shown;
    /** @type {?} */
    PickerContainerComponent.prototype.parent;
    /** @type {?} */
    PickerContainerComponent.prototype.button;
    /** @type {?} */
    PickerContainerComponent.prototype.adjustWidth;
    /** @type {?} */
    PickerContainerComponent.prototype.placement;
    /** @type {?} */
    PickerContainerComponent.prototype.padding;
    /** @type {?} */
    PickerContainerComponent.prototype.berforeHide;
    /** @type {?} */
    PickerContainerComponent.prototype.afterHide;
    /** @type {?} */
    PickerContainerComponent.prototype.berforeShow;
    /** @type {?} */
    PickerContainerComponent.prototype.afterShow;
    /** @type {?} */
    PickerContainerComponent.prototype.picker;
    /**
     * @type {?}
     * @private
     */
    PickerContainerComponent.prototype._thePopper;
    /**
     * @type {?}
     * @private
     */
    PickerContainerComponent.prototype.changeDetectorRef;
}
