// tamin-framework/esm2015/lib/components/tamin-tab/tab-item/tab-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, Input } from '@angular/core';
import { TaminTabBodyComponent } from '../tab-body/tab-body.component';
import { TaminTabLabelComponent } from '../tab-label/tab-label.component';
export class TaminTabItemComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
TaminTabItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tab-item',
                template: "\n",
                styles: [""]
            }] }
];
/** @nocollapse */
TaminTabItemComponent.ctorParameters = () => [];
TaminTabItemComponent.propDecorators = {
    label: [{ type: Input }],
    isActive: [{ type: Input }],
    bodyComponent: [{ type: ContentChild, args: [TaminTabBodyComponent,] }],
    labelComponent: [{ type: ContentChild, args: [TaminTabLabelComponent,] }]
};
if (false) {
    /** @type {?} */
    TaminTabItemComponent.prototype.label;
    /** @type {?} */
    TaminTabItemComponent.prototype.isActive;
    /** @type {?} */
    TaminTabItemComponent.prototype.bodyComponent;
    /** @type {?} */
    TaminTabItemComponent.prototype.labelComponent;
}
