// tamin-framework/esm2015/lib/components/fields/tamin-field-natioanl-id/tamin-field-natioanl-id.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldNatioanlIdComponent extends FieldBaseComponent {
    /**
     * @return {?}
     */
    initializeComponent() {
        this._setInputMask();
    }
    /**
     * @private
     * @return {?}
     */
    _setInputMask() {
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: '0000000000'
        });
    }
}
TaminFieldNatioanlIdComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-natioanl-id',
                template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  #inputElement>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldNatioanlIdComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
