// tamin-framework/esm5/lib/models/tamin-data-column.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var DataColumnViewType = {
    Label: 'Label',
    ThousandSeperated: 'ThousandSeperated',
    PersianDate: 'PersianDate',
    PersianDateTime: 'PersianDateTime',
    RowNumber: 'RowNumber',
    Custom: 'Custom',
    CustomRow: 'CustomRow',
};
export { DataColumnViewType };
var TaminDataColumn = /** @class */ (function () {
    function TaminDataColumn() {
    }
    return TaminDataColumn;
}());
export { TaminDataColumn };
if (false) {
    /** @type {?} */
    TaminDataColumn.prototype.columnName;
    /** @type {?} */
    TaminDataColumn.prototype.columnCaption;
    /** @type {?} */
    TaminDataColumn.prototype.columnWidth;
    /** @type {?} */
    TaminDataColumn.prototype.columnViewType;
    /** @type {?} */
    TaminDataColumn.prototype.columnSortable;
    /** @type {?} */
    TaminDataColumn.prototype.columnEditable;
    /** @type {?} */
    TaminDataColumn.prototype.columnTranslator;
    /** @type {?} */
    TaminDataColumn.prototype.columnStyle;
}
