// tamin-framework/esm5/lib/components/tamin-wizard/tamin-wizard/tamin-wizard.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, EventEmitter, Output, QueryList } from '@angular/core';
import { TaminWizardItemComponent } from './tamin-wizard-item/tamin-wizard-item.component';
var TaminWizardComponent = /** @class */ (function () {
    function TaminWizardComponent() {
        this.stepChanging = new EventEmitter();
        this.stepChanged = new EventEmitter();
    }
    Object.defineProperty(TaminWizardComponent.prototype, "currentItem", {
        get: /**
         * @return {?}
         */
        function () {
            return this.items.toArray()[this._current];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaminWizardComponent.prototype, "nextItem", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._current < this.items.length) {
                return this.items.toArray()[this._current + 1];
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaminWizardComponent.prototype, "prevItem", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._current > 0) {
                return this.items.toArray()[this._current - 1];
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (this.items.length > 0) {
            this._current = 0;
            this.refresh();
        }
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.onNextStep = /**
     * @return {?}
     */
    function () {
        if (this.items.length > 0 && this.items.length > this._current) {
            if (this.stepChanging) {
                this.stepChanging.emit(new WizardNavigationData(this._current, 'forward'));
            }
            else {
                this.setNext();
                this.refresh();
                this.stepChanged.emit(new WizardNavigationData(this._current, 'forward'));
            }
        }
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.onPrevStep = /**
     * @return {?}
     */
    function () {
        if (this.items.length > 0 && this._current > 0) {
            if (this.stepChanging) {
                this.stepChanging.emit(new WizardNavigationData(this._current, 'backward'));
            }
            else {
                this.setPrev();
                this.refresh();
                this.stepChanged.emit(new WizardNavigationData(this._current, 'backward'));
            }
        }
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.setNext = /**
     * @return {?}
     */
    function () {
        if (this.items.length > 0 && this.items.length > this._current) {
            ++this._current;
        }
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.setPrev = /**
     * @return {?}
     */
    function () {
        if (this.items.length > 0 && this._current > 0) {
            --this._current;
        }
    };
    /**
     * @param {?} step
     * @return {?}
     */
    TaminWizardComponent.prototype.setStep = /**
     * @param {?} step
     * @return {?}
     */
    function (step) {
        if (this.items.length > 0 && step > 0 && step < this.items.length) {
            this._current = step;
        }
    };
    /**
     * @return {?}
     */
    TaminWizardComponent.prototype.refresh = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.items.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            if (item.visible) {
                item.hide();
            }
            if (_this._current === index) {
                item.show();
            }
        }));
    };
    TaminWizardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-wizard',
                    template: "<div class=\"wizard\">\n  <div class=\"wizard-header\">{{ currentItem?.stepTitle}}</div>\n  <div class=\"wizard-content\">\n    <ng-content></ng-content>\n  </div>\n  <div class=\"wizard-footer\">\n    <button style=\"float: right; margin-bottom: 5px\" class=\"button button-default normal\" *ngIf=\"nextItem\" (click)=\"onNextStep()\">{{nextItem?.buttonTitle}}</button>\n    <button style=\"float: left; margin-bottom: 5px\" class=\"button button-default normal\" *ngIf=\"prevItem\" (click)=\"onPrevStep()\">\n      {{prevItem?.buttonTitle}}\n    </button>\n  </div>\n</div>\n\n",
                    styles: [".wizard{border:1px solid #cacaca;margin:5px;overflow:hidden}.wizard-content{padding:5px}.wizard-header{padding:5px;border-bottom:1px solid #cacaca;vertical-align:middle;text-align:center}.wizard-footer{padding:5px;border-top:1px solid #cacaca}"]
                }] }
    ];
    /** @nocollapse */
    TaminWizardComponent.ctorParameters = function () { return []; };
    TaminWizardComponent.propDecorators = {
        items: [{ type: ContentChildren, args: [TaminWizardItemComponent,] }],
        stepChanging: [{ type: Output }],
        stepChanged: [{ type: Output }]
    };
    return TaminWizardComponent;
}());
export { TaminWizardComponent };
if (false) {
    /** @type {?} */
    TaminWizardComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    TaminWizardComponent.prototype._current;
    /** @type {?} */
    TaminWizardComponent.prototype.stepChanging;
    /** @type {?} */
    TaminWizardComponent.prototype.stepChanged;
}
var WizardNavigationData = /** @class */ (function () {
    function WizardNavigationData(currentStep, direction) {
        this.currentStep = currentStep;
        this.direction = direction;
    }
    return WizardNavigationData;
}());
export { WizardNavigationData };
if (false) {
    /** @type {?} */
    WizardNavigationData.prototype.currentStep;
    /** @type {?} */
    WizardNavigationData.prototype.direction;
}
