// tamin-framework/esm5/lib/components/fields/tamin-field-label/tamin-field-label.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldLabelComponent = /** @class */ (function () {
    function TaminFieldLabelComponent(controlContainer) {
        this.controlContainer = controlContainer;
    }
    /**
     * @return {?}
     */
    TaminFieldLabelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.control = this.controlContainer.control.get(this.formControlName);
    };
    Object.defineProperty(TaminFieldLabelComponent.prototype, "required", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.control.validator === null) {
                return false;
            }
            /** @type {?} */
            var validator = this.control.validator((/** @type {?} */ ({})));
            return (validator && (validator.required || validator.requiredNumber));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldLabelComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldLabelComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    TaminFieldLabelComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaminFieldLabelComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
    };
    TaminFieldLabelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-label',
                    template: "<label class=\"tamin-form-control-label\"\n  [class.tamin-form-control-label]='!required'\n  [class.tamin-form-control-label-required]='required'\n  [class.tamin-form-control-label-invalid]='control.touched && !control.valid'>\n  <ng-content></ng-content>\n</label>\n\n\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldLabelComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminFieldLabelComponent.ctorParameters = function () { return [
        { type: ControlContainer }
    ]; };
    TaminFieldLabelComponent.propDecorators = {
        formControlName: [{ type: Input }]
    };
    return TaminFieldLabelComponent;
}());
export { TaminFieldLabelComponent };
if (false) {
    /** @type {?} */
    TaminFieldLabelComponent.prototype.formControlName;
    /** @type {?} */
    TaminFieldLabelComponent.prototype.control;
    /**
     * @type {?}
     * @private
     */
    TaminFieldLabelComponent.prototype.controlContainer;
}
