// tamin-framework/esm2015/lib/components/fields/tamin-tag/tamin-tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
export class TaminTagComponent extends FieldBaseComponent {
}
TaminTagComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tag',
                template: "<div class=\"tamin-form-control-tag\" #inputElement>\n  <div class=\"tamin-form-control-tag-element-container\">\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n  </div>\n  <span class=\"icon\" #icon style=\"float: left\">\n    <i class=\"icon-down-open\" style=\"font-size: 1.5em\"></i>\n  </span>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminTagComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
if (false) {
    /** @type {?} */
    TaminTagComponent.prototype.selectedItem;
}
