// tamin-framework/esm5/lib/components/tamin-tab/tabs/tabs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, ElementRef, HostListener, Input, QueryList, ViewChild } from '@angular/core';
import { TaminTabItemComponent } from '../tab-item/tab-item.component';
import { delay, map, startWith } from 'rxjs/operators';
var TaminTabsComponent = /** @class */ (function () {
    function TaminTabsComponent() {
        this.layout = 'horizontal';
        this.tabPosition = 'top';
    }
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.tabItems = this.tabs.changes
            .pipe(startWith(''))
            .pipe(delay(0))
            .pipe(map((/**
         * @return {?}
         */
        function () { return _this.tabs.toArray(); })));
    };
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.ngAfterContentChecked = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.activeTab) {
            Promise.resolve().then((/**
             * @return {?}
             */
            function () {
                _this.activeTab = _this.tabs.first;
            }));
        }
    };
    /**
     * @param {?} tabItem
     * @return {?}
     */
    TaminTabsComponent.prototype.selectTab = /**
     * @param {?} tabItem
     * @return {?}
     */
    function (tabItem) {
        if (this.activeTab === tabItem) {
            return;
        }
        if (this.activeTab) {
            this.activeTab.isActive = false;
        }
        this.activeTab = tabItem;
        tabItem.isActive = true;
    };
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.getBorderClass = /**
     * @return {?}
     */
    function () {
        return this.tabPosition === 'top' ? 'border-bottom' : 'border-top';
    };
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.onResize = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var hasScroll = this.hasScroll(this.tabHeader.nativeElement, 'horizontal');
        this.scroll = hasScroll;
    };
    /**
     * @private
     * @param {?} el
     * @param {?} direction
     * @return {?}
     */
    TaminTabsComponent.prototype.hasScroll = /**
     * @private
     * @param {?} el
     * @param {?} direction
     * @return {?}
     */
    function (el, direction) {
        direction = (direction === 'vertical') ? 'scrollTop' : 'scrollLeft';
        /** @type {?} */
        var result = !!el[direction];
        if (!result) {
            el[direction] = 1;
            result = !!el[direction];
            el[direction] = 0;
        }
        return result;
    };
    /**
     * @param {?} direction
     * @return {?}
     */
    TaminTabsComponent.prototype.scrollTo = /**
     * @param {?} direction
     * @return {?}
     */
    function (direction) {
        this.tabHeader.nativeElement.scrollLeft += direction === 'left' ? -100 : 100;
    };
    /**
     * @return {?}
     */
    TaminTabsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.onResize();
        }));
    };
    TaminTabsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tabs',
                    template: "<div class=\"tabs\">\n  <div *ngIf=\"tabPosition==='top'\" class=\"tab-header\" [ngClass]=\"getBorderClass()\">\n    <div class=\"flex-container\">\n      <div class=\"paddle\" *ngIf=\"scroll\" (click)=\"scrollTo('right')\"><i class=\"icon-right-open\"></i></div>\n      <ul #tabHeader>\n        <li\n          *ngFor=\"let item of tabItems | async\" (click)=\"selectTab(item)\"\n          [ngClass]=\"{'tab-header-item-active' : activeTab == item}\" class=\"tab-header\">\n          <ng-container *ngIf=\"item.labelComponent\">\n            <ng-container *ngTemplateOutlet=\"item.labelComponent.labelContent\">\n            </ng-container>\n          </ng-container>\n        </li>\n      </ul>\n      <div class=\"paddle\" *ngIf=\"scroll\" (click)=\"scrollTo('left')\">\n        <i class=\"icon-left-open\" style=\"margin-right: 5px;margin-left: 0;\"></i>\n      </div>\n    </div>\n  </div>\n  <div class=\"tab-content\" *ngFor=\"let item of tabItems | async\" [hidden]=\"activeTab !== item\">\n    <ng-container *ngTemplateOutlet=\"item.bodyComponent.bodyContent\">\n    </ng-container>\n  </div>\n</div>\n",
                    styles: [".flex-container{display:flex;align-items:stretch;background-color:#f1f1f1}ul{list-style-type:none;margin:auto;padding:0;overflow-x:hidden;white-space:nowrap;flex-grow:10;background-color:#fff}li{display:inline-block;padding:10px;cursor:pointer}.tabs{border:1px solid #e0e0e0;margin:3px;background-color:#fff}.tab-header-item{padding:15px;float:right}.tab-header-item-active{border-bottom:3px solid #00f;cursor:auto}.tab-content{padding:10px}li:hover:not(.tab-header-item-active){background-color:#dee2e6;border-bottom:3px solid #dee2e6}.border-top{border-top:1px solid #e0e0e0}.border-bottom{border-bottom:1px solid #e0e0e0}.vertical-item{border:1px solid #e0e0e0;margin:3px}i{margin-left:5px}.vertical-item-header{padding:10px;cursor:pointer}.vertical-item-header:hover{background-color:#dee2e6}.vertical-item-body{padding:10px}.paddle{display:flex;align-items:center;justify-content:center}"]
                }] }
    ];
    TaminTabsComponent.propDecorators = {
        tabs: [{ type: ContentChildren, args: [TaminTabItemComponent,] }],
        tabPosition: [{ type: Input }],
        tabHeader: [{ type: ViewChild, args: ['tabHeader',] }],
        onResize: [{ type: HostListener, args: ['window:resize',] }]
    };
    return TaminTabsComponent;
}());
export { TaminTabsComponent };
if (false) {
    /** @type {?} */
    TaminTabsComponent.prototype.tabs;
    /** @type {?} */
    TaminTabsComponent.prototype.tabItems;
    /** @type {?} */
    TaminTabsComponent.prototype.activeTab;
    /** @type {?} */
    TaminTabsComponent.prototype.layout;
    /** @type {?} */
    TaminTabsComponent.prototype.tabPosition;
    /** @type {?} */
    TaminTabsComponent.prototype.tabHeader;
    /** @type {?} */
    TaminTabsComponent.prototype.scroll;
}
