// tamin-framework/esm5/lib/components/fields/tamin-field-password/tamin-field-password.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef, Input, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { fromEvent } from 'rxjs';
var TaminFieldPasswordComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldPasswordComponent, _super);
    function TaminFieldPasswordComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.showPasswordMeter = false;
        _this.complexity = '';
        _this.score = '';
        return _this;
        // initPwdChk() {
        // this.scorebar.nativeElement.style.backgroundPosition = '0';
        // }
    }
    /**
     * @return {?}
     */
    TaminFieldPasswordComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var keyups = fromEvent(_this.inputElement.nativeElement, 'keyup');
            _this.subscription.add(keyups.subscribe((/**
             * @param {?} evt
             * @return {?}
             */
            function (evt) {
                if (_this.showPasswordMeter) {
                    _this.checkPassword(evt.target.value);
                }
            })));
        }), 0);
    };
    /**
     * @param {?} pwd
     * @return {?}
     */
    TaminFieldPasswordComponent.prototype.checkPassword = /**
     * @param {?} pwd
     * @return {?}
     */
    function (pwd) {
        this.color = 'white';
        this.complexity = '';
        this.score = '';
        if (pwd.trim().length === 0) {
            return;
        }
        /** @type {?} */
        var nScore = 0;
        /** @type {?} */
        var nLength = 0;
        /** @type {?} */
        var nAlphaUC = 0;
        /** @type {?} */
        var nAlphaLC = 0;
        /** @type {?} */
        var nNumber = 0;
        /** @type {?} */
        var nSymbol = 0;
        /** @type {?} */
        var nMidChar = 0;
        /** @type {?} */
        var nRequirements = 0;
        /** @type {?} */
        var nAlphasOnly = 0;
        /** @type {?} */
        var nNumbersOnly = 0;
        /** @type {?} */
        var nUnqChar = 0;
        /** @type {?} */
        var nRepChar = 0;
        /** @type {?} */
        var nRepInc = 0;
        /** @type {?} */
        var nConsecAlphaUC = 0;
        /** @type {?} */
        var nConsecAlphaLC = 0;
        /** @type {?} */
        var nConsecNumber = 0;
        /** @type {?} */
        var nConsecSymbol = 0;
        /** @type {?} */
        var nConsecCharType = 0;
        /** @type {?} */
        var nSeqAlpha = 0;
        /** @type {?} */
        var nSeqNumber = 0;
        /** @type {?} */
        var nSeqSymbol = 0;
        /** @type {?} */
        var nSeqChar = 0;
        /** @type {?} */
        var nReqChar = 0;
        // nMultConsecCharType = 0;
        // const nMultRepChar = 1, nMultConsecSymbol = 1;
        /** @type {?} */
        var nMultMidChar = 2;
        /** @type {?} */
        var 
        /*nMultRequirements = 2,*/
        nMultConsecAlphaUC = 2;
        /** @type {?} */
        var nMultConsecAlphaLC = 2;
        /** @type {?} */
        var nMultConsecNumber = 2;
        /** @type {?} */
        var nReqCharType = 3;
        /** @type {?} */
        var 
        /*nMultAlphaUC = 3,*/
        /*nMultAlphaLC = 3,*/
        nMultSeqAlpha = 3;
        /** @type {?} */
        var nMultSeqNumber = 3;
        /** @type {?} */
        var nMultSeqSymbol = 3;
        /** @type {?} */
        var nMultLength = 4;
        /** @type {?} */
        var nMultNumber = 4;
        /** @type {?} */
        var nMultSymbol = 6;
        /** @type {?} */
        var nTmpAlphaUC = '';
        /** @type {?} */
        var nTmpAlphaLC = '';
        /** @type {?} */
        var nTmpNumber = '';
        /** @type {?} */
        var nTmpSymbol = '';
        /** @type {?} */
        var sAlphaUC = '0';
        /** @type {?} */
        var sAlphaLC = '0';
        /** @type {?} */
        var sNumber = '0';
        /** @type {?} */
        var sSymbol = '0';
        /** @type {?} */
        var sMidChar = '0';
        /** @type {?} */
        var 
        // sRequirements = '0',
        sAlphasOnly = '0';
        /** @type {?} */
        var sNumbersOnly = '0';
        /** @type {?} */
        var sRepChar = '0';
        /** @type {?} */
        var sConsecAlphaUC = '0';
        /** @type {?} */
        var sConsecAlphaLC = '0';
        /** @type {?} */
        var sConsecNumber = '0';
        /** @type {?} */
        var sSeqAlpha = '0';
        /** @type {?} */
        var sSeqNumber = '0';
        /** @type {?} */
        var sSeqSymbol = '0';
        /** @type {?} */
        var sAlphas = 'abcdefghijklmnopqrstuvwxyz';
        /** @type {?} */
        var sNumerics = '01234567890';
        /** @type {?} */
        var sSymbols = ')!@#$%^&*()';
        /*let sComplexity = '';*/
        /** @type {?} */
        var sStandards = 'Below';
        /** @type {?} */
        var nMinPwdLen = 8;
        if (document.all) {
            /** @type {?} */
            var nd = 0;
        }
        else {
            /** @type {?} */
            var nd = 1;
        }
        if (pwd) {
            nScore = Number(pwd.length * nMultLength);
            nLength = pwd.length;
            /** @type {?} */
            var arrPwd = pwd.replace(/\s+/g, '').split(/\s*/);
            /** @type {?} */
            var arrPwdLen = arrPwd.length;
            for (var a = 0; a < arrPwdLen; a++) {
                if (arrPwd[a].match(/[A-Z]/g)) {
                    if (nTmpAlphaUC !== '') {
                        if (Number(nTmpAlphaUC + 1) === a) {
                            nConsecAlphaUC++;
                            nConsecCharType++;
                        }
                    }
                    nTmpAlphaUC = a.toString();
                    nAlphaUC++;
                }
                else if (arrPwd[a].match(/[a-z]/g)) {
                    if (nTmpAlphaLC !== '') {
                        if (Number(nTmpAlphaLC + 1) === a) {
                            nConsecAlphaLC++;
                            nConsecCharType++;
                        }
                    }
                    nTmpAlphaLC = a.toString();
                    nAlphaLC++;
                }
                else if (arrPwd[a].match(/[0-9]/g)) {
                    if (a > 0 && a < (arrPwdLen - 1)) {
                        nMidChar++;
                    }
                    if (nTmpNumber !== '') {
                        if (Number(nTmpNumber + 1) === a) {
                            nConsecNumber++;
                            nConsecCharType++;
                        }
                    }
                    nTmpNumber = a.toString();
                    nNumber++;
                }
                else if (arrPwd[a].match(/[^a-zA-Z0-9_]/g)) {
                    if (a > 0 && a < (arrPwdLen - 1)) {
                        nMidChar++;
                    }
                    if (nTmpSymbol !== '') {
                        if (Number(nTmpSymbol + 1) === a) {
                            nConsecSymbol++;
                            nConsecCharType++;
                        }
                    }
                    nTmpSymbol = a.toString();
                    nSymbol++;
                }
                /** @type {?} */
                var bCharExists = false;
                for (var b = 0; b < arrPwdLen; b++) {
                    if (arrPwd[a] === arrPwd[b] && a !== b) { /* repeat character exists */
                        bCharExists = true;
                        nRepInc += Math.abs(arrPwdLen / (b - a));
                    }
                }
                if (bCharExists) {
                    nRepChar++;
                    nUnqChar = arrPwdLen - nRepChar;
                    nRepInc = (nUnqChar) ? Math.ceil(nRepInc / nUnqChar) : Math.ceil(nRepInc);
                }
            }
            /* Check for sequential alpha string patterns (forward and reverse) */
            for (var s = 0; s < 23; s++) {
                /** @type {?} */
                var sFwd = sAlphas.substring(s, Number(s + 3));
                /** @type {?} */
                var sRev = this.strReverse(sFwd);
                if (pwd.toLowerCase().indexOf(sFwd) !== -1 || pwd.toLowerCase().indexOf(sRev) !== -1) {
                    nSeqAlpha++;
                    nSeqChar++;
                }
            }
            for (var s = 0; s < 8; s++) {
                /** @type {?} */
                var sFwd = sNumerics.substring(s, Number(s + 3));
                /** @type {?} */
                var sRev = this.strReverse(sFwd);
                if (pwd.toLowerCase().indexOf(sFwd) !== -1 || pwd.toLowerCase().indexOf(sRev) !== -1) {
                    nSeqNumber++;
                    nSeqChar++;
                }
            }
            for (var s = 0; s < 8; s++) {
                /** @type {?} */
                var sFwd = sSymbols.substring(s, Number(s + 3));
                /** @type {?} */
                var sRev = this.strReverse(sFwd);
                if (pwd.toLowerCase().indexOf(sFwd) !== -1 || pwd.toLowerCase().indexOf(sRev) !== -1) {
                    nSeqSymbol++;
                    nSeqChar++;
                }
            }
            if (nAlphaUC > 0 && nAlphaUC < nLength) {
                nScore = Number(nScore + ((nLength - nAlphaUC) * 2));
                sAlphaUC = '+ ' + Number((nLength - nAlphaUC) * 2);
            }
            if (nAlphaLC > 0 && nAlphaLC < nLength) {
                nScore = Number(nScore + ((nLength - nAlphaLC) * 2));
                sAlphaLC = '+ ' + Number((nLength - nAlphaLC) * 2);
            }
            if (nNumber > 0 && nNumber < nLength) {
                nScore = Number(nScore + (nNumber * nMultNumber));
                sNumber = '+ ' + Number(nNumber * nMultNumber);
            }
            if (nSymbol > 0) {
                nScore = Number(nScore + (nSymbol * nMultSymbol));
                sSymbol = '+ ' + Number(nSymbol * nMultSymbol);
            }
            if (nMidChar > 0) {
                nScore = Number(nScore + (nMidChar * nMultMidChar));
                sMidChar = '+ ' + Number(nMidChar * nMultMidChar);
            }
            if ((nAlphaLC > 0 || nAlphaUC > 0) && nSymbol === 0 && nNumber === 0) { // Only Letters
                nScore = Number(nScore - nLength);
                nAlphasOnly = nLength;
                sAlphasOnly = '- ' + nLength;
            }
            if (nAlphaLC === 0 && nAlphaUC === 0 && nSymbol === 0 && nNumber > 0) { // Only Numbers
                nScore = Number(nScore - nLength);
                nNumbersOnly = nLength;
                sNumbersOnly = '- ' + nLength;
            }
            if (nRepChar > 0) { // Same character exists more than once
                nScore = Number(nScore - nRepInc);
                sRepChar = '- ' + nRepInc;
            }
            if (nConsecAlphaUC > 0) { // Consecutive Uppercase Letters exist
                nScore = Number(nScore - (nConsecAlphaUC * nMultConsecAlphaUC));
                sConsecAlphaUC = '- ' + Number(nConsecAlphaUC * nMultConsecAlphaUC);
            }
            if (nConsecAlphaLC > 0) { // Consecutive Lowercase Letters exist
                nScore = Number(nScore - (nConsecAlphaLC * nMultConsecAlphaLC));
                sConsecAlphaLC = '- ' + Number(nConsecAlphaLC * nMultConsecAlphaLC);
            }
            if (nConsecNumber > 0) { // Consecutive Numbers exist
                nScore = Number(nScore - (nConsecNumber * nMultConsecNumber));
                sConsecNumber = '- ' + Number(nConsecNumber * nMultConsecNumber);
            }
            if (nSeqAlpha > 0) { // Sequential alpha strings exist (3 characters or more)
                nScore = Number(nScore - (nSeqAlpha * nMultSeqAlpha));
                sSeqAlpha = '- ' + Number(nSeqAlpha * nMultSeqAlpha);
            }
            if (nSeqNumber > 0) { // Sequential numeric strings exist (3 characters or more)
                nScore = Number(nScore - (nSeqNumber * nMultSeqNumber));
                sSeqNumber = '- ' + Number(nSeqNumber * nMultSeqNumber);
            }
            if (nSeqSymbol > 0) { // Sequential symbol strings exist (3 characters or more)
                nScore = Number(nScore - (nSeqSymbol * nMultSeqSymbol));
                sSeqSymbol = '- ' + Number(nSeqSymbol * nMultSeqSymbol);
            }
            /** @type {?} */
            var arrChars = [nLength, nAlphaUC, nAlphaLC, nNumber, nSymbol];
            /** @type {?} */
            var arrCharsIds = ['nLength', 'nAlphaUC', 'nAlphaLC', 'nNumber', 'nSymbol'];
            /** @type {?} */
            var arrCharsLen = arrChars.length;
            /** @type {?} */
            var minVal = void 0;
            for (var c = 0; c < arrCharsLen; c++) {
                // let oImg = $('div_' + arrCharsIds[c]);
                // let oBonus = $(arrCharsIds[c] + 'Bonus');
                // $(arrCharsIds[c]).innerHTML = arrChars[c];
                if (arrCharsIds[c] === 'nLength') {
                    minVal = Number(nMinPwdLen - 1);
                }
                else {
                    minVal = 0;
                }
                if (arrChars[c] === Number(minVal + 1)) {
                    nReqChar++;
                }
                else if (arrChars[c] > Number(minVal + 1)) {
                    nReqChar++;
                }
            }
            nRequirements = nReqChar;
            /** @type {?} */
            var nMinReqChars = void 0;
            if (pwd.length >= nMinPwdLen) {
                nMinReqChars = 3;
            }
            else {
                nMinReqChars = 4;
            }
            if (nRequirements > nMinReqChars) { // One or more required characters exist
                nScore = Number(nScore + (nRequirements * 2));
            }
            arrChars = [nMidChar, nRequirements];
            arrCharsIds = ['nMidChar', 'nRequirements'];
            arrCharsLen = arrChars.length;
            for (var c = 0; c < arrCharsLen; c++) {
                if (arrCharsIds[c] === 'nRequirements') {
                    minVal = nMinReqChars;
                }
                else {
                    minVal = 0;
                }
            }
            arrChars = [nAlphasOnly, nNumbersOnly, nRepChar, nConsecAlphaUC, nConsecAlphaLC, nConsecNumber, nSeqAlpha, nSeqNumber, nSeqSymbol];
            arrCharsIds = ['nAlphasOnly', 'nNumbersOnly', 'nRepChar', 'nConsecAlphaUC', 'nConsecAlphaLC', 'nConsecNumber', 'nSeqAlpha', 'nSeqNumber', 'nSeqSymbol'];
            arrCharsLen = arrChars.length;
            for (var c = 0; c < arrCharsLen; c++) {
                // const oImg = $('div_' + arrCharsIds[c]);
                // const oBonus = $(arrCharsIds[c] + 'Bonus');
                // $(arrCharsIds[c]).innerHTML = arrChars[c];
                if (arrChars[c] > 0) {
                    // oImg.className = 'warn';
                    // oBonus.parentNode.className = 'warn';
                }
                else {
                    // oImg.className = 'pass';
                    // oBonus.parentNode.className = 'pass';
                }
            }
            if (nScore > 100) {
                nScore = 100;
            }
            else if (nScore < 0) {
                nScore = 0;
            }
            if (nScore >= 0 && nScore < 20) {
                this.complexity = 'خیلی ضعیف';
                this.meter.nativeElement.style.width = nScore + '%';
                this.color = 'darkred';
            }
            else if (nScore >= 20 && nScore < 40) {
                this.meter.nativeElement.style.width = nScore + '%';
                this.color = 'red';
                this.complexity = 'ضعیف';
            }
            else if (nScore >= 40 && nScore < 60) {
                this.meter.nativeElement.style.width = nScore + '%';
                this.color = 'greenyellow';
                this.complexity = 'خوب';
            }
            else if (nScore >= 60 && nScore < 80) {
                this.meter.nativeElement.style.width = nScore + '%';
                this.color = 'lightgreen';
                this.complexity = 'قوی';
            }
            else if (nScore >= 80 && nScore <= 100) {
                this.meter.nativeElement.style.width = nScore + '%';
                this.color = 'green';
                this.complexity = 'خیلی قوی';
            }
            this.score = nScore.toString();
            this.meter.nativeElement.style.width = nScore + '%';
            // oScore.innerHTML  = nScore + '%';
            // oComplexity.innerHTML = sComplexity;
        }
        else {
            this.score = nScore.toString() + '%';
            // this.initPwdChk();
            // oScore.innerHTML = nScore + '%';
            // oComplexity.innerHTML = sComplexity;
        }
    };
    /**
     * @param {?} str
     * @return {?}
     */
    TaminFieldPasswordComponent.prototype.strReverse = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        /** @type {?} */
        var newstring = '';
        for (var s = 0; s < str.length; s++) {
            newstring = str.charAt(s) + newstring;
        }
        return newstring;
    };
    TaminFieldPasswordComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-password',
                    template: "<div>\n  <input\n    type=\"password\"\n    style=\"direction: ltr\"\n    class=\"tamin-form-control-input\"\n    [disabled]=\"control.disabled\"\n    [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n    [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n    #inputElement>\n  <div #meterContainer\n       style=\"height: 20px\"\n       class=\"meter-container\"\n       [class.meter-container-bkg-color]=\"complexity !== ''\"\n       [style.visibility]=\"showPasswordMeter ? 'visible' : 'hidden'\">\n    <div #meter [style.background-color]=\"color\" class=\"meter\" [hidden]=\"complexity == ''\"\n         style=\"color:white; padding-right: 5px\">\n      <span style=\"white-space: nowrap\">{{complexity}}</span>&nbsp;<span>(%{{score}})</span>\n    </div>\n  </div>\n</div>\n\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldPasswordComponent; })),
                            multi: true
                        }
                    ],
                    styles: [".meter-container{height:20px}.meter-container-bkg-color{background-color:#bbb}.meter{float:right;height:20px}"]
                }] }
    ];
    TaminFieldPasswordComponent.propDecorators = {
        meter: [{ type: ViewChild, args: ['meter',] }],
        meterContainer: [{ type: ViewChild, args: ['meterContainer',] }],
        showPasswordMeter: [{ type: Input }]
    };
    return TaminFieldPasswordComponent;
}(FieldBaseComponent));
export { TaminFieldPasswordComponent };
if (false) {
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.meter;
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.meterContainer;
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.showPasswordMeter;
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.color;
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.complexity;
    /** @type {?} */
    TaminFieldPasswordComponent.prototype.score;
}
