// tamin-framework/esm5/lib/components/tamin-tab/tab-item/tab-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, Input } from '@angular/core';
import { TaminTabBodyComponent } from '../tab-body/tab-body.component';
import { TaminTabLabelComponent } from '../tab-label/tab-label.component';
var TaminTabItemComponent = /** @class */ (function () {
    function TaminTabItemComponent() {
    }
    /**
     * @return {?}
     */
    TaminTabItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    TaminTabItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tab-item',
                    template: "\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminTabItemComponent.ctorParameters = function () { return []; };
    TaminTabItemComponent.propDecorators = {
        label: [{ type: Input }],
        isActive: [{ type: Input }],
        bodyComponent: [{ type: ContentChild, args: [TaminTabBodyComponent,] }],
        labelComponent: [{ type: ContentChild, args: [TaminTabLabelComponent,] }]
    };
    return TaminTabItemComponent;
}());
export { TaminTabItemComponent };
if (false) {
    /** @type {?} */
    TaminTabItemComponent.prototype.label;
    /** @type {?} */
    TaminTabItemComponent.prototype.isActive;
    /** @type {?} */
    TaminTabItemComponent.prototype.bodyComponent;
    /** @type {?} */
    TaminTabItemComponent.prototype.labelComponent;
}
