// tamin-framework/esm5/lib/components/tamin-site/tamin-site-top-menu/tamin-site-top-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
  Revised by: a_amiri
*/
import { ChangeDetectorRef, Component, ElementRef, HostListener, Input, ViewChild } from '@angular/core';
import { TaminMenuModel } from '../../../models/tamin-menu.model';
import { Router } from '@angular/router';
import { TaminRestService } from '../../../services/tamin-rest-service/tamin-rest.service';
var TaminSiteTopMenuComponent = /** @class */ (function () {
    function TaminSiteTopMenuComponent(taminRestService, router, ref) {
        this.taminRestService = taminRestService;
        this.router = router;
        this.ref = ref;
        this.menuData = new TaminMenuModel();
        this.stack = [];
        this.dummy = true;
    }
    /**
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.taminRestService.getAll(this.restUrl)
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.menuData = data.data;
            _this.stack.push(data.data);
        }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
        }));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.handleMenu = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item.disable) {
            alertify.alert('پیام سیستم', 'این خدمت بزودی در اختیار کاربران محترم سازمان تامین اجتماعی قرار خواهد گرفت.');
            return;
        }
        if (item.url !== '') {
            if (item['url-type'] === 'external') {
                window.open(item.url);
            }
            this.router.navigateByUrl('/' + item.url);
            this.dummy = false;
            this.ref.detectChanges();
            this.dummy = true;
            this.ref.detectChanges();
        }
    };
    /**
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.toggleMobileMenu = /**
     * @return {?}
     */
    function () {
        if (this.mobileMenuBody.nativeElement.style.display === 'block') {
            this.mobileMenuBody.nativeElement.style.display = 'none';
            this.mobileMenuButton.nativeElement.style.background = '#000a';
            this.mobileMenuButton.nativeElement.style.color = 'white';
        }
        else {
            this.mobileMenuBody.nativeElement.style.display = 'block';
            /** @type {?} */
            var rect = window.document.body.getBoundingClientRect();
            this.mobileMenuBody.nativeElement.style.width = rect.width + 'px';
            this.mobileMenuButton.nativeElement.style.background = '#00779caa';
            this.mobileMenuButton.nativeElement.style.color = 'white';
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.windowResizeHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var rect = window.document.body.getBoundingClientRect();
        this.mobileMenuBody.nativeElement.style.width = rect.width + 'px';
    };
    /**
     * @param {?} menuItem
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.onMobileMenuItemClick = /**
     * @param {?} menuItem
     * @return {?}
     */
    function (menuItem) {
        if (menuItem.items !== undefined && menuItem.items.length !== 0) {
            this.stack.push(menuItem);
        }
    };
    /**
     * @return {?}
     */
    TaminSiteTopMenuComponent.prototype.onMobileMenuReturn = /**
     * @return {?}
     */
    function () {
        this.stack.pop();
    };
    TaminSiteTopMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-site-top-menu',
                    template: "<!--\n  Created by: k_kaviani\n  Revised by: a_amiri\n-->\n<nav class=\"desktop\" *ngIf=\"dummy\">\n  <ul>\n    <li *ngFor=\"let item1 of menuData.items\">\n      <a (click)=\"handleMenu(item1)\" [ngClass]=\"{'disable':item1.disable}\">{{item1.caption}}</a>\n      <ul *ngIf=\"item1.items !== undefined\" class=\"child\">\n        <li *ngFor=\"let item2 of item1.items\">\n          <a (click)=\"handleMenu(item2)\" [ngClass]=\"{'hasSubItem':item2.items !== undefined, 'disable':item2.disable}\">{{item2.caption}}</a>\n          <ul *ngIf=\"item2.items !== undefined\" class=\"child\">\n            <li *ngFor=\"let item3 of item2.items\" [ngClass]=\"{'disable':item3.disable}\">\n              <a (click)=\"handleMenu(item3)\">{{item3.caption}}</a>\n            </li>\n          </ul>\n        </li>\n      </ul>\n    </li>\n  </ul>\n</nav>\n\n<nav class=\"mobile\" *ngIf=\"dummy\">\n  <ul>\n    <li>\n      <a (click)=\"toggleMobileMenu()\" #mobileMenuButton style=\"border-radius:0\"><i class=\"icon-menu\"></i></a>\n      <ul class=\"child\" #mobileMenuBody>\n        <li *ngIf=\"stack.length !== 1\"><a (click)=\"onMobileMenuReturn()\"><i class=\"icon-right-open\"\n                                                                            style=\"margin-left: 5px\"></i>\u0628\u0627\u0632\u06AF\u0634\u062A</a></li>\n        <li *ngFor=\"let item of stack[stack.length-1]?.items\" (click)=\"onMobileMenuItemClick(item)\">\n          <a (click)=\"handleMenu(item)\">{{item.caption}}</a></li>\n      </ul>\n    </li>\n  </ul>\n</nav>\n",
                    styles: [".disable{background-color:rgba(119,119,119,.67)}nav.desktop{text-align:right;width:100%;position:static}nav.desktop a{margin:0;background-color:transparent;cursor:pointer}nav.desktop ul li:hover>ul{display:block}nav.desktop ul{background:rgba(173,173,173,.6);border:1px solid #0009;list-style:none;position:relative;z-index:1000}nav.desktop ul:after{content:\"\";clear:both;display:block}nav.desktop ul li{float:right;position:relative;font-weight:700}nav.desktop ul li:hover{background:rgba(0,0,0,.67)}nav.desktop ul li:hover a{color:#fff}nav.desktop ul li a{display:block;padding:10px 15px;color:#333;text-decoration:none}nav.desktop ul ul{display:none;background:rgba(0,0,0,.67);border-radius:0;padding:0;position:absolute;top:100%;right:0;min-width:300px}nav.desktop ul ul li{float:none;position:relative}nav.desktop ul ul li a{padding:10px 15px;color:#fff}nav.desktop a.hasSubItem:after{font-family:fontello;content:'\\e804';float:left;margin-top:4px}nav.desktop ul ul li a:hover{background:rgba(44,123,175,.67)}nav.desktop ul ul ul{position:absolute;right:100%;top:0}nav.mobile{text-align:right;width:100%;display:none}nav.mobile a{margin:0;background-color:transparent;cursor:pointer}nav.mobile ul{background:rgba(51,64,75,.6);border:1px solid #555;list-style:none;position:relative;z-index:1000}nav.mobile ul:after{content:\"\";clear:both;display:block}nav.mobile ul li{float:right;position:relative;font-weight:700}nav.mobile ul li a{display:block;padding:10px 15px;color:#fff;text-decoration:none}nav.mobile ul ul{display:none;background:rgba(0,0,0,.85);border-radius:0;padding:0;position:absolute;top:100%;right:0}nav.mobile ul ul li{float:none;position:relative}nav.mobile ul ul li a{padding:10px 15px;color:#fff}nav.mobile a.hasSubItem:after{font-family:fontello;content:'\\e804';float:left;margin-top:4px}nav.mobile ul ul li a:hover{background:rgba(255,255,255,.33)}nav.mobile ul ul ul{position:absolute;right:100%;top:0}@media screen and (max-width:767px),(orientation:portrait){nav.desktop{display:none}nav.mobile{display:block}}"]
                }] }
    ];
    /** @nocollapse */
    TaminSiteTopMenuComponent.ctorParameters = function () { return [
        { type: TaminRestService },
        { type: Router },
        { type: ChangeDetectorRef }
    ]; };
    TaminSiteTopMenuComponent.propDecorators = {
        desktopMenu: [{ type: ViewChild, args: ['desktopMenu',] }],
        mobileMenu: [{ type: ViewChild, args: ['mobileMenu',] }],
        topMenu: [{ type: ViewChild, args: ['topMenu',] }],
        menuElement: [{ type: ViewChild, args: ['menuElement',] }],
        mobileMenuBody: [{ type: ViewChild, args: ['mobileMenuBody',] }],
        mobileMenuButton: [{ type: ViewChild, args: ['mobileMenuButton',] }],
        restUrl: [{ type: Input }],
        windowResizeHandler: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
    };
    return TaminSiteTopMenuComponent;
}());
export { TaminSiteTopMenuComponent };
if (false) {
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.desktopMenu;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.mobileMenu;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.topMenu;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.menuElement;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.mobileMenuBody;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.mobileMenuButton;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.restUrl;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.menuData;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.stack;
    /** @type {?} */
    TaminSiteTopMenuComponent.prototype.dummy;
    /**
     * @type {?}
     * @private
     */
    TaminSiteTopMenuComponent.prototype.taminRestService;
    /**
     * @type {?}
     * @private
     */
    TaminSiteTopMenuComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    TaminSiteTopMenuComponent.prototype.ref;
}
