// tamin-framework/esm5/lib/components/fields/tamin-field-mask/tamin-field-mask.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
var TaminFieldMaskComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldMaskComponent, _super);
    function TaminFieldMaskComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(TaminFieldMaskComponent.prototype, "mask", {
        get: /**
         * @return {?}
         */
        function () {
            return this.maskRef.mask;
        },
        set: /**
         * @param {?} mask
         * @return {?}
         */
        function (mask) {
            var _this = this;
            /** @type {?} */
            var me = this;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.maskRef = new IMask(_this.inputElement.nativeElement, {
                    mask: mask,
                });
            }), 0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaminFieldMaskComponent.prototype, "regex", {
        get: /**
         * @return {?}
         */
        function () {
            return this.maskRef.regex;
        },
        set: /**
         * @param {?} regex
         * @return {?}
         */
        function (regex) {
            var _this = this;
            /** @type {?} */
            var me = this;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.maskRef = new IMask(_this.inputElement.nativeElement, {
                    mask: new RegExp(regex),
                });
            }), 0);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminFieldMaskComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
    };
    TaminFieldMaskComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-mask',
                    template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  class=\"tamin-form-control-input tamin-form-control-mask\"\n  [class.tamin-form-control-valid]=\"control.touched && control?.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control?.valid\"\n  #inputElement>\n<span>{{hint}}</span>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldMaskComponent; })),
                            multi: true
                        }
                    ],
                    styles: [".tamin-form-control-mask{direction:ltr;text-align:right}"]
                }] }
    ];
    TaminFieldMaskComponent.propDecorators = {
        mask: [{ type: Input }],
        regex: [{ type: Input }]
    };
    return TaminFieldMaskComponent;
}(FieldBaseComponent));
export { TaminFieldMaskComponent };
