// tamin-framework/esm5/lib/components/fields/tamin-field-persian-text/tamin-field-persian-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
var TaminFieldPersianTextComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldPersianTextComponent, _super);
    function TaminFieldPersianTextComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.maxLength = Number.MAX_SAFE_INTEGER;
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldPersianTextComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: /^[\u0600-\u06FF\s]+$/,
        });
    };
    TaminFieldPersianTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-persian-text',
                    template: "<input\n  type=\"text\"\n  class=\"tamin-form-control-input\"\n  [disabled] = \"control.disabled\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [attr.maxLength] = \"maxLength\"\n  [attr.placeholder] =\"placeHolder\"\n  #inputElement>\n<span>{{hint}}</span>\n\n\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldPersianTextComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    TaminFieldPersianTextComponent.propDecorators = {
        maxLength: [{ type: Input }]
    };
    return TaminFieldPersianTextComponent;
}(FieldBaseComponent));
export { TaminFieldPersianTextComponent };
if (false) {
    /** @type {?} */
    TaminFieldPersianTextComponent.prototype.maxLength;
}
