// tamin-framework/esm5/lib/components/tamin-document-viewer/tamin-document-viewer-modal/tamin-document-viewer-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { TaminDocumentViewerComponent } from '../tamin-document-viewer.component';
import { TaminModalComponent } from '../../tamin-modal/tamin-modal.component';
var TaminDocumentViewerModalComponent = /** @class */ (function () {
    function TaminDocumentViewerModalComponent() {
    }
    /**
     * @return {?}
     */
    TaminDocumentViewerModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} url
     * @return {?}
     */
    TaminDocumentViewerModalComponent.prototype.openImage = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.theModal.show();
        this.documentViewer.loadImage(url);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    TaminDocumentViewerModalComponent.prototype.openPdf = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.theModal.show();
        this.documentViewer.loadPdf(url);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    TaminDocumentViewerModalComponent.prototype.openPdfBased64 = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.theModal.show();
        this.documentViewer.loadPdfBased64(url);
    };
    TaminDocumentViewerModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-document-viewer-modal',
                    template: "<tamin-modal #theModal>\n  <tamin-document-viewer #documentViewer></tamin-document-viewer>\n</tamin-modal>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminDocumentViewerModalComponent.ctorParameters = function () { return []; };
    TaminDocumentViewerModalComponent.propDecorators = {
        documentViewer: [{ type: ViewChild, args: ['documentViewer',] }],
        theModal: [{ type: ViewChild, args: ['theModal',] }]
    };
    return TaminDocumentViewerModalComponent;
}());
export { TaminDocumentViewerModalComponent };
if (false) {
    /** @type {?} */
    TaminDocumentViewerModalComponent.prototype.documentViewer;
    /** @type {?} */
    TaminDocumentViewerModalComponent.prototype.theModal;
}
