// tamin-framework/esm5/lib/components/tamin-data-grid/tamin-data-grid-action-menu/tamin-data-grid-action-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PickerContainerComponent } from '../../containers/picker-container/picker-container.component';
var TaminDataGridActionMenuComponent = /** @class */ (function () {
    function TaminDataGridActionMenuComponent() {
        this.actionColumns = [];
        this.menuItemClicked = new EventEmitter();
        this.dropdownClicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TaminDataGridActionMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // this.picker.parent = this.dropdown;
        // this.picker.parent = new ElementRef(this.dropdown.nativeElement.parentNode);
    };
    /**
     * @param {?} actionColumn
     * @return {?}
     */
    TaminDataGridActionMenuComponent.prototype.theMenuItemClicked = /**
     * @param {?} actionColumn
     * @return {?}
     */
    function (actionColumn) {
        this.picker.toggle();
        /** @type {?} */
        var item = this.dataItem;
        this.menuItemClicked.emit({ actionColumn: actionColumn, item: item });
    };
    /**
     * @return {?}
     */
    TaminDataGridActionMenuComponent.prototype.dropdownClickedHandler = /**
     * @return {?}
     */
    function () {
        this.picker.toggle();
        this.dropdownClicked.emit();
    };
    /**
     * @return {?}
     */
    TaminDataGridActionMenuComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.picker.parent = new ElementRef(this.dropdown.nativeElement.parentNode);
    };
    TaminDataGridActionMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-data-grid-action-menu',
                    template: "<div class=\"dropdown\" #dropdown (click)=\"dropdownClickedHandler()\">\n  <ng-container *ngIf=\"actionColumns.length != 0\">\n    <span>\u0639\u0645\u0644\u06CC\u0627\u062A</span>&nbsp;&nbsp;<i class=\" icon-down-open\"></i>\n  </ng-container>\n  <ng-container *ngIf=\"actionColumns.length == 0\">\n    <span style=\"color: #aaaaaa\">\u0639\u0645\u0644\u06CC\u0627\u062A</span>&nbsp;&nbsp;<i class=\" icon-down-open\" style=\"color: #aaaaaa\"></i>\n  </ng-container>\n</div>\n<tamin-picker-container #picker>\n  <ng-container *ngFor=\"let actionColumn of actionColumns\">\n    <div class=\"data-grid-menu-item\" *ngIf=\"actionColumn.visible\"\n         (click)=\"theMenuItemClicked(actionColumn)\">\n      <i [ngClass]=\"actionColumn.icon\"></i>\n      <span>{{actionColumn.columnCaption}}</span>\n    </div>\n  </ng-container>\n</tamin-picker-container>\n\n\n",
                    styles: [".dropdown{color:#0b7dda;background-color:transparent;cursor:pointer;padding-left:10px;padding-right:10px}.dropdown:hover{font-weight:700}.data-grid-menu-item{cursor:pointer;text-align:right;padding:5px 15px}.data-grid-menu-item:hover{background-color:#0b7dda;color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    TaminDataGridActionMenuComponent.ctorParameters = function () { return []; };
    TaminDataGridActionMenuComponent.propDecorators = {
        dropdown: [{ type: ViewChild, args: ['dropdown',] }],
        picker: [{ type: ViewChild, args: ['picker',] }],
        actionColumns: [{ type: Input }],
        dataItem: [{ type: Input }],
        menuItemClicked: [{ type: Output }],
        dropdownClicked: [{ type: Output }]
    };
    return TaminDataGridActionMenuComponent;
}());
export { TaminDataGridActionMenuComponent };
if (false) {
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.dropdown;
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.picker;
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.actionColumns;
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.dataItem;
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.menuItemClicked;
    /** @type {?} */
    TaminDataGridActionMenuComponent.prototype.dropdownClicked;
}
