// tamin-framework/esm2015/lib/components/fields/tamin-field-combo-box-static/tamin-field-combo-box-static.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, forwardRef, Input, ViewChild } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { PickerContainerComponent } from '../../containers/picker-container/picker-container.component';
export class TaminFieldComboBoxStaticComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.dataItems = [];
    }
    /**
     * @return {?}
     */
    initializeComponent() {
        this.picker.parent = this.inputElement;
        this.picker.button = this.iconElement;
    }
    /**
     * @return {?}
     */
    onShowPicker() {
        this.picker.toggle();
    }
    /**
     * @param {?} itemData
     * @return {?}
     */
    renderItem(itemData) {
        if (this.itemRenderer) {
            return this.itemRenderer(itemData);
        }
        return itemData[this.displayField];
    }
    /**
     * @return {?}
     */
    onClearContent() {
        this.onChange('');
        this.inputElement.nativeElement.value = '';
        this.picker.hide();
    }
    /**
     * @param {?} item
     * @return {?}
     */
    onSelect(item) {
        this.onChange(item[this.valueField]);
        this.inputElement.nativeElement.value = item[this.displayField];
        this.picker.toggle();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        /** @type {?} */
        const theValue = value == null ? '' : value;
        /** @type {?} */
        const tmp = this.dataItems.find((/**
         * @param {?} c
         * @return {?}
         */
        c => c[this.valueField] === theValue));
        if (tmp) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.inputElement.nativeElement.value = tmp[this.displayField];
            }), 0);
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.inputElement.nativeElement.value = '';
            }), 0);
        }
        this.onChange(value);
    }
}
TaminFieldComboBoxStaticComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-combo-box-static',
                template: "<div class=\"input-wrapper\">\n  <input\n    id=\"textbox\"\n    readonly=\"readonly\"\n    [disabled]=\"control.disabled\"\n    style=\"padding-left: 35px;;cursor: pointer\"\n    type=\"text\"\n    class=\"tamin-form-control-input\"\n    [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n    [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n    (click)=\"onShowPicker()\" [attr.placeholder]=\"placeHolder\"\n    #inputElement>\n  <span class=\"icon\" style=\"margin-left:20px\" *ngIf=\"showClearButton\" (click)=\"onClearContent()\">\n    <i\n      class=\"icon-cancel\"\n      style=\"font-size: 1.5em\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"></i>\n  </span>\n  <span class=\"icon\" #icon (click)=\"onShowPicker()\">\n    <i\n      class=\"icon-down-open\"\n      style=\"font-size: 1.5em\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"></i>\n  </span>\n</div>\n<tamin-picker-container #picker>\n  <div *ngFor=\"let item of dataItems\" class=\"item\" (click)=\"onSelect(item)\" [innerHtml]=\"renderItem(item)\"></div>\n</tamin-picker-container>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldComboBoxStaticComponent)),
                        multi: true
                    }
                ],
                styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
            }] }
];
TaminFieldComboBoxStaticComponent.propDecorators = {
    displayField: [{ type: Input }],
    valueField: [{ type: Input }],
    itemRenderer: [{ type: Input }],
    showClearButton: [{ type: Input }],
    picker: [{ type: ViewChild, args: ['picker',] }],
    iconElement: [{ type: ViewChild, args: ['icon',] }],
    dataItems: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.displayField;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.valueField;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.itemRenderer;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.showClearButton;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.picker;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.iconElement;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.dataItems;
}
