// tamin-framework/esm5/lib/components/tamin-process-indicator/tamin-process-indicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var TaminProcessIndicatorComponent = /** @class */ (function () {
    function TaminProcessIndicatorComponent() {
        this.steps = [];
        this.currentStep = 0;
    }
    /**
     * @return {?}
     */
    TaminProcessIndicatorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TaminProcessIndicatorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-process-indicator',
                    template: "<div class=\"process-container\">\n  <ng-container *ngFor=\"let item of steps; let idx = index\">\n    <span class=\"process-item\" [class.process-current]=\"idx <= currentStep\">{{item}}</span>\n    <i class=\"icon-left-open process-item\"  [class.process-current]=\"idx < currentStep\" *ngIf=\"idx !== steps.length-1\"></i>\n  </ng-container>\n</div>\n",
                    styles: [".process-container{display:flex;flex-direction:row;flex-flow:wrap;padding:10px;border:1px solid #cacaca;color:#bbb}.process-item{padding:5px;white-space:nowrap}.process-current{font-weight:700;color:#000}"]
                }] }
    ];
    /** @nocollapse */
    TaminProcessIndicatorComponent.ctorParameters = function () { return []; };
    TaminProcessIndicatorComponent.propDecorators = {
        steps: [{ type: Input }],
        currentStep: [{ type: Input }]
    };
    return TaminProcessIndicatorComponent;
}());
export { TaminProcessIndicatorComponent };
if (false) {
    /** @type {?} */
    TaminProcessIndicatorComponent.prototype.steps;
    /** @type {?} */
    TaminProcessIndicatorComponent.prototype.currentStep;
}
