// tamin-framework/esm2015/lib/helpers/guid.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class Guid {
    /**
     * @private
     * @param {?} guid
     */
    constructor(guid) {
        if (!guid) {
            throw new TypeError('Invalid argument; `value` has no value.');
        }
        this.value = Guid.EMPTY;
        if (guid && Guid.isGuid(guid)) {
            this.value = guid;
        }
    }
    /**
     * @param {?} guid
     * @return {?}
     */
    static isGuid(guid) {
        /** @type {?} */
        const value = guid.toString();
        return guid && (guid instanceof Guid || Guid.validator.test(value));
    }
    /**
     * @return {?}
     */
    static create() {
        return new Guid([Guid.gen(2), Guid.gen(1), Guid.gen(1), Guid.gen(1), Guid.gen(3)].join('-'));
    }
    /**
     * @return {?}
     */
    static createEmpty() {
        return new Guid('emptyguid');
    }
    /**
     * @param {?} guid
     * @return {?}
     */
    static parse(guid) {
        return new Guid(guid);
    }
    /**
     * @return {?}
     */
    static raw() {
        return [Guid.gen(2), Guid.gen(1), Guid.gen(1), Guid.gen(1), Guid.gen(3)].join('-');
    }
    /**
     * @private
     * @param {?} count
     * @return {?}
     */
    static gen(count) {
        /** @type {?} */
        let out = '';
        for (let i = 0; i < count; i++) {
            // tslint:disable-next-line:no-bitwise
            out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return out;
    }
    /**
     * @param {?} other
     * @return {?}
     */
    equals(other) {
        // Comparing string `value` against provided `guid` will auto-call
        // toString on `guid` for comparison
        return Guid.isGuid(other) && this.value === other.toString();
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return this.value === Guid.EMPTY;
    }
    /**
     * @return {?}
     */
    toString() {
        return this.value;
    }
    /**
     * @return {?}
     */
    toJSON() {
        return {
            value: this.value,
        };
    }
}
Guid.validator = new RegExp('^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$', 'i');
Guid.EMPTY = '00000000-0000-0000-0000-000000000000';
if (false) {
    /** @type {?} */
    Guid.validator;
    /** @type {?} */
    Guid.EMPTY;
    /**
     * @type {?}
     * @private
     */
    Guid.prototype.value;
}
