// tamin-framework/esm5/lib/helpers/tamin.data.grid.configuration/tamin.data.grid.configuration-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { TaminDataGridConfiguration } from './tamin.data.grid.configuration';
var TaminDataGridConfigurationFactory = /** @class */ (function () {
    function TaminDataGridConfigurationFactory() {
        this._data = new TaminDataGridConfiguration();
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} url
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.addUrl = /**
     * @template THIS
     * @this {THIS}
     * @param {?} url
     * @return {THIS}
     */
    function (url) {
        (/** @type {?} */ (this))._data.serviceUrl = url;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.addVisibleColumn = /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    function (column) {
        (/** @type {?} */ (this))._data.visibleColumns.push((/** @type {?} */ (column)));
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.clearVisibleColumns = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this))._data.visibleColumns = [];
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.addActionColumn = /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    function (column) {
        (/** @type {?} */ (this))._data.actionColumns.push((/** @type {?} */ (column)));
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.clearActionColumns = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this))._data.actionColumns = [];
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.addSearchParam = /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    function (param) {
        (/** @type {?} */ (this))._data.searchParams.push((/** @type {?} */ (param)));
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.clearSearchParams = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this))._data.searchParams = [];
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.addSortParam = /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    function (param) {
        (/** @type {?} */ (this))._data.sortParams.push((/** @type {?} */ (param)));
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.clearSortParams = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this))._data.sortParams = [];
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} viewType
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setViewType = /**
     * @template THIS
     * @this {THIS}
     * @param {?} viewType
     * @return {THIS}
     */
    function (viewType) {
        (/** @type {?} */ (this))._data.viewType = viewType;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setShowActionColumn = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.showActionColumn = val;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} caption
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setActionColumnCaption = /**
     * @template THIS
     * @this {THIS}
     * @param {?} caption
     * @return {THIS}
     */
    function (caption) {
        (/** @type {?} */ (this))._data.actionColumnCaption = caption;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} id
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setId = /**
     * @template THIS
     * @this {THIS}
     * @param {?} id
     * @return {THIS}
     */
    function (id) {
        (/** @type {?} */ (this))._data.id = id;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setRowEditable = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.rowEditable = val.toString();
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setRowDeletable = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.rowDeletable = val.toString();
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} items
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setDataItems = /**
     * @template THIS
     * @this {THIS}
     * @param {?} items
     * @return {THIS}
     */
    function (items) {
        (/** @type {?} */ (this))._data.dataItems = items;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setShowFooter = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.showFooter = val.toString();
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setShowPager = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.showPager = val;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setPagerSize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.pagerSize = val;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setFirstLoad = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.loadDataOnInit = val;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    TaminDataGridConfigurationFactory.prototype.setPagerCurrentPage = /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    function (val) {
        (/** @type {?} */ (this))._data.pagerCurrentPage = val;
        return (/** @type {?} */ (this));
    };
    /**
     * @return {?}
     */
    TaminDataGridConfigurationFactory.prototype.getData = /**
     * @return {?}
     */
    function () {
        return this._data;
    };
    return TaminDataGridConfigurationFactory;
}());
export { TaminDataGridConfigurationFactory };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminDataGridConfigurationFactory.prototype._data;
}
