// tamin-framework/esm2015/lib/components/fields/tamin-field-validator/tamin-field-validator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldValidatorComponent {
    /**
     * @param {?} controlContainer
     */
    constructor(controlContainer) {
        this.controlContainer = controlContainer;
    }
    /**
     * @param {?} errorMessages
     * @return {?}
     */
    set errorMessages(errorMessages) {
        // errorMessages.forEach((item) => {
        //   const temp = (<ValidationMessageModel[]>this._errorMessages).find(c => c.name === item.name);
        //   if (!temp) {
        //     this._errorMessages.push(item);
        //   } else {
        //     temp.value = item.value;
        //   }
        // });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.configureErrorMessages();
        this._control = this.controlContainer.control.get(this.formControlName);
    }
    /**
     * @return {?}
     */
    get control() {
        return this._control;
    }
    /**
     * @return {?}
     */
    get errorMessage() {
        if (!this._control.valid) {
            for (const propertyName in this._control.errors) {
                if (this._control.errors.hasOwnProperty(propertyName)) {
                    return ((/** @type {?} */ (this._errorMessages))).find((/**
                     * @param {?} p
                     * @return {?}
                     */
                    p => p.name === propertyName)).value;
                }
            }
        }
        return '';
    }
    /**
     * @return {?}
     */
    configureErrorMessages() {
        this._errorMessages = [
            { name: 'required', value: 'مقدار فیلد الزامی می باشد' },
            { name: 'pattern', value: 'الگوی مقدار صحیح نمی باشد' },
            { name: 'email', value: 'ایمیل نادرست است' },
            { name: 'minlength', value: 'مقدار فیلد از حد مجاز کمتر است' },
            { name: 'maxlength', value: 'مقدار فیلد از حد مجاز بیشتر است' },
            { name: 'min', value: 'حداقل مقدار رعایت نشده است' },
            { name: 'max', value: 'حداکثر مقدار رعایت نشده است' },
            { name: 'requiredTrue', value: '' },
            { name: 'nullValidator', value: '' },
            { name: 'compose', value: '' },
            { name: 'composeAsync', value: '' },
            { name: 'requiredNumber', value: 'مقدار فیلد الزامی می باشد' },
            { name: 'persianDate', value: 'تاریخ نادرست است' },
            { name: 'nationalId', value: 'کد ملی نادرست است' }
        ];
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
    }
}
TaminFieldValidatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-validator',
                template: "<span *ngIf=\"control?.touched && !control?.valid\" class=\"tamin-form-control-error\">\n  <span style=\"color: #b00020;\">{{errorMessage}}</span>\n</span>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldValidatorComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
TaminFieldValidatorComponent.ctorParameters = () => [
    { type: ControlContainer }
];
TaminFieldValidatorComponent.propDecorators = {
    formControlName: [{ type: Input }],
    errorMessages: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldValidatorComponent.prototype._errorMessages;
    /** @type {?} */
    TaminFieldValidatorComponent.prototype.formControlName;
    /** @type {?} */
    TaminFieldValidatorComponent.prototype._control;
    /**
     * @type {?}
     * @private
     */
    TaminFieldValidatorComponent.prototype.controlContainer;
}
