// tamin-framework/esm2015/lib/components/tamin-document-viewer/tamin-document-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, Input, ViewChild } from '@angular/core';
import { PdfViewerComponent } from './pdf-viewer/pdf-viewer.component';
import { ImageViewerComponent } from './image-viewer/image-viewer.component';
export class TaminDocumentViewerComponent {
    /**
     * @param {?} changeDetectorRef
     */
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.width = '100%';
        this.height = '600px';
        this.pdfJsPath = 'assets/pdfjs';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadPdf(url) {
        this.viewerType = 'pdf';
        this.changeDetectorRef.detectChanges();
        this.pdfViewer.open(url);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadPdfBased64(url) {
        this.viewerType = 'pdf';
        this.changeDetectorRef.detectChanges();
        this.pdfViewer.openBased64(url);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadImage(url) {
        this.viewerType = 'image';
        this.changeDetectorRef.detectChanges();
        this.imageViewer.open(url);
    }
}
TaminDocumentViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-document-viewer',
                template: "<div [ngStyle]=\"{'width':width, 'height':height, 'text-align': 'center', 'border':'solid 1px #cccccc'}\">\n  <tamin-pdf-viewer2 #pdfViewer [width]=\"width\" [height]=\"height\" *ngIf=\"viewerType === 'pdf'\">\n  </tamin-pdf-viewer2>\n  <tamin-image-viewer #imageViewer [width]=\"width\" [height]=\"height\" *ngIf=\"viewerType === 'image'\">\n  </tamin-image-viewer>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
TaminDocumentViewerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
TaminDocumentViewerComponent.propDecorators = {
    pdfViewer: [{ type: ViewChild, args: ['pdfViewer',] }],
    imageViewer: [{ type: ViewChild, args: ['imageViewer',] }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    pdfJsPath: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.pdfViewer;
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.imageViewer;
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.width;
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.height;
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.pdfJsPath;
    /** @type {?} */
    TaminDocumentViewerComponent.prototype.viewerType;
    /**
     * @type {?}
     * @private
     */
    TaminDocumentViewerComponent.prototype.changeDetectorRef;
}
