// tamin-framework/esm2015/lib/components/fields/tamin-field-file-upload/tamin-field-file-upload.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldFileUploadComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.focus = new EventEmitter();
        this.blur = new EventEmitter();
    }
    /**
     * @return {?}
     */
    onShowPicker() {
        this.fileInput.nativeElement.click();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeFileUpload(event) {
        this.inputElement.nativeElement.value = event.target.files[0].name;
        this.onChange(event.target.files[0]);
    }
}
TaminFieldFileUploadComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-file-upload',
                template: "<div class=\"input-wrapper\">\n  <input\n    id=\"textbox\"\n    style=\"padding-left: 35px;\"\n    type=\"text\"\n    [disabled]=\"control.disabled\"\n    class=\"tamin-form-control-input\"\n    [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n    [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n    #inputElement>\n  <span class=\"icon\" #icon>\n    <i\n      class=\"icon-folder-empty\"\n      style=\"font-size: 1.5em\"\n      (click)=\"onShowPicker()\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"\n      #iconElement></i>\n  </span>\n</div>\n\n<input type=\"file\" #fileInput style=\"display:none\" (change)=\"changeFileUpload($event)\" [attr.accept]=\"accept\"/>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldFileUploadComponent)),
                        multi: true
                    }
                ],
                styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
            }] }
];
TaminFieldFileUploadComponent.propDecorators = {
    maxLength: [{ type: Input }],
    accept: [{ type: Input }],
    focus: [{ type: Output }],
    blur: [{ type: Output }],
    fileInput: [{ type: ViewChild, args: ['fileInput',] }]
};
if (false) {
    /** @type {?} */
    TaminFieldFileUploadComponent.prototype.maxLength;
    /** @type {?} */
    TaminFieldFileUploadComponent.prototype.accept;
    /** @type {?} */
    TaminFieldFileUploadComponent.prototype.focus;
    /** @type {?} */
    TaminFieldFileUploadComponent.prototype.blur;
    /** @type {?} */
    TaminFieldFileUploadComponent.prototype.fileInput;
}
