// tamin-framework/esm2015/lib/components/tamin-document-viewer/pdf-viewer/pdf-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { TaminLazyLoadService } from '../../../services/tamin-lazy-load/tamin-lazy-load.service';
import { Subscription } from 'rxjs';
import { DomSanitizer } from '@angular/platform-browser';
export class PdfViewerComponent {
    /**
     * @param {?} taminLazyLoadService
     * @param {?} domSanitizer
     */
    constructor(taminLazyLoadService, domSanitizer) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.domSanitizer = domSanitizer;
        this.baseUrl = 'assets/pdfjs/viewer.html';
        // viewerUrl: any;
        this.subscription = new Subscription();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} b64Data
     * @param {?} contentType
     * @param {?=} sliceSize
     * @return {?}
     */
    based64toBlob(b64Data, contentType, sliceSize = 512) {
        contentType = contentType || '';
        sliceSize = sliceSize || 512;
        /** @type {?} */
        const byteCharacters = atob(b64Data);
        /** @type {?} */
        const byteArrays = [];
        for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
            /** @type {?} */
            const slice = byteCharacters.slice(offset, offset + sliceSize);
            /** @type {?} */
            const byteNumbers = new Array(slice.length);
            for (let i = 0; i < slice.length; i++) {
                byteNumbers[i] = slice.charCodeAt(i);
            }
            /** @type {?} */
            const byteArray = new Uint8Array(byteNumbers);
            byteArrays.push(byteArray);
        }
        return new Blob(byteArrays, { type: contentType });
    }
    /**
     * @param {?} data
     * @return {?}
     */
    openBased64(data) {
        /** @type {?} */
        const tmp = URL.createObjectURL(this.based64toBlob(data, 'application/pdf'));
        this.open(tmp);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    open(url) {
        this.iframe.nativeElement.src = `${this.baseUrl}?file=${url}`;
    }
}
PdfViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-pdf-viewer2',
                template: "<iframe frameborder=\"0\" [ngStyle]=\"{'width':width, 'height':height}\" #iframe>\n</iframe>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
PdfViewerComponent.ctorParameters = () => [
    { type: TaminLazyLoadService },
    { type: DomSanitizer }
];
PdfViewerComponent.propDecorators = {
    iframe: [{ type: ViewChild, args: ['iframe',] }],
    width: [{ type: Input }],
    height: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PdfViewerComponent.prototype.iframe;
    /** @type {?} */
    PdfViewerComponent.prototype.width;
    /** @type {?} */
    PdfViewerComponent.prototype.height;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.taminLazyLoadService;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.domSanitizer;
}
