// tamin-framework/esm5/lib/components/tamin-wizard/tamin-wizard/tamin-wizard-item/tamin-wizard-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
var TaminWizardItemComponent = /** @class */ (function () {
    function TaminWizardItemComponent() {
    }
    Object.defineProperty(TaminWizardItemComponent.prototype, "visible", {
        get: /**
         * @return {?}
         */
        function () {
            return this._visible;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminWizardItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.hide();
    };
    /**
     * @return {?}
     */
    TaminWizardItemComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        this.wizardItem.nativeElement.style.display = 'block';
        this._visible = true;
    };
    /**
     * @return {?}
     */
    TaminWizardItemComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.wizardItem.nativeElement.style.display = 'none';
        this._visible = false;
    };
    TaminWizardItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-wizard-item',
                    template: "<div #wizardItem class=\"wizard-item\">\r\n  <ng-content></ng-content>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminWizardItemComponent.ctorParameters = function () { return []; };
    TaminWizardItemComponent.propDecorators = {
        stepTitle: [{ type: Input }],
        buttonTitle: [{ type: Input }],
        wizardItem: [{ type: ViewChild, args: ['wizardItem',] }]
    };
    return TaminWizardItemComponent;
}());
export { TaminWizardItemComponent };
if (false) {
    /** @type {?} */
    TaminWizardItemComponent.prototype.stepTitle;
    /** @type {?} */
    TaminWizardItemComponent.prototype.buttonTitle;
    /** @type {?} */
    TaminWizardItemComponent.prototype.wizardItem;
    /**
     * @type {?}
     * @private
     */
    TaminWizardItemComponent.prototype._visible;
}
