// tamin-framework/esm5/lib/services/platform.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Platform } from '@angular/cdk/platform';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/platform";
var PlatformService = /** @class */ (function () {
    function PlatformService(platform) {
        this.platform = platform;
    }
    /**
     * @return {?}
     */
    PlatformService.prototype.isAndroid = /**
     * @return {?}
     */
    function () {
        return this.platform.ANDROID;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isIOS = /**
     * @return {?}
     */
    function () {
        return this.platform.IOS;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isBlink = /**
     * @return {?}
     */
    function () {
        return this.platform.BLINK;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isEdge = /**
     * @return {?}
     */
    function () {
        return this.platform.EDGE;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isFirefox = /**
     * @return {?}
     */
    function () {
        return this.platform.FIREFOX;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isBrowser = /**
     * @return {?}
     */
    function () {
        return this.platform.isBrowser;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isSafari = /**
     * @return {?}
     */
    function () {
        return this.platform.SAFARI;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isTrident = /**
     * @return {?}
     */
    function () {
        return this.platform.TRIDENT;
    };
    /**
     * @return {?}
     */
    PlatformService.prototype.isWebkit = /**
     * @return {?}
     */
    function () {
        return this.platform.WEBKIT;
    };
    PlatformService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    PlatformService.ctorParameters = function () { return [
        { type: Platform }
    ]; };
    /** @nocollapse */ PlatformService.ngInjectableDef = i0.defineInjectable({ factory: function PlatformService_Factory() { return new PlatformService(i0.inject(i1.Platform)); }, token: PlatformService, providedIn: "root" });
    return PlatformService;
}());
export { PlatformService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PlatformService.prototype.platform;
}
