// tamin-framework/esm2015/lib/components/tamin-wizard/tamin-wizard/tamin-wizard-item/tamin-wizard-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
export class TaminWizardItemComponent {
    constructor() {
    }
    /**
     * @return {?}
     */
    get visible() {
        return this._visible;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.hide();
    }
    /**
     * @return {?}
     */
    show() {
        this.wizardItem.nativeElement.style.display = 'block';
        this._visible = true;
    }
    /**
     * @return {?}
     */
    hide() {
        this.wizardItem.nativeElement.style.display = 'none';
        this._visible = false;
    }
}
TaminWizardItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-wizard-item',
                template: "<div #wizardItem class=\"wizard-item\">\r\n  <ng-content></ng-content>\r\n</div>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
TaminWizardItemComponent.ctorParameters = () => [];
TaminWizardItemComponent.propDecorators = {
    stepTitle: [{ type: Input }],
    buttonTitle: [{ type: Input }],
    wizardItem: [{ type: ViewChild, args: ['wizardItem',] }]
};
if (false) {
    /** @type {?} */
    TaminWizardItemComponent.prototype.stepTitle;
    /** @type {?} */
    TaminWizardItemComponent.prototype.buttonTitle;
    /** @type {?} */
    TaminWizardItemComponent.prototype.wizardItem;
    /**
     * @type {?}
     * @private
     */
    TaminWizardItemComponent.prototype._visible;
}
