// tamin-framework/esm5/lib/components/fields/tamin-field-radio-button-group/tamin-field-radio-button-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, forwardRef, Input, QueryList } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { TaminFieldRadioButtonComponent } from '../tamin-field-radio-button/tamin-field-radio-button.component';
var TaminFieldRadioButtonGroupComponent = /** @class */ (function () {
    function TaminFieldRadioButtonGroupComponent() {
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
        this.onChange = (/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
        });
    }
    /**
     * @return {?}
     */
    TaminFieldRadioButtonGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.radioButtons.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                item.name = _this.formControlName;
                item.change.subscribe((/**
                 * @param {?} value
                 * @return {?}
                 */
                function (value) {
                    _this.onChange(value);
                }));
            }));
        }));
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldRadioButtonGroupComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaminFieldRadioButtonGroupComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TaminFieldRadioButtonGroupComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.radioButtons.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.value === obj) {
                    item.setSelected();
                }
            }));
        }));
    };
    TaminFieldRadioButtonGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-radio-button-group',
                    template: "<ng-container>\n  <ng-content></ng-content>\n</ng-container>\n<span #inputElement></span>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldRadioButtonGroupComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminFieldRadioButtonGroupComponent.ctorParameters = function () { return []; };
    TaminFieldRadioButtonGroupComponent.propDecorators = {
        radioButtons: [{ type: ContentChildren, args: [TaminFieldRadioButtonComponent,] }],
        formControlName: [{ type: Input }]
    };
    return TaminFieldRadioButtonGroupComponent;
}());
export { TaminFieldRadioButtonGroupComponent };
if (false) {
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.radioButtons;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.onChange;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.onTouched;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.formControlName;
}
