// tamin-framework/esm2015/lib/components/fields/tamin-field-label/tamin-field-label.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldLabelComponent {
    /**
     * @param {?} controlContainer
     */
    constructor(controlContainer) {
        this.controlContainer = controlContainer;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.control = this.controlContainer.control.get(this.formControlName);
    }
    /**
     * @return {?}
     */
    get required() {
        if (this.control.validator === null) {
            return false;
        }
        /** @type {?} */
        const validator = this.control.validator((/** @type {?} */ ({})));
        return (validator && (validator.required || validator.requiredNumber));
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
    }
}
TaminFieldLabelComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-label',
                template: "<label class=\"tamin-form-control-label\"\n  [class.tamin-form-control-label]='!required'\n  [class.tamin-form-control-label-required]='required'\n  [class.tamin-form-control-label-invalid]='control.touched && !control.valid'>\n  <ng-content></ng-content>\n</label>\n\n\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldLabelComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
TaminFieldLabelComponent.ctorParameters = () => [
    { type: ControlContainer }
];
TaminFieldLabelComponent.propDecorators = {
    formControlName: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldLabelComponent.prototype.formControlName;
    /** @type {?} */
    TaminFieldLabelComponent.prototype.control;
    /**
     * @type {?}
     * @private
     */
    TaminFieldLabelComponent.prototype.controlContainer;
}
