// tamin-framework/esm5/lib/components/fields/tamin-field-combo-box-static/tamin-field-combo-box-static.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef, Input, ViewChild } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { PickerContainerComponent } from '../../containers/picker-container/picker-container.component';
var TaminFieldComboBoxStaticComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldComboBoxStaticComponent, _super);
    function TaminFieldComboBoxStaticComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.dataItems = [];
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        this.picker.parent = this.inputElement;
        this.picker.button = this.iconElement;
    };
    /**
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.onShowPicker = /**
     * @return {?}
     */
    function () {
        this.picker.toggle();
    };
    /**
     * @param {?} itemData
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.renderItem = /**
     * @param {?} itemData
     * @return {?}
     */
    function (itemData) {
        if (this.itemRenderer) {
            return this.itemRenderer(itemData);
        }
        return itemData[this.displayField];
    };
    /**
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.onClearContent = /**
     * @return {?}
     */
    function () {
        this.onChange('');
        this.inputElement.nativeElement.value = '';
        this.picker.hide();
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.onSelect = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.onChange(item[this.valueField]);
        this.inputElement.nativeElement.value = item[this.displayField];
        this.picker.toggle();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldComboBoxStaticComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        /** @type {?} */
        var theValue = value == null ? '' : value;
        /** @type {?} */
        var tmp = this.dataItems.find((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c[_this.valueField] === theValue; }));
        if (tmp) {
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.inputElement.nativeElement.value = tmp[_this.displayField];
            }), 0);
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.inputElement.nativeElement.value = '';
            }), 0);
        }
        this.onChange(value);
    };
    TaminFieldComboBoxStaticComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-combo-box-static',
                    template: "<div class=\"input-wrapper\">\n  <input\n    id=\"textbox\"\n    readonly=\"readonly\"\n    [disabled]=\"control.disabled\"\n    style=\"padding-left: 35px;;cursor: pointer\"\n    type=\"text\"\n    class=\"tamin-form-control-input\"\n    [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n    [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n    (click)=\"onShowPicker()\" [attr.placeholder]=\"placeHolder\"\n    #inputElement>\n  <span class=\"icon\" style=\"margin-left:20px\" *ngIf=\"showClearButton\" (click)=\"onClearContent()\">\n    <i\n      class=\"icon-cancel\"\n      style=\"font-size: 1.5em\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"></i>\n  </span>\n  <span class=\"icon\" #icon (click)=\"onShowPicker()\">\n    <i\n      class=\"icon-down-open\"\n      style=\"font-size: 1.5em\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"></i>\n  </span>\n</div>\n<tamin-picker-container #picker>\n  <div *ngFor=\"let item of dataItems\" class=\"item\" (click)=\"onSelect(item)\" [innerHtml]=\"renderItem(item)\"></div>\n</tamin-picker-container>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldComboBoxStaticComponent; })),
                            multi: true
                        }
                    ],
                    styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
                }] }
    ];
    TaminFieldComboBoxStaticComponent.propDecorators = {
        displayField: [{ type: Input }],
        valueField: [{ type: Input }],
        itemRenderer: [{ type: Input }],
        showClearButton: [{ type: Input }],
        picker: [{ type: ViewChild, args: ['picker',] }],
        iconElement: [{ type: ViewChild, args: ['icon',] }],
        dataItems: [{ type: Input }]
    };
    return TaminFieldComboBoxStaticComponent;
}(FieldBaseComponent));
export { TaminFieldComboBoxStaticComponent };
if (false) {
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.displayField;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.valueField;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.itemRenderer;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.showClearButton;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.picker;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.iconElement;
    /** @type {?} */
    TaminFieldComboBoxStaticComponent.prototype.dataItems;
}
