// tamin-framework/esm2015/lib/components/containers/tamin-tab-panel/tamin-tab-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { TaminTabComponent } from '../tamin-tab/tamin-tab.component';
export class TaminTabPanelComponent {
    constructor() {
        this.currentTab = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        console.log('ngAfterContentInit');
        this.tabComponents.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            if (index === this.currentTab) {
                item.present = true;
            }
        }));
    }
    /**
     * @param {?} tab
     * @param {?} index
     * @return {?}
     */
    onClick(tab, index) {
        this.currentTab = index;
        this.tabComponents.forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => item.present = item === tab));
    }
}
TaminTabPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tab-panel',
                template: "<div class=\"tab\">\r\n  <ng-container *ngFor=\"let tab of tabComponents, let idx = index\">\r\n    <button style=\"font-family: main-font, tahoma, verdana;\"\r\n      *ngIf=\"tab.visible\" class=\"tablinks\" (click)=\"onClick(tab, idx)\" [class.active] =\"idx === currentTab\">{{tab.title}}\r\n    </button>\r\n  </ng-container>\r\n</div>\r\n<ng-content></ng-content>\r\n",
                styles: [".tab{overflow:hidden;border:1px solid #ccc;background-color:#f1f1f1;margin:3px 3px 0}.tab button{background-color:inherit;float:right;border:1px solid #aaa;border-bottom:0;border-radius:10px 10px 0 0;outline:0;cursor:pointer;padding:12px 14px;margin:2px 1px 0;transition:.3s}.tab button:hover{background-color:#e4e4e4}.tab button.active{background-color:#dfefff;border-color:#0af}"]
            }] }
];
/** @nocollapse */
TaminTabPanelComponent.ctorParameters = () => [];
TaminTabPanelComponent.propDecorators = {
    tabComponents: [{ type: ContentChildren, args: [TaminTabComponent,] }],
    currentTab: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminTabPanelComponent.prototype.tabComponents;
    /** @type {?} */
    TaminTabPanelComponent.prototype.currentTab;
}
