// tamin-framework/esm5/lib/components/tamin-document-viewer/pdf-viewer/pdf-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { TaminLazyLoadService } from '../../../services/tamin-lazy-load/tamin-lazy-load.service';
import { Subscription } from 'rxjs';
import { DomSanitizer } from '@angular/platform-browser';
var PdfViewerComponent = /** @class */ (function () {
    function PdfViewerComponent(taminLazyLoadService, domSanitizer) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.domSanitizer = domSanitizer;
        this.baseUrl = 'assets/pdfjs/viewer.html';
        // viewerUrl: any;
        this.subscription = new Subscription();
    }
    /**
     * @return {?}
     */
    PdfViewerComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PdfViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} b64Data
     * @param {?} contentType
     * @param {?=} sliceSize
     * @return {?}
     */
    PdfViewerComponent.prototype.based64toBlob = /**
     * @param {?} b64Data
     * @param {?} contentType
     * @param {?=} sliceSize
     * @return {?}
     */
    function (b64Data, contentType, sliceSize) {
        if (sliceSize === void 0) { sliceSize = 512; }
        contentType = contentType || '';
        sliceSize = sliceSize || 512;
        /** @type {?} */
        var byteCharacters = atob(b64Data);
        /** @type {?} */
        var byteArrays = [];
        for (var offset = 0; offset < byteCharacters.length; offset += sliceSize) {
            /** @type {?} */
            var slice = byteCharacters.slice(offset, offset + sliceSize);
            /** @type {?} */
            var byteNumbers = new Array(slice.length);
            for (var i = 0; i < slice.length; i++) {
                byteNumbers[i] = slice.charCodeAt(i);
            }
            /** @type {?} */
            var byteArray = new Uint8Array(byteNumbers);
            byteArrays.push(byteArray);
        }
        return new Blob(byteArrays, { type: contentType });
    };
    /**
     * @param {?} data
     * @return {?}
     */
    PdfViewerComponent.prototype.openBased64 = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var tmp = URL.createObjectURL(this.based64toBlob(data, 'application/pdf'));
        this.open(tmp);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    PdfViewerComponent.prototype.open = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.iframe.nativeElement.src = this.baseUrl + "?file=" + url;
    };
    PdfViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-pdf-viewer2',
                    template: "<iframe frameborder=\"0\" [ngStyle]=\"{'width':width, 'height':height}\" #iframe>\n</iframe>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PdfViewerComponent.ctorParameters = function () { return [
        { type: TaminLazyLoadService },
        { type: DomSanitizer }
    ]; };
    PdfViewerComponent.propDecorators = {
        iframe: [{ type: ViewChild, args: ['iframe',] }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return PdfViewerComponent;
}());
export { PdfViewerComponent };
if (false) {
    /** @type {?} */
    PdfViewerComponent.prototype.iframe;
    /** @type {?} */
    PdfViewerComponent.prototype.width;
    /** @type {?} */
    PdfViewerComponent.prototype.height;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.baseUrl;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.taminLazyLoadService;
    /**
     * @type {?}
     * @private
     */
    PdfViewerComponent.prototype.domSanitizer;
}
