// tamin-framework/esm5/lib/components/tamin-document-viewer/image-viewer/image-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { TaminLazyLoadService } from '../../../services/tamin-lazy-load/tamin-lazy-load.service';
import { Subscription } from 'rxjs';
var ImageViewerComponent = /** @class */ (function () {
    function ImageViewerComponent(taminLazyLoadService) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.width = '100%';
        this.height = '300px';
        this.subscription = new Subscription();
    }
    /**
     * @return {?}
     */
    ImageViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} url
     * @return {?}
     */
    ImageViewerComponent.prototype.open = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        var _this = this;
        this.subscription.add(this.taminLazyLoadService.loadJs('assets/viewerjs/dist/viewer.js').subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.subscription.add(_this.taminLazyLoadService.loadCss('assets/viewerjs/dist/viewer.css').subscribe((/**
             * @param {?} value1
             * @return {?}
             */
            function (value1) {
                _this.img.nativeElement.src = url;
                /** @type {?} */
                var viewer = new Viewer(_this.img.nativeElement, {
                    inline: false,
                    viewed: /**
                     * @return {?}
                     */
                    function () {
                        viewer.zoomTo(1);
                    },
                });
            })));
        })));
    };
    /**
     * @return {?}
     */
    ImageViewerComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscription.unsubscribe();
    };
    ImageViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-image-viewer',
                    template: "<img alt=\"image\" #img style=\"max-width:100%;max-height:100%;cursor:pointer\">\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ImageViewerComponent.ctorParameters = function () { return [
        { type: TaminLazyLoadService }
    ]; };
    ImageViewerComponent.propDecorators = {
        img: [{ type: ViewChild, args: ['img',] }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return ImageViewerComponent;
}());
export { ImageViewerComponent };
if (false) {
    /** @type {?} */
    ImageViewerComponent.prototype.img;
    /** @type {?} */
    ImageViewerComponent.prototype.width;
    /** @type {?} */
    ImageViewerComponent.prototype.height;
    /**
     * @type {?}
     * @private
     */
    ImageViewerComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    ImageViewerComponent.prototype.taminLazyLoadService;
}
