// tamin-framework/esm2015/lib/helpers/tamin.data.grid.configuration/tamin.data.grid.configuration-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { TaminDataGridConfiguration } from './tamin.data.grid.configuration';
export class TaminDataGridConfigurationFactory {
    constructor() {
        this._data = new TaminDataGridConfiguration();
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} url
     * @return {THIS}
     */
    addUrl(url) {
        (/** @type {?} */ (this))._data.serviceUrl = url;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    addVisibleColumn(column) {
        (/** @type {?} */ (this))._data.visibleColumns.push((/** @type {?} */ (column)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    clearVisibleColumns() {
        (/** @type {?} */ (this))._data.visibleColumns = [];
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} column
     * @return {THIS}
     */
    addActionColumn(column) {
        (/** @type {?} */ (this))._data.actionColumns.push((/** @type {?} */ (column)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    clearActionColumns() {
        (/** @type {?} */ (this))._data.actionColumns = [];
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    addSearchParam(param) {
        (/** @type {?} */ (this))._data.searchParams.push((/** @type {?} */ (param)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    clearSearchParams() {
        (/** @type {?} */ (this))._data.searchParams = [];
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} param
     * @return {THIS}
     */
    addSortParam(param) {
        (/** @type {?} */ (this))._data.sortParams.push((/** @type {?} */ (param)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    clearSortParams() {
        (/** @type {?} */ (this))._data.sortParams = [];
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} viewType
     * @return {THIS}
     */
    setViewType(viewType) {
        (/** @type {?} */ (this))._data.viewType = viewType;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setShowActionColumn(val) {
        (/** @type {?} */ (this))._data.showActionColumn = val;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} caption
     * @return {THIS}
     */
    setActionColumnCaption(caption) {
        (/** @type {?} */ (this))._data.actionColumnCaption = caption;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} id
     * @return {THIS}
     */
    setId(id) {
        (/** @type {?} */ (this))._data.id = id;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setRowEditable(val) {
        (/** @type {?} */ (this))._data.rowEditable = val.toString();
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setRowDeletable(val) {
        (/** @type {?} */ (this))._data.rowDeletable = val.toString();
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} items
     * @return {THIS}
     */
    setDataItems(items) {
        (/** @type {?} */ (this))._data.dataItems = items;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setShowFooter(val) {
        (/** @type {?} */ (this))._data.showFooter = val.toString();
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setShowPager(val) {
        (/** @type {?} */ (this))._data.showPager = val;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setPagerSize(val) {
        (/** @type {?} */ (this))._data.pagerSize = val;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setFirstLoad(val) {
        (/** @type {?} */ (this))._data.loadDataOnInit = val;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} val
     * @return {THIS}
     */
    setPagerCurrentPage(val) {
        (/** @type {?} */ (this))._data.pagerCurrentPage = val;
        return (/** @type {?} */ (this));
    }
    /**
     * @return {?}
     */
    getData() {
        return this._data;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminDataGridConfigurationFactory.prototype._data;
}
