// tamin-framework/esm2015/lib/components/fields/tamin-field-text-area/tamin-field-text-area.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldTextAreaComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.rows = 4;
        this.cols = 50;
    }
    /**
     * @return {?}
     */
    initializeComponent() {
    }
}
TaminFieldTextAreaComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-text-area',
                template: "<textarea\n  style=\"resize: none;height: 7em\"\n  class=\"tamin-form-control-input\"\n  [disabled]=\"control.disabled\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [attr.rows]=\"rows\"\n  [attr.cols]=\"cols\"\n  #inputElement></textarea>\n<span>{{hint}}</span>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldTextAreaComponent)),
                        multi: true
                    }
                ],
                styles: [".tamin-form-control-input{height:inherit!important}"]
            }] }
];
TaminFieldTextAreaComponent.propDecorators = {
    rows: [{ type: Input }],
    cols: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldTextAreaComponent.prototype.rows;
    /** @type {?} */
    TaminFieldTextAreaComponent.prototype.cols;
}
