// tamin-framework/esm5/lib/components/tamin-data-grid/tamin-data-pager/tamin-data-pager.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var TaminDataPagerComponent = /** @class */ (function () {
    function TaminDataPagerComponent() {
        this.pageChanged = new EventEmitter();
        this.refreshData = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TaminDataPagerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.pagerCurrentPage > this.pagerTotal) {
            this.pagerCurrentPage = this.pagerTotal;
        }
        else if (this.pagerCurrentPage < 1) {
            this.pagerCurrentPage = 1;
        }
    };
    /**
     * @param {?} changeAmount
     * @return {?}
     */
    TaminDataPagerComponent.prototype.changePage = /**
     * @param {?} changeAmount
     * @return {?}
     */
    function (changeAmount) {
        /** @type {?} */
        var oldPage = this.pagerCurrentPage;
        if (changeAmount > 1) {
            this.pagerCurrentPage = this.pagerTotal;
        }
        else if (changeAmount < -1) {
            this.pagerCurrentPage = 1;
        }
        else {
            this.pagerCurrentPage = parseFloat(this.pagerCurrentPage.toString()) + parseFloat(changeAmount.toString());
            if (this.pagerCurrentPage < 1) {
                this.pagerCurrentPage = 1;
            }
            if (this.pagerCurrentPage > this.pagerTotal) {
                this.pagerCurrentPage = this.pagerTotal;
            }
        }
        if (oldPage !== this.pagerCurrentPage) {
            this.pageChanged.emit(this.pagerCurrentPage);
        }
    };
    /**
     * @return {?}
     */
    TaminDataPagerComponent.prototype.onRefresh = /**
     * @return {?}
     */
    function () {
        this.refreshData.emit();
    };
    TaminDataPagerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-data-pager',
                    template: "<table class=\"pager\">\n  <tbody>\n  <tr>\n    <td>\n      <button type=\"button\" class=\"button pager icon-to-end\" (click)=\"changePage(-2)\"></button>\n    </td>\n    <td>\n      <button type=\"button\" class=\"button pager icon-play\" (click)=\"changePage(-1)\"></button>\n    </td>\n    <td class=\"pagerText\">\n      <span>\u0635\u0641\u062D\u0647</span> {{pagerCurrentPage | persianNumber}} <span>\u0627\u0632</span> {{this.pagerTotal | persianNumber}}\n    </td>\n    <td>\n      <button type=\"button\" class=\"button pager icon-play\" style=\"transform: rotate(180deg);\" (click)=\"changePage(1)\"></button>\n    </td>\n    <td>\n      <button type=\"button\" class=\"button pager icon-to-start\" (click)=\"changePage(2)\"></button>\n    </td>\n    <td>\n      <button type=\"button\" class=\"button pager icon-arrows-cw\" (click)=\"onRefresh()\"></button>\n    </td>\n  </tr>\n  </tbody>\n</table>\n",
                    styles: [".pager{border-spacing:0;border:none;text-align:right}.pager td{border:none}.pagerText{padding:5px;font-size:11px}.button.pager{text-align:center;vertical-align:middle;padding:3px 3px 0;color:#333;background-color:transparent}.button.pager:hover{color:#057;background-color:#eaeaea}"]
                }] }
    ];
    /** @nocollapse */
    TaminDataPagerComponent.ctorParameters = function () { return []; };
    TaminDataPagerComponent.propDecorators = {
        pagerSize: [{ type: Input }],
        pagerCurrentPage: [{ type: Input }],
        pagerTotal: [{ type: Input }],
        pageChanged: [{ type: Output }],
        refreshData: [{ type: Output }]
    };
    return TaminDataPagerComponent;
}());
export { TaminDataPagerComponent };
if (false) {
    /** @type {?} */
    TaminDataPagerComponent.prototype.pagerSize;
    /** @type {?} */
    TaminDataPagerComponent.prototype.pagerCurrentPage;
    /** @type {?} */
    TaminDataPagerComponent.prototype.pagerTotal;
    /** @type {?} */
    TaminDataPagerComponent.prototype.pageChanged;
    /** @type {?} */
    TaminDataPagerComponent.prototype.refreshData;
}
