// tamin-framework/esm5/lib/components/fields/tamin-field-static-text/tamin-field-static-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldStaticTextComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldStaticTextComponent, _super);
    function TaminFieldStaticTextComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    TaminFieldStaticTextComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldStaticTextComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var theValue = value == null ? '' : value;
            _this.inputElement.nativeElement.value = _this.persianNumberPipe.transform(value);
            _this.onChange(value);
        }), 0);
    };
    TaminFieldStaticTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-static-text',
                    template: "\n<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  readonly=\"readonly\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  #inputElement>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldStaticTextComponent; })),
                            multi: true
                        }
                    ],
                    styles: [":host{width:100%}"]
                }] }
    ];
    return TaminFieldStaticTextComponent;
}(FieldBaseComponent));
export { TaminFieldStaticTextComponent };
