// tamin-framework/esm5/lib/components/tamin-pdf-viewer/tamin-pdf-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { timeout } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { TaminModalComponent } from '../tamin-modal/tamin-modal.component';
var TaminPdfViewerComponent = /** @class */ (function () {
    function TaminPdfViewerComponent(http) {
        this.http = http;
        this.errorMessage = 'خطا در دریافت اطلاعات';
        this.beforeLoad = new EventEmitter();
        this.afterLoad = new EventEmitter();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    TaminPdfViewerComponent.prototype.getBlob = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(url, { responseType: 'blob' })
                .pipe(timeout(6000))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(error); }));
        }));
    };
    /**
     * @param {?} url
     * @return {?}
     */
    TaminPdfViewerComponent.prototype.open = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        var _this = this;
        this.beforeLoad.emit();
        this.getBlob(url)
            .then((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.iframe.nativeElement.src = URL.createObjectURL(value);
            _this.afterLoad.emit();
            _this.theModal.show();
        }))
            .catch((/**
         * @param {?} reason
         * @return {?}
         */
        function (reason) {
            _this.afterLoad.emit();
        }));
    };
    /**
     * @return {?}
     */
    TaminPdfViewerComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.theModal.hide();
    };
    /**
     * @return {?}
     */
    TaminPdfViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TaminPdfViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-pdf-viewer',
                    template: "<tamin-modal #theModal [title]=\"title\">\n  <iframe #iframe style=\"border:none; width:100%; height:80vh\"></iframe>\n  <ng-content></ng-content>\n</tamin-modal>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminPdfViewerComponent.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    TaminPdfViewerComponent.propDecorators = {
        theModal: [{ type: ViewChild, args: ['theModal',] }],
        iframe: [{ type: ViewChild, args: ['iframe',] }],
        title: [{ type: Input }],
        errorMessage: [{ type: Input }],
        beforeLoad: [{ type: Output }],
        afterLoad: [{ type: Output }]
    };
    return TaminPdfViewerComponent;
}());
export { TaminPdfViewerComponent };
if (false) {
    /** @type {?} */
    TaminPdfViewerComponent.prototype.theModal;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.iframe;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.title;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.errorMessage;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.beforeLoad;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.afterLoad;
    /**
     * @type {?}
     * @private
     */
    TaminPdfViewerComponent.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    TaminPdfViewerComponent.prototype.http;
}
