// tamin-framework/esm5/lib/components/tamin-tab/tab-body/tab-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, TemplateRef, ViewChild } from '@angular/core';
var TaminTabBodyComponent = /** @class */ (function () {
    function TaminTabBodyComponent() {
    }
    /**
     * @return {?}
     */
    TaminTabBodyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TaminTabBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tab-body',
                    template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminTabBodyComponent.ctorParameters = function () { return []; };
    TaminTabBodyComponent.propDecorators = {
        bodyContent: [{ type: ViewChild, args: [TemplateRef,] }]
    };
    return TaminTabBodyComponent;
}());
export { TaminTabBodyComponent };
if (false) {
    /** @type {?} */
    TaminTabBodyComponent.prototype.bodyContent;
}
