// tamin-framework/esm2015/lib/components/fields/tamin-field-static-text/tamin-field-static-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldStaticTextComponent extends FieldBaseComponent {
    /**
     * @return {?}
     */
    initializeComponent() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const theValue = value == null ? '' : value;
            this.inputElement.nativeElement.value = this.persianNumberPipe.transform(value);
            this.onChange(value);
        }), 0);
    }
}
TaminFieldStaticTextComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-static-text',
                template: "\n<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  readonly=\"readonly\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  #inputElement>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldStaticTextComponent)),
                        multi: true
                    }
                ],
                styles: [":host{width:100%}"]
            }] }
];
