// tamin-framework/esm5/lib/components/tamin-accordion-group/tamin-accordion-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
*/
import { Component, ContentChildren, QueryList } from '@angular/core';
import { TaminAccordionComponent } from '../tamin-accordion/tamin-accordion.component';
var TaminAccordionGroupComponent = /** @class */ (function () {
    function TaminAccordionGroupComponent() {
        this.subscriptions = [];
    }
    /**
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._accordions = this.accordions;
        this.removeSubscriptions();
        this.addSubscriptions();
        this.accordions.changes.subscribe((/**
         * @param {?} rex
         * @return {?}
         */
        function (rex) {
            _this._accordions = rex;
            _this.removeSubscriptions();
            _this.addSubscriptions();
        }));
    };
    /**
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.addSubscriptions = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._accordions.forEach((/**
         * @param {?} a
         * @return {?}
         */
        function (a) {
            /** @type {?} */
            var subscription = a.toggleAccordion.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                _this.toogleAccordion(a);
            }));
            _this.subscriptions.push(subscription);
        }));
    };
    /**
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.removeSubscriptions = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) {
            sub.unsubscribe();
        }));
    };
    /**
     * @param {?} accordion
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.toogleAccordion = /**
     * @param {?} accordion
     * @return {?}
     */
    function (accordion) {
        if (!accordion.active) {
            this.accordions.forEach((/**
             * @param {?} a
             * @return {?}
             */
            function (a) { return a.active = false; }));
        }
        // set active accordion
        accordion.active = !accordion.active;
    };
    /**
     * @return {?}
     */
    TaminAccordionGroupComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeSubscriptions();
    };
    TaminAccordionGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-accordion-group',
                    template: "<!--\r\n  Created by: k_kaviani\r\n-->\r\n\r\n<ng-content></ng-content>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TaminAccordionGroupComponent.ctorParameters = function () { return []; };
    TaminAccordionGroupComponent.propDecorators = {
        accordions: [{ type: ContentChildren, args: [TaminAccordionComponent,] }]
    };
    return TaminAccordionGroupComponent;
}());
export { TaminAccordionGroupComponent };
if (false) {
    /** @type {?} */
    TaminAccordionGroupComponent.prototype.accordions;
    /**
     * @type {?}
     * @private
     */
    TaminAccordionGroupComponent.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    TaminAccordionGroupComponent.prototype._accordions;
}
