// tamin-framework/esm5/lib/components/tamin-stepper/tamin-stepper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { TaminStepperStepComponent } from './tamin-stepper-step/tamin-stepper-step.component';
import { BreakpointObserver, Breakpoints } from '@angular/cdk/layout';
var TaminStepperComponent = /** @class */ (function () {
    function TaminStepperComponent(breakpointObserver) {
        this.breakpointObserver = breakpointObserver;
        this.breakPoints = [Breakpoints.Small, Breakpoints.Handset];
        this.renderMode = 'desktop';
    }
    Object.defineProperty(TaminStepperComponent.prototype, "activeStepName", {
        get: /**
         * @return {?}
         */
        function () {
            return this._activeStepName;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            this._activeStepName = value;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.steps.forEach((/**
                 * @param {?} item
                 * @param {?} index
                 * @return {?}
                 */
                function (item, index) {
                    item.active = item.name === value;
                }));
            }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaminStepperComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.breakpointObserver
            .observe(this.breakPoints)
            .subscribe((/**
         * @param {?} state
         * @return {?}
         */
        function (state) {
            if (state.matches) {
                if (_this.renderMode !== 'mobile') {
                    _this.renderMode = 'mobile';
                }
            }
            else {
                if (_this.renderMode !== 'desktop') {
                    _this.renderMode = 'desktop';
                }
            }
        }));
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.steps.forEach((/**
             * @param {?} item
             * @param {?} index
             * @return {?}
             */
            function (item, index) {
                item.iconText = (index + 1).toString();
            }));
        }));
    };
    TaminStepperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tamin-stepper',
                    template: "<div\n  [ngClass]=\"{'row stepper-desktop': renderMode === 'desktop', 'stepper-mobile': renderMode === 'mobile'}\">\n  <ng-content></ng-content>\n</div>\n",
                    styles: [".stepper-desktop,.stepper-mobile{position:relative}.stepper-desktop:before{content:'';position:absolute;top:50%;left:0;border-top:1px solid #cacaca;width:100%;z-index:-1}.stepper-mobile:after{content:'';position:absolute;top:0;bottom:0;right:29px;border-left:1px solid #cacaca;z-index:-1}"]
                }] }
    ];
    /** @nocollapse */
    TaminStepperComponent.ctorParameters = function () { return [
        { type: BreakpointObserver }
    ]; };
    TaminStepperComponent.propDecorators = {
        steps: [{ type: ContentChildren, args: [TaminStepperStepComponent,] }],
        activeStepName: [{ type: Input }]
    };
    return TaminStepperComponent;
}());
export { TaminStepperComponent };
if (false) {
    /** @type {?} */
    TaminStepperComponent.prototype.steps;
    /**
     * @type {?}
     * @private
     */
    TaminStepperComponent.prototype._activeStepName;
    /** @type {?} */
    TaminStepperComponent.prototype.breakPoints;
    /** @type {?} */
    TaminStepperComponent.prototype.renderMode;
    /** @type {?} */
    TaminStepperComponent.prototype.breakpointObserver;
}
