// tamin-framework/esm2015/lib/services/tamin-lazy-load/tamin-lazy-load.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { DOCUMENT } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class TaminLazyLoadService {
    /**
     * @param {?} document
     */
    constructor(document) {
        this.document = document;
        this.loadedLibraries = {};
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadJs(url) {
        if (this.loadedLibraries[url]) {
            return this.loadedLibraries[url].asObservable();
        }
        this.loadedLibraries[url] = new ReplaySubject();
        /** @type {?} */
        const script = this.document.createElement('script');
        script.type = 'text/javascript';
        script.src = url;
        script.onload = (/**
         * @return {?}
         */
        () => {
            this.loadedLibraries[url].next('');
            this.loadedLibraries[url].complete();
        });
        this.document.body.appendChild(script);
        return this.loadedLibraries[url].asObservable();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadCss(url) {
        if (this.loadedLibraries[url]) {
            return this.loadedLibraries[url].asObservable();
        }
        this.loadedLibraries[url] = new ReplaySubject();
        /** @type {?} */
        const script = this.document.createElement('link');
        script.rel = 'stylesheet';
        script.href = url;
        script.onload = (/**
         * @return {?}
         */
        () => {
            this.loadedLibraries[url].next('');
            this.loadedLibraries[url].complete();
        });
        this.document.body.appendChild(script);
        return this.loadedLibraries[url].asObservable();
    }
}
TaminLazyLoadService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
TaminLazyLoadService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/** @nocollapse */ TaminLazyLoadService.ngInjectableDef = i0.defineInjectable({ factory: function TaminLazyLoadService_Factory() { return new TaminLazyLoadService(i0.inject(i1.DOCUMENT)); }, token: TaminLazyLoadService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminLazyLoadService.prototype.loadedLibraries;
    /**
     * @type {?}
     * @private
     */
    TaminLazyLoadService.prototype.document;
}
