// tamin-framework/esm2015/lib/components/tmn/field-txt/field-txt.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
export class FieldTxtComponent {
    /**
     * @param {?} controlContainer
     */
    constructor(controlContainer) {
        this.controlContainer = controlContainer;
        this.value = '';
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.blur = new EventEmitter();
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => {
        });
        this.onTouched = (/**
         * @return {?}
         */
        () => {
            this.blur.emit();
        });
    }
    /**
     * @return {?}
     */
    updateChanges() {
        this.onChange(this.value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        this.value = obj;
        this.updateChanges();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.control = this.controlContainer.control.get(this.formControlName);
    }
}
FieldTxtComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'tmn-field-txt',
                template: `<input
    type="text"
    [(ngModel)]="value"
    (ngModelChange)="updateChanges()"
    class="tamin-form-control-input"
    (blur)="onTouched()"
    [disabled]="control?.disabled"
    [class.tamin-form-control-valid]="control?.touched && control?.valid"
    [class.tamin-form-control-invalid]="control?.touched && !control?.valid"
    [attr.maxLength]="maxLength"
    [attr.placeholder]="placeHolder"
    #element>`,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => FieldTxtComponent)),
                        multi: true
                    }],
                styles: [`:host {
      width: 100%;
    }`]
            }] }
];
/** @nocollapse */
FieldTxtComponent.ctorParameters = () => [
    { type: ControlContainer }
];
FieldTxtComponent.propDecorators = {
    element: [{ type: ViewChild, args: ['element',] }],
    maxLength: [{ type: Input }],
    formControlName: [{ type: Input }],
    placeHolder: [{ type: Input }],
    blur: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FieldTxtComponent.prototype.value;
    /** @type {?} */
    FieldTxtComponent.prototype.control;
    /** @type {?} */
    FieldTxtComponent.prototype.element;
    /** @type {?} */
    FieldTxtComponent.prototype.maxLength;
    /** @type {?} */
    FieldTxtComponent.prototype.formControlName;
    /** @type {?} */
    FieldTxtComponent.prototype.placeHolder;
    /** @type {?} */
    FieldTxtComponent.prototype.blur;
    /** @type {?} */
    FieldTxtComponent.prototype.onChange;
    /** @type {?} */
    FieldTxtComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    FieldTxtComponent.prototype.controlContainer;
}
