// tamin-framework/esm5/lib/components/tmn/field-txt/field-txt.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
var FieldTxtComponent = /** @class */ (function () {
    function FieldTxtComponent(controlContainer) {
        var _this = this;
        this.controlContainer = controlContainer;
        this.value = '';
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.blur = new EventEmitter();
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) {
        });
        this.onTouched = (/**
         * @return {?}
         */
        function () {
            _this.blur.emit();
        });
    }
    /**
     * @return {?}
     */
    FieldTxtComponent.prototype.updateChanges = /**
     * @return {?}
     */
    function () {
        this.onChange(this.value);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    FieldTxtComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    FieldTxtComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FieldTxtComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.value = obj;
        this.updateChanges();
    };
    /**
     * @return {?}
     */
    FieldTxtComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.control = this.controlContainer.control.get(this.formControlName);
    };
    FieldTxtComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'tmn-field-txt',
                    template: "<input\n    type=\"text\"\n    [(ngModel)]=\"value\"\n    (ngModelChange)=\"updateChanges()\"\n    class=\"tamin-form-control-input\"\n    (blur)=\"onTouched()\"\n    [disabled]=\"control?.disabled\"\n    [class.tamin-form-control-valid]=\"control?.touched && control?.valid\"\n    [class.tamin-form-control-invalid]=\"control?.touched && !control?.valid\"\n    [attr.maxLength]=\"maxLength\"\n    [attr.placeholder]=\"placeHolder\"\n    #element>",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return FieldTxtComponent; })),
                            multi: true
                        }],
                    styles: [":host {\n      width: 100%;\n    }"]
                }] }
    ];
    /** @nocollapse */
    FieldTxtComponent.ctorParameters = function () { return [
        { type: ControlContainer }
    ]; };
    FieldTxtComponent.propDecorators = {
        element: [{ type: ViewChild, args: ['element',] }],
        maxLength: [{ type: Input }],
        formControlName: [{ type: Input }],
        placeHolder: [{ type: Input }],
        blur: [{ type: Output }]
    };
    return FieldTxtComponent;
}());
export { FieldTxtComponent };
if (false) {
    /** @type {?} */
    FieldTxtComponent.prototype.value;
    /** @type {?} */
    FieldTxtComponent.prototype.control;
    /** @type {?} */
    FieldTxtComponent.prototype.element;
    /** @type {?} */
    FieldTxtComponent.prototype.maxLength;
    /** @type {?} */
    FieldTxtComponent.prototype.formControlName;
    /** @type {?} */
    FieldTxtComponent.prototype.placeHolder;
    /** @type {?} */
    FieldTxtComponent.prototype.blur;
    /** @type {?} */
    FieldTxtComponent.prototype.onChange;
    /** @type {?} */
    FieldTxtComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    FieldTxtComponent.prototype.controlContainer;
}
