// tamin-framework/esm5/lib/components/tmn/field-txt/field-num.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
var FieldNumComponent = /** @class */ (function () {
    function FieldNumComponent(controlContainer) {
        var _this = this;
        this.controlContainer = controlContainer;
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.min = Number.MIN_SAFE_INTEGER;
        this.max = Number.MAX_SAFE_INTEGER;
        this.negativeNumberColor = 'red';
        this.blur = new EventEmitter();
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) {
        });
        this.onTouched = (/**
         * @return {?}
         */
        function () {
            _this.blur.emit();
        });
    }
    Object.defineProperty(FieldNumComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.maskRef.value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.maskRef.value = val.toString();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FieldNumComponent.prototype.updateChanges = /**
     * @return {?}
     */
    function () {
        this.onChange(this.value);
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    FieldNumComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    FieldNumComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FieldNumComponent.prototype.writeValue = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        this.value = obj;
        this.updateChanges();
    };
    /**
     * @return {?}
     */
    FieldNumComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.control = this.controlContainer.control.get(this.formControlName);
        /** @type {?} */
        var me = this;
        this.maskRef = new IMask(this.element.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max
        });
    };
    FieldNumComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'tmn-field-num',
                    template: "<input\n    type=\"text\"\n    [(ngModel)]=\"value\"\n    (ngModelChange)=\"updateChanges()\"\n    class=\"tamin-form-control-input\"\n    style=\"direction: ltr\"\n    (blur)=\"onTouched()\"\n    [disabled]=\"control?.disabled\"\n    [class.tamin-form-control-valid]=\"control?.touched && control?.valid\"\n    [class.tamin-form-control-invalid]=\"control?.touched && !control?.valid\"\n    [style.color]=\"value < 0 ? negativeNumberColor : ''\"\n    [attr.placeholder]=\"placeHolder\"\n    #element>",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return FieldNumComponent; })),
                            multi: true
                        }],
                    styles: [":host {\n      width: 100%;\n    }"]
                }] }
    ];
    /** @nocollapse */
    FieldNumComponent.ctorParameters = function () { return [
        { type: ControlContainer }
    ]; };
    FieldNumComponent.propDecorators = {
        element: [{ type: ViewChild, args: ['element',] }],
        maxLength: [{ type: Input }],
        formControlName: [{ type: Input }],
        placeHolder: [{ type: Input }],
        min: [{ type: Input }],
        max: [{ type: Input }],
        negativeNumberColor: [{ type: Input }],
        blur: [{ type: Output }]
    };
    return FieldNumComponent;
}());
export { FieldNumComponent };
if (false) {
    /** @type {?} */
    FieldNumComponent.prototype.control;
    /** @type {?} */
    FieldNumComponent.prototype.maskRef;
    /** @type {?} */
    FieldNumComponent.prototype.element;
    /** @type {?} */
    FieldNumComponent.prototype.maxLength;
    /** @type {?} */
    FieldNumComponent.prototype.formControlName;
    /** @type {?} */
    FieldNumComponent.prototype.placeHolder;
    /** @type {?} */
    FieldNumComponent.prototype.min;
    /** @type {?} */
    FieldNumComponent.prototype.max;
    /** @type {?} */
    FieldNumComponent.prototype.negativeNumberColor;
    /** @type {?} */
    FieldNumComponent.prototype.blur;
    /** @type {?} */
    FieldNumComponent.prototype.onChange;
    /** @type {?} */
    FieldNumComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    FieldNumComponent.prototype.controlContainer;
}
