// tamin-framework/esm2015/lib/components/tmn/field-txt/field-num.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, forwardRef, Input, Output, ViewChild } from '@angular/core';
import { ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
export class FieldNumComponent {
    /**
     * @param {?} controlContainer
     */
    constructor(controlContainer) {
        this.controlContainer = controlContainer;
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.min = Number.MIN_SAFE_INTEGER;
        this.max = Number.MAX_SAFE_INTEGER;
        this.negativeNumberColor = 'red';
        this.blur = new EventEmitter();
        this.onChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => {
        });
        this.onTouched = (/**
         * @return {?}
         */
        () => {
            this.blur.emit();
        });
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this.maskRef.value = val.toString();
    }
    /**
     * @return {?}
     */
    get value() {
        return this.maskRef.value;
    }
    /**
     * @return {?}
     */
    updateChanges() {
        this.onChange(this.value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        this.value = obj;
        this.updateChanges();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.control = this.controlContainer.control.get(this.formControlName);
        /** @type {?} */
        const me = this;
        this.maskRef = new IMask(this.element.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max
        });
    }
}
FieldNumComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'tmn-field-num',
                template: `<input
    type="text"
    [(ngModel)]="value"
    (ngModelChange)="updateChanges()"
    class="tamin-form-control-input"
    style="direction: ltr"
    (blur)="onTouched()"
    [disabled]="control?.disabled"
    [class.tamin-form-control-valid]="control?.touched && control?.valid"
    [class.tamin-form-control-invalid]="control?.touched && !control?.valid"
    [style.color]="value < 0 ? negativeNumberColor : ''"
    [attr.placeholder]="placeHolder"
    #element>`,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => FieldNumComponent)),
                        multi: true
                    }],
                styles: [`:host {
      width: 100%;
    }`]
            }] }
];
/** @nocollapse */
FieldNumComponent.ctorParameters = () => [
    { type: ControlContainer }
];
FieldNumComponent.propDecorators = {
    element: [{ type: ViewChild, args: ['element',] }],
    maxLength: [{ type: Input }],
    formControlName: [{ type: Input }],
    placeHolder: [{ type: Input }],
    min: [{ type: Input }],
    max: [{ type: Input }],
    negativeNumberColor: [{ type: Input }],
    blur: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FieldNumComponent.prototype.control;
    /** @type {?} */
    FieldNumComponent.prototype.maskRef;
    /** @type {?} */
    FieldNumComponent.prototype.element;
    /** @type {?} */
    FieldNumComponent.prototype.maxLength;
    /** @type {?} */
    FieldNumComponent.prototype.formControlName;
    /** @type {?} */
    FieldNumComponent.prototype.placeHolder;
    /** @type {?} */
    FieldNumComponent.prototype.min;
    /** @type {?} */
    FieldNumComponent.prototype.max;
    /** @type {?} */
    FieldNumComponent.prototype.negativeNumberColor;
    /** @type {?} */
    FieldNumComponent.prototype.blur;
    /** @type {?} */
    FieldNumComponent.prototype.onChange;
    /** @type {?} */
    FieldNumComponent.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    FieldNumComponent.prototype.controlContainer;
}
