// tamin-framework/esm5/lib/components/tamin-accordion/tamin-accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
*/
import { Component, EventEmitter, Input, Output } from '@angular/core';
var TaminAccordionComponent = /** @class */ (function () {
    function TaminAccordionComponent() {
        this.active = false;
        this.toggleAccordion = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    TaminAccordionComponent.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.toggleAccordion.emit(this.active);
    };
    /**
     * @return {?}
     */
    TaminAccordionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TaminAccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-accordion',
                    template: "<!--\r\n  Created by: k_kaviani\r\n-->\r\n\r\n<div class=\"accordion-head\" (click)=\"onClick($event)\">{{ title }}</div>\r\n<div class=\"accordion-body\" [class.active]=\"active\">\r\n  <ng-content></ng-content>\r\n</div>\r\n",
                    styles: [".accordion-head{cursor:pointer;padding:5px;margin:1px;font-size:1.1em;background-color:#dfdfdf;border:1px solid #949494;color:#555;border-radius:4px}.accordion-head:hover{background-color:#d3d3d3}.accordion-body{display:none;color:#46a8ff;padding-right:10px}.accordion-body.active{display:block;-webkit-animation:.3s fadeIn;animation:.3s fadeIn}@-webkit-keyframes fadeIn{from{opacity:0;transform:translateY(50px)}to{opacity:1;transform:translateY(0)}}@keyframes fadeIn{from{opacity:0;transform:translateY(50px)}to{opacity:1;transform:translateY(0)}}"]
                }] }
    ];
    /** @nocollapse */
    TaminAccordionComponent.ctorParameters = function () { return []; };
    TaminAccordionComponent.propDecorators = {
        title: [{ type: Input }],
        active: [{ type: Input }],
        toggleAccordion: [{ type: Output }]
    };
    return TaminAccordionComponent;
}());
export { TaminAccordionComponent };
if (false) {
    /** @type {?} */
    TaminAccordionComponent.prototype.title;
    /** @type {?} */
    TaminAccordionComponent.prototype.active;
    /** @type {?} */
    TaminAccordionComponent.prototype.toggleAccordion;
}
