// tamin-framework/esm5/lib/components/fields/tamin-field-static-multiline-text/tamin-field-static-multiline-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldStaticMultilineTextComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldStaticMultilineTextComponent, _super);
    function TaminFieldStaticMultilineTextComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rows = 4;
        _this.cols = 50;
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldStaticMultilineTextComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldStaticMultilineTextComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var theValue = value == null ? '' : value;
        this.inputElement.nativeElement.value = this.persianNumberPipe.transform(value);
        this.onChange(value);
    };
    TaminFieldStaticMultilineTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-static-multiline-text',
                    template: "<textarea\n  [disabled] = \"control.disabled\"\n  readonly=\"readonly\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  [attr.rows]=\"rows\"\n  [attr.cols]=\"cols\"\n  #inputElement></textarea>\n<span>{{hint}}</span>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldStaticMultilineTextComponent; })),
                            multi: true
                        }
                    ],
                    styles: [":host{width:100%}.tamin-form-control-input{height:inherit!important}"]
                }] }
    ];
    TaminFieldStaticMultilineTextComponent.propDecorators = {
        rows: [{ type: Input }],
        cols: [{ type: Input }]
    };
    return TaminFieldStaticMultilineTextComponent;
}(FieldBaseComponent));
export { TaminFieldStaticMultilineTextComponent };
if (false) {
    /** @type {?} */
    TaminFieldStaticMultilineTextComponent.prototype.rows;
    /** @type {?} */
    TaminFieldStaticMultilineTextComponent.prototype.cols;
}
