// tamin-framework/esm2015/lib/components/fields/tamin-field-number/tamin-field-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
export class TaminFieldNumberComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.min = Number.MIN_SAFE_INTEGER;
        this.max = Number.MAX_SAFE_INTEGER;
        this.negativeNumberColor = 'red';
    }
    /**
     * @return {?}
     */
    initializeComponent() {
        /** @type {?} */
        const me = this;
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max,
            radix: '.',
            scale: 2,
        });
    }
}
TaminFieldNumberComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-number',
                template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  style=\"direction: ltr\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [style.color]=\"inputElement?.nativeElement?.value < 0 ? negativeNumberColor : ''\"\n  #inputElement>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldNumberComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
TaminFieldNumberComponent.propDecorators = {
    min: [{ type: Input }],
    max: [{ type: Input }],
    negativeNumberColor: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldNumberComponent.prototype.min;
    /** @type {?} */
    TaminFieldNumberComponent.prototype.max;
    /** @type {?} */
    TaminFieldNumberComponent.prototype.negativeNumberColor;
}
