// tamin-framework/esm2015/lib/components/tamin-stepper/tamin-stepper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { TaminStepperStepComponent } from './tamin-stepper-step/tamin-stepper-step.component';
import { BreakpointObserver, Breakpoints } from '@angular/cdk/layout';
export class TaminStepperComponent {
    /**
     * @param {?} breakpointObserver
     */
    constructor(breakpointObserver) {
        this.breakpointObserver = breakpointObserver;
        this.breakPoints = [Breakpoints.Small, Breakpoints.Handset];
        this.renderMode = 'desktop';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeStepName(value) {
        this._activeStepName = value;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.steps.forEach((/**
             * @param {?} item
             * @param {?} index
             * @return {?}
             */
            (item, index) => {
                item.active = item.name === value;
            }));
        }));
    }
    /**
     * @return {?}
     */
    get activeStepName() {
        return this._activeStepName;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.breakpointObserver
            .observe(this.breakPoints)
            .subscribe((/**
         * @param {?} state
         * @return {?}
         */
        (state) => {
            if (state.matches) {
                if (this.renderMode !== 'mobile') {
                    this.renderMode = 'mobile';
                }
            }
            else {
                if (this.renderMode !== 'desktop') {
                    this.renderMode = 'desktop';
                }
            }
        }));
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.steps.forEach((/**
             * @param {?} item
             * @param {?} index
             * @return {?}
             */
            (item, index) => {
                item.iconText = (index + 1).toString();
            }));
        }));
    }
}
TaminStepperComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tamin-stepper',
                template: "<div\n  [ngClass]=\"{'row stepper-desktop': renderMode === 'desktop', 'stepper-mobile': renderMode === 'mobile'}\">\n  <ng-content></ng-content>\n</div>\n",
                styles: [".stepper-desktop,.stepper-mobile{position:relative}.stepper-desktop:before{content:'';position:absolute;top:50%;left:0;border-top:1px solid #cacaca;width:100%;z-index:-1}.stepper-mobile:after{content:'';position:absolute;top:0;bottom:0;right:29px;border-left:1px solid #cacaca;z-index:-1}"]
            }] }
];
/** @nocollapse */
TaminStepperComponent.ctorParameters = () => [
    { type: BreakpointObserver }
];
TaminStepperComponent.propDecorators = {
    steps: [{ type: ContentChildren, args: [TaminStepperStepComponent,] }],
    activeStepName: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminStepperComponent.prototype.steps;
    /**
     * @type {?}
     * @private
     */
    TaminStepperComponent.prototype._activeStepName;
    /** @type {?} */
    TaminStepperComponent.prototype.breakPoints;
    /** @type {?} */
    TaminStepperComponent.prototype.renderMode;
    /** @type {?} */
    TaminStepperComponent.prototype.breakpointObserver;
}
