// tamin-framework/esm5/lib/components/tamin-site/tamin-site-bottom-menu/tamin-site-bottom-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
  Revised by: a_amiri
*/
import { Component } from '@angular/core';
import { Router } from '@angular/router';
var TaminSiteBottomMenuComponent = /** @class */ (function () {
    function TaminSiteBottomMenuComponent(router) {
        this.router = router;
    }
    /**
     * @return {?}
     */
    TaminSiteBottomMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} item
     * @return {?}
     */
    TaminSiteBottomMenuComponent.prototype.handleMenu = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item.url !== '') {
            this.router.navigate([item.url]);
        }
    };
    TaminSiteBottomMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-site-bottom-menu',
                    template: "<!--\n  Created by: k_kaviani\n  Revised by: a_amiri\n-->\n<div class=\"bottom-menu row\">\n  <div class=\"col-3 col-s-3\" style=\"padding: 0px;\">\n    <button style=\"width: 150px;\" type=\"button\" class=\"button normal orange bottom-menu\" (click)=\"handleMenu('login')\">\n      <i class=\"icon-user\"></i>\n      <span>\u0648\u0631\u0648\u062F \u0628\u0647 \u0633\u0627\u0645\u0627\u0646\u0647</span>\n    </button>\n  </div>\n  <div class=\"col-3 col-s-3\" style=\"padding: 0px;\">\n    <button style=\"width: 150px;\" type=\"button\" class=\"button normal orange\" (click)=\"handleMenu('login')\">\n      <i class=\"icon-user\"></i>\n      <span>\u0648\u0631\u0648\u062F \u0628\u0647 \u0633\u0627\u0645\u0627\u0646\u0647</span>\n    </button>\n  </div>\n  <div class=\"col-3 col-s-3\" style=\"padding: 0px;\">\n    <button style=\"width: 150px;\" type=\"button\" class=\"button normal orange\" (click)=\"handleMenu('login')\">\n      <i class=\"icon-user\"></i>\n      <span>\u0648\u0631\u0648\u062F \u0628\u0647 \u0633\u0627\u0645\u0627\u0646\u0647</span>\n    </button>\n  </div>\n  <div class=\"col-3 col-s-3\" style=\"padding: 0px;\">\n    <button style=\"width: 150px;\" type=\"button\" class=\"button normal orange\" (click)=\"handleMenu('login')\">\n      <i class=\"icon-user\"></i>\n      <span>\u0648\u0631\u0648\u062F \u0628\u0647 \u0633\u0627\u0645\u0627\u0646\u0647</span>\n    </button>\n  </div>\n</div>",
                    styles: [".bottom-menu{width:100%;height:32px;position:fixed;bottom:0;background-color:rgba(0,0,0,.219);vertical-align:middle;text-align:left;font-size:10px;color:#fff;padding:7px}.button.bottom-menu{height:24px}"]
                }] }
    ];
    /** @nocollapse */
    TaminSiteBottomMenuComponent.ctorParameters = function () { return [
        { type: Router }
    ]; };
    return TaminSiteBottomMenuComponent;
}());
export { TaminSiteBottomMenuComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminSiteBottomMenuComponent.prototype.router;
}
