// tamin-framework/esm2015/lib/components/tamin-tab/tabs/tabs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, ElementRef, HostListener, Input, QueryList, ViewChild } from '@angular/core';
import { TaminTabItemComponent } from '../tab-item/tab-item.component';
import { delay, map, startWith } from 'rxjs/operators';
export class TaminTabsComponent {
    constructor() {
        this.layout = 'horizontal';
        this.tabPosition = 'top';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.tabItems = this.tabs.changes
            .pipe(startWith(''))
            .pipe(delay(0))
            .pipe(map((/**
         * @return {?}
         */
        () => this.tabs.toArray())));
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        if (!this.activeTab) {
            Promise.resolve().then((/**
             * @return {?}
             */
            () => {
                this.activeTab = this.tabs.first;
            }));
        }
    }
    /**
     * @param {?} tabItem
     * @return {?}
     */
    selectTab(tabItem) {
        if (this.activeTab === tabItem) {
            return;
        }
        if (this.activeTab) {
            this.activeTab.isActive = false;
        }
        this.activeTab = tabItem;
        tabItem.isActive = true;
    }
    /**
     * @return {?}
     */
    getBorderClass() {
        return this.tabPosition === 'top' ? 'border-bottom' : 'border-top';
    }
    /**
     * @return {?}
     */
    onResize() {
        /** @type {?} */
        const hasScroll = this.hasScroll(this.tabHeader.nativeElement, 'horizontal');
        this.scroll = hasScroll;
    }
    /**
     * @private
     * @param {?} el
     * @param {?} direction
     * @return {?}
     */
    hasScroll(el, direction) {
        direction = (direction === 'vertical') ? 'scrollTop' : 'scrollLeft';
        /** @type {?} */
        let result = !!el[direction];
        if (!result) {
            el[direction] = 1;
            result = !!el[direction];
            el[direction] = 0;
        }
        return result;
    }
    /**
     * @param {?} direction
     * @return {?}
     */
    scrollTo(direction) {
        this.tabHeader.nativeElement.scrollLeft += direction === 'left' ? -100 : 100;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.onResize();
        }));
    }
}
TaminTabsComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tabs',
                template: "<div class=\"tabs\">\n  <div *ngIf=\"tabPosition==='top'\" class=\"tab-header\" [ngClass]=\"getBorderClass()\">\n    <div class=\"flex-container\">\n      <div class=\"paddle\" *ngIf=\"scroll\" (click)=\"scrollTo('right')\"><i class=\"icon-right-open\"></i></div>\n      <ul #tabHeader>\n        <li\n          *ngFor=\"let item of tabItems | async\" (click)=\"selectTab(item)\"\n          [ngClass]=\"{'tab-header-item-active' : activeTab == item}\" class=\"tab-header\">\n          <ng-container *ngIf=\"item.labelComponent\">\n            <ng-container *ngTemplateOutlet=\"item.labelComponent.labelContent\">\n            </ng-container>\n          </ng-container>\n        </li>\n      </ul>\n      <div class=\"paddle\" *ngIf=\"scroll\" (click)=\"scrollTo('left')\">\n        <i class=\"icon-left-open\" style=\"margin-right: 5px;margin-left: 0;\"></i>\n      </div>\n    </div>\n  </div>\n  <div class=\"tab-content\" *ngFor=\"let item of tabItems | async\" [hidden]=\"activeTab !== item\">\n    <ng-container *ngTemplateOutlet=\"item.bodyComponent.bodyContent\">\n    </ng-container>\n  </div>\n</div>\n",
                styles: [".flex-container{display:flex;align-items:stretch;background-color:#f1f1f1}ul{list-style-type:none;margin:auto;padding:0;overflow-x:hidden;white-space:nowrap;flex-grow:10;background-color:#fff}li{display:inline-block;padding:10px;cursor:pointer}.tabs{border:1px solid #e0e0e0;margin:3px;background-color:#fff}.tab-header-item{padding:15px;float:right}.tab-header-item-active{border-bottom:3px solid #00f;cursor:auto}.tab-content{padding:10px}li:hover:not(.tab-header-item-active){background-color:#dee2e6;border-bottom:3px solid #dee2e6}.border-top{border-top:1px solid #e0e0e0}.border-bottom{border-bottom:1px solid #e0e0e0}.vertical-item{border:1px solid #e0e0e0;margin:3px}i{margin-left:5px}.vertical-item-header{padding:10px;cursor:pointer}.vertical-item-header:hover{background-color:#dee2e6}.vertical-item-body{padding:10px}.paddle{display:flex;align-items:center;justify-content:center}"]
            }] }
];
TaminTabsComponent.propDecorators = {
    tabs: [{ type: ContentChildren, args: [TaminTabItemComponent,] }],
    tabPosition: [{ type: Input }],
    tabHeader: [{ type: ViewChild, args: ['tabHeader',] }],
    onResize: [{ type: HostListener, args: ['window:resize',] }]
};
if (false) {
    /** @type {?} */
    TaminTabsComponent.prototype.tabs;
    /** @type {?} */
    TaminTabsComponent.prototype.tabItems;
    /** @type {?} */
    TaminTabsComponent.prototype.activeTab;
    /** @type {?} */
    TaminTabsComponent.prototype.layout;
    /** @type {?} */
    TaminTabsComponent.prototype.tabPosition;
    /** @type {?} */
    TaminTabsComponent.prototype.tabHeader;
    /** @type {?} */
    TaminTabsComponent.prototype.scroll;
}
