// tamin-framework/esm5/lib/components/fields/tamin-field-text-area/tamin-field-text-area.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldTextAreaComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldTextAreaComponent, _super);
    function TaminFieldTextAreaComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rows = 4;
        _this.cols = 50;
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldTextAreaComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
    };
    TaminFieldTextAreaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-text-area',
                    template: "<textarea\n  style=\"resize: none;height: 7em\"\n  class=\"tamin-form-control-input\"\n  [disabled]=\"control.disabled\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [attr.rows]=\"rows\"\n  [attr.cols]=\"cols\"\n  #inputElement></textarea>\n<span>{{hint}}</span>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldTextAreaComponent; })),
                            multi: true
                        }
                    ],
                    styles: [".tamin-form-control-input{height:inherit!important}"]
                }] }
    ];
    TaminFieldTextAreaComponent.propDecorators = {
        rows: [{ type: Input }],
        cols: [{ type: Input }]
    };
    return TaminFieldTextAreaComponent;
}(FieldBaseComponent));
export { TaminFieldTextAreaComponent };
if (false) {
    /** @type {?} */
    TaminFieldTextAreaComponent.prototype.rows;
    /** @type {?} */
    TaminFieldTextAreaComponent.prototype.cols;
}
