// tamin-framework/esm5/lib/components/fields/tamin-field-number/tamin-field-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
var TaminFieldNumberComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldNumberComponent, _super);
    function TaminFieldNumberComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.min = Number.MIN_SAFE_INTEGER;
        _this.max = Number.MAX_SAFE_INTEGER;
        _this.negativeNumberColor = 'red';
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldNumberComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var me = this;
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max,
            radix: '.',
            scale: 2,
        });
    };
    TaminFieldNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-number',
                    template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  style=\"direction: ltr\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [style.color]=\"inputElement?.nativeElement?.value < 0 ? negativeNumberColor : ''\"\n  #inputElement>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldNumberComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    TaminFieldNumberComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        negativeNumberColor: [{ type: Input }]
    };
    return TaminFieldNumberComponent;
}(FieldBaseComponent));
export { TaminFieldNumberComponent };
if (false) {
    /** @type {?} */
    TaminFieldNumberComponent.prototype.min;
    /** @type {?} */
    TaminFieldNumberComponent.prototype.max;
    /** @type {?} */
    TaminFieldNumberComponent.prototype.negativeNumberColor;
}
