// tamin-framework/esm5/lib/services/generic-rest-service/generic-rest.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
  Revised by: a_amiri
*/
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
// import { Observable } from "rxjs";
import { timeout } from 'rxjs/operators';
/**
 * @template T
 */
var GenericRestService = /** @class */ (function () {
    function GenericRestService(http) {
        this.http = http;
        this.restUrl = '';
        this.timeout = 180000;
        this.pagingPath = '';
    }
    /**
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    GenericRestService.prototype.getAll = /**
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    function (querySearchParams, querySortParams) {
        var _this = this;
        /** @type {?} */
        var headers = new HttpHeaders();
        /** @type {?} */
        var params = new HttpParams();
        params.append('filter', JSON.stringify(querySearchParams));
        params.append('sort', JSON.stringify(querySortParams));
        params = params.append('_dc', (new Date()).getTime().toString());
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(_this.restUrl, { headers: headers, params: params })
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} includeMetaData
     * @param {?} pageNo
     * @param {?} pageSize
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    GenericRestService.prototype.getPagedEntity = /**
     * @param {?} includeMetaData
     * @param {?} pageNo
     * @param {?} pageSize
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    function (includeMetaData, pageNo, pageSize, querySearchParams, querySortParams) {
        var _this = this;
        /** @type {?} */
        var params = new HttpParams();
        /** @type {?} */
        var headers = new HttpHeaders();
        params = params.append('filter', JSON.stringify(querySearchParams));
        params = params.append('sort', JSON.stringify(querySortParams));
        //params = params.append('start', pageNo.toString());
        params = params.append('start', ((pageNo - 1) * pageSize).toString());
        params = params.append('limit', pageSize.toString());
        params = params.append('_dc', (new Date()).getTime().toString());
        /** @type {?} */
        var url = this.restUrl + "/" + this.pagingPath;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(url, { headers: headers, params: params })
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} includeMetaData
     * @param {?} pageNo
     * @param {?} pageSize
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    GenericRestService.prototype.getPagedObject = /**
     * @param {?} includeMetaData
     * @param {?} pageNo
     * @param {?} pageSize
     * @param {?} querySearchParams
     * @param {?} querySortParams
     * @return {?}
     */
    function (includeMetaData, pageNo, pageSize, querySearchParams, querySortParams) {
        var _this = this;
        /** @type {?} */
        var params = new HttpParams();
        /** @type {?} */
        var headers = new HttpHeaders();
        params = params.append('filter', JSON.stringify(querySearchParams));
        params = params.append('sort', JSON.stringify(querySortParams));
        // params = params.append('start', pageNo.toString());
        params = params.append('start', ((pageNo - 1) * pageSize).toString());
        params = params.append('limit', pageSize.toString());
        params = params.append('_dc', (new Date()).getTime().toString());
        /** @type {?} */
        var url = this.restUrl + "/" + this.pagingPath;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(url, { headers: headers, params: params })
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    GenericRestService.prototype.getById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var url = this.restUrl + "/" + id;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(url)
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} ids
     * @return {?}
     */
    GenericRestService.prototype.getByMultipleId = /**
     * @param {?} ids
     * @return {?}
     */
    function (ids) {
        var _this = this;
        /** @type {?} */
        var url = "" + this.restUrl;
        ids.forEach((/**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            url += "/" + id;
        }));
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http.get(url)
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} dataItem
     * @return {?}
     */
    GenericRestService.prototype.create = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        var _this = this;
        /** @type {?} */
        var url = this.restUrl;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http
                .post(url, dataItem)
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} dataItem
     * @return {?}
     */
    GenericRestService.prototype.update = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        var _this = this;
        /** @type {?} */
        var url = this.restUrl;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http
                .put(url, dataItem)
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    GenericRestService.prototype.delete = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        var _this = this;
        /** @type {?} */
        var url = this.restUrl + "/" + id;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.http
                .delete(url)
                .pipe(timeout(_this.timeout))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return resolve(response); }))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            function (error) { return reject(_this.handleError(error)); }));
        }));
    };
    /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    GenericRestService.prototype.handleError = /**
     * @protected
     * @param {?} error
     * @return {?}
     */
    function (error) {
        return error;
    };
    GenericRestService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GenericRestService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return GenericRestService;
}());
export { GenericRestService };
if (false) {
    /** @type {?} */
    GenericRestService.prototype.restUrl;
    /** @type {?} */
    GenericRestService.prototype.timeout;
    /**
     * @type {?}
     * @private
     */
    GenericRestService.prototype.pagingPath;
    /**
     * @type {?}
     * @private
     */
    GenericRestService.prototype.http;
}
