// tamin-framework/esm2015/lib/components/fields/tamin-field-radio-button-group/tamin-field-radio-button-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, forwardRef, Input, QueryList } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { TaminFieldRadioButtonComponent } from '../tamin-field-radio-button/tamin-field-radio-button.component';
export class TaminFieldRadioButtonGroupComponent {
    constructor() {
        this.onTouched = (/**
         * @return {?}
         */
        () => {
        });
        this.onChange = (/**
         * @param {?} value
         * @return {?}
         */
        (value) => {
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.radioButtons.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                item.name = this.formControlName;
                item.change.subscribe((/**
                 * @param {?} value
                 * @return {?}
                 */
                value => {
                    this.onChange(value);
                }));
            }));
        }));
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.radioButtons.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                if (item.value === obj) {
                    item.setSelected();
                }
            }));
        }));
    }
}
TaminFieldRadioButtonGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-radio-button-group',
                template: "<ng-container>\n  <ng-content></ng-content>\n</ng-container>\n<span #inputElement></span>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldRadioButtonGroupComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
/** @nocollapse */
TaminFieldRadioButtonGroupComponent.ctorParameters = () => [];
TaminFieldRadioButtonGroupComponent.propDecorators = {
    radioButtons: [{ type: ContentChildren, args: [TaminFieldRadioButtonComponent,] }],
    formControlName: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.radioButtons;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.onChange;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.onTouched;
    /** @type {?} */
    TaminFieldRadioButtonGroupComponent.prototype.formControlName;
}
