// tamin-framework/esm5/lib/components/fields/tamin-field-static-persian-date/tamin-field-static-persian-date.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { TaminPersianDate } from '../../../tamin-persian-date';
var TaminFieldStaticPersianDateComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldStaticPersianDateComponent, _super);
    function TaminFieldStaticPersianDateComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    TaminFieldStaticPersianDateComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        // this._taminPersianService = this.injector.get(TaminPersianService);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldStaticPersianDateComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var theValue = value == null ? '' : value;
        // this.inputElement.nativeElement.value = value ? this._taminPersianService.getPersianDate(new Date(value)) : '';
        this.inputElement.nativeElement.value = value ? TaminPersianDate.GregorianToJalali(new Date(value)) : '';
        this.onChange(value);
    };
    TaminFieldStaticPersianDateComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-static-persian-date',
                    template: "\n<input\n  type=\"text\"\n  readonly=\"readonly\"\n  [disabled] = \"control.disabled\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  #inputElement>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldStaticPersianDateComponent; })),
                            multi: true
                        }
                    ],
                    styles: [":host{width:100%}"]
                }] }
    ];
    return TaminFieldStaticPersianDateComponent;
}(FieldBaseComponent));
export { TaminFieldStaticPersianDateComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminFieldStaticPersianDateComponent.prototype._taminPersianService;
}
