// tamin-framework/esm2015/lib/components/tamin-pdf-viewer/tamin-pdf-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { timeout } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { TaminModalComponent } from '../tamin-modal/tamin-modal.component';
export class TaminPdfViewerComponent {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.errorMessage = 'خطا در دریافت اطلاعات';
        this.beforeLoad = new EventEmitter();
        this.afterLoad = new EventEmitter();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    getBlob(url) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            this.http.get(url, { responseType: 'blob' })
                .pipe(timeout(6000))
                .toPromise()
                .then((/**
             * @param {?} response
             * @return {?}
             */
            (response) => resolve(response)))
                .catch((/**
             * @param {?} error
             * @return {?}
             */
            (error) => reject(error)));
        }));
    }
    /**
     * @param {?} url
     * @return {?}
     */
    open(url) {
        this.beforeLoad.emit();
        this.getBlob(url)
            .then((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.iframe.nativeElement.src = URL.createObjectURL(value);
            this.afterLoad.emit();
            this.theModal.show();
        }))
            .catch((/**
         * @param {?} reason
         * @return {?}
         */
        reason => {
            this.afterLoad.emit();
        }));
    }
    /**
     * @return {?}
     */
    close() {
        this.theModal.hide();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
TaminPdfViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-pdf-viewer',
                template: "<tamin-modal #theModal [title]=\"title\">\n  <iframe #iframe style=\"border:none; width:100%; height:80vh\"></iframe>\n  <ng-content></ng-content>\n</tamin-modal>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
TaminPdfViewerComponent.ctorParameters = () => [
    { type: HttpClient }
];
TaminPdfViewerComponent.propDecorators = {
    theModal: [{ type: ViewChild, args: ['theModal',] }],
    iframe: [{ type: ViewChild, args: ['iframe',] }],
    title: [{ type: Input }],
    errorMessage: [{ type: Input }],
    beforeLoad: [{ type: Output }],
    afterLoad: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaminPdfViewerComponent.prototype.theModal;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.iframe;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.title;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.errorMessage;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.beforeLoad;
    /** @type {?} */
    TaminPdfViewerComponent.prototype.afterLoad;
    /**
     * @type {?}
     * @private
     */
    TaminPdfViewerComponent.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    TaminPdfViewerComponent.prototype.http;
}
