// tamin-framework/esm5/lib/components/fields/tamin-tag/tamin-tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
var TaminTagComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminTagComponent, _super);
    function TaminTagComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TaminTagComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tag',
                    template: "<div class=\"tamin-form-control-tag\" #inputElement>\n  <div class=\"tamin-form-control-tag-element-container\">\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n    <div class=\"tamin-form-control-tag-element\"><i class=\"icon-cancel tamin-form-control-tag-element-close\"></i><span\n      class=\"tamin-form-control-tag-element-text\">\u0622\u06CC\u062A\u0645 \u0627\u0648\u0644</span></div>\n  </div>\n  <span class=\"icon\" #icon style=\"float: left\">\n    <i class=\"icon-down-open\" style=\"font-size: 1.5em\"></i>\n  </span>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminTagComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    return TaminTagComponent;
}(FieldBaseComponent));
export { TaminTagComponent };
if (false) {
    /** @type {?} */
    TaminTagComponent.prototype.selectedItem;
}
