// tamin-framework/esm2015/lib/components/tamin-menu-button/tamin-menu-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class TaminMenuButtonComponent {
    constructor() {
        this.menuItems = [];
        this.expanded = 'false';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.menuId = 'id_' + new Date().getTime().toString();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    onMenuClicked(name) {
    }
    /**
     * @return {?}
     */
    onButtonClicked() {
        this.expanded = this.expanded === 'false' ? 'true' : 'false';
    }
}
TaminMenuButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-menu-button',
                template: "<div data-tamin-menu=\"\">\n  <button\n    [attr.data-tamin-menu-opens]=\"menuId\"\n    aria-haspopup=\"true\"\n    [attr.aria-expanded]=\"expanded\"\n    (click)=\"onButtonClicked()\"\n    class=\"tmb-button\">\n    {{menuTitle}}\n    <span aria-hidden=\"true\">\u25BE</span>\n  </button>\n  <div [attr.id]=\"menuId\" data-tamin-menu-from=\"right\" role=\"menu\" hidden=\"\">\n    <button\n      role=\"menuitem\"\n      [disabled]=\"item.disabled\"\n      tabindex=\"-1\"\n      *ngFor=\"let item of menuItems\"\n      (click)=\"onMenuClicked(item.name)\">\n      {{item.title}}\n    </button>\n  </div>\n</div>\n",
                styles: ["[data-tamin-menu]{position:relative;display:inline-block}[data-tamin-menu-opens],[data-tamin-menu] [role^=menuitem]{text-align:left;border:0}[data-tamin-menu] [role=menu]{position:absolute;left:0}[data-tamin-menu] [data-tamin-menu-from=right]{left:auto;right:0}[data-tamin-menu] [role^=menuitem]{display:block;min-width:100%;white-space:nowrap}[data-tamin-menu] [role^=menuitem][aria-checked=true]::before{content:'\\2713\\0020'}.tmb-button{font-family:main-font,tahoma,verdana;text-align:center;padding:5px 10px}"]
            }] }
];
/** @nocollapse */
TaminMenuButtonComponent.ctorParameters = () => [];
TaminMenuButtonComponent.propDecorators = {
    menuTitle: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuId;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuTitle;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.checkable;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuItems;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.expanded;
}
export class MenuItem {
    /**
     * @param {?} title
     * @param {?} disabled
     * @param {?} name
     */
    constructor(title, disabled, name) {
        this.title = title;
        this.disabled = disabled;
        this.name = name;
    }
}
if (false) {
    /** @type {?} */
    MenuItem.prototype.title;
    /** @type {?} */
    MenuItem.prototype.disabled;
    /** @type {?} */
    MenuItem.prototype.name;
}
