// tamin-framework/esm2015/lib/components/tamin-document-viewer/image-viewer/image-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { TaminLazyLoadService } from '../../../services/tamin-lazy-load/tamin-lazy-load.service';
import { Subscription } from 'rxjs';
export class ImageViewerComponent {
    /**
     * @param {?} taminLazyLoadService
     */
    constructor(taminLazyLoadService) {
        this.taminLazyLoadService = taminLazyLoadService;
        this.width = '100%';
        this.height = '300px';
        this.subscription = new Subscription();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} url
     * @return {?}
     */
    open(url) {
        this.subscription.add(this.taminLazyLoadService.loadJs('assets/viewerjs/dist/viewer.js').subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.subscription.add(this.taminLazyLoadService.loadCss('assets/viewerjs/dist/viewer.css').subscribe((/**
             * @param {?} value1
             * @return {?}
             */
            value1 => {
                this.img.nativeElement.src = url;
                /** @type {?} */
                const viewer = new Viewer(this.img.nativeElement, {
                    inline: false,
                    /**
                     * @return {?}
                     */
                    viewed() {
                        viewer.zoomTo(1);
                    },
                });
            })));
        })));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
}
ImageViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-image-viewer',
                template: "<img alt=\"image\" #img style=\"max-width:100%;max-height:100%;cursor:pointer\">\n",
                styles: [""]
            }] }
];
/** @nocollapse */
ImageViewerComponent.ctorParameters = () => [
    { type: TaminLazyLoadService }
];
ImageViewerComponent.propDecorators = {
    img: [{ type: ViewChild, args: ['img',] }],
    width: [{ type: Input }],
    height: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ImageViewerComponent.prototype.img;
    /** @type {?} */
    ImageViewerComponent.prototype.width;
    /** @type {?} */
    ImageViewerComponent.prototype.height;
    /**
     * @type {?}
     * @private
     */
    ImageViewerComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    ImageViewerComponent.prototype.taminLazyLoadService;
}
