// tamin-framework/esm5/lib/components/fields/tamin-field-calculable/tamin-field-calculable.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldCalculableComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldCalculableComponent, _super);
    function TaminFieldCalculableComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.min = Number.MIN_SAFE_INTEGER;
        _this.max = Number.MAX_SAFE_INTEGER;
        _this.negativeNumberColor = 'red';
        _this.calculate = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    TaminFieldCalculableComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var me = this;
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max
        });
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldCalculableComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        /** @type {?} */
        var theValue = value == null ? '' : value;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.maskRef) {
                _this.maskRef.value = value.toString();
            }
            else {
                _this.inputElement.nativeElement.value = value.toString();
            }
        }));
        this.onChange(value);
    };
    /**
     * @return {?}
     */
    TaminFieldCalculableComponent.prototype.onCalculate = /**
     * @return {?}
     */
    function () {
        this.calculate.emit();
    };
    TaminFieldCalculableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tamin-field-calculable',
                    template: "<div class=\"input-wrapper\">\n<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  style=\"direction: ltr; padding-left: 35px;\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [style.color]=\"inputElement?.nativeElement?.value < 0 ? negativeNumberColor : ''\"\n  #inputElement>\n<span class=\"icon\" #icon>\n    <i\n      class=\"icon-calc\"\n      style=\"font-size: 1.5em\"\n      (click)=\"onCalculate()\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"\n      #iconElement></i>\n  </span>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldCalculableComponent; })),
                            multi: true
                        }
                    ],
                    styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
                }] }
    ];
    TaminFieldCalculableComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        negativeNumberColor: [{ type: Input }],
        calculate: [{ type: Output, args: ['calculate',] }]
    };
    return TaminFieldCalculableComponent;
}(FieldBaseComponent));
export { TaminFieldCalculableComponent };
if (false) {
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.min;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.max;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.negativeNumberColor;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.calculate;
}
