// tamin-framework/esm5/lib/services/persian-service/tamin-persian.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as momentNs from 'jalali-moment';
import * as i0 from "@angular/core";
/** @type {?} */
var moment = momentNs;
var TaminPersianService = /** @class */ (function () {
    function TaminPersianService() {
        this._numberRange = '[\u06F0-\u06F9]';
        this._charRange = ['[\u06A9\u06AF\u06C0\u06CC\u060C',
            '\u062A\u062B\u062C\u062D\u062E\u062F',
            '\u063A\u064A\u064B\u064C\u064D\u064E',
            '\u064F\u067E\u0670\u0686\u0698\u200C',
            '\u0621-\u0629\u0630-\u0639\u0641-\u0654]'].join('');
        this._rtlPunctuations = '(،|؟|«|»|؛|٬)';
        this._ltrPunctuations = '(\\.|:|\\!|\\-|\\[|\\]|\\(|\\)|/)';
        this._number = new RegExp('^' + this._numberRange + '+$');
        this._letter = new RegExp('^' + this._charRange + '+$');
        this._punctuation = new RegExp('^' + this.combineRegExps(this._rtlPunctuations, this._ltrPunctuations) + '+$');
        this._text = new RegExp('^' + this.combineRegExps(this._numberRange, this._charRange, this._rtlPunctuations, this._ltrPunctuations, '\\s') + '+$');
        this._rtl = new RegExp('^' + this.combineRegExps(this._charRange, this._numberRange, this._rtlPunctuations, '\\s') + '+$');
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TaminPersianService.prototype.getPersianDate = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            /** @type {?} */
            var tmp = moment.from(value.toLocaleDateString(), 'en');
            if (tmp.isValid()) {
                return tmp.locale('fa').format('YYYY/MM/DD');
            }
            return null;
        }
        catch (e) {
            return '';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminPersianService.prototype.getPersianDayOfWeek = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            /** @type {?} */
            var dayOfWeeks = ['یکشنبه', 'دوشنبه', 'سه شنبه', 'چهارشنبه', 'پنج شنبه', 'جمعه', 'شنبه'];
            /** @type {?} */
            var tmp = moment.from(value.toLocaleDateString(), 'en');
            if (tmp.isValid()) {
                return dayOfWeeks[tmp.locale('fa').day()];
            }
            return null;
        }
        catch (e) {
            return '';
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TaminPersianService.prototype.getGregorianDate = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var tmp = moment.from(value, 'fa', 'YYYY/MM/DD');
        if (tmp.isValid()) {
            return tmp.toDate();
        }
        return null;
    };
    /**
     * @private
     * @param {...?} params
     * @return {?}
     */
    TaminPersianService.prototype.combineRegExps = /**
     * @private
     * @param {...?} params
     * @return {?}
     */
    function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        /** @type {?} */
        var combined = '(';
        for (var i = 0; i < arguments.length; i++) {
            combined += '(';
            if (i !== arguments.length - 1) {
                combined += arguments[i] + ')|';
            }
            else {
                combined += arguments[i] + ')';
            }
        }
        return combined + ')';
    };
    TaminPersianService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TaminPersianService.ctorParameters = function () { return []; };
    /** @nocollapse */ TaminPersianService.ngInjectableDef = i0.defineInjectable({ factory: function TaminPersianService_Factory() { return new TaminPersianService(); }, token: TaminPersianService, providedIn: "root" });
    return TaminPersianService;
}());
export { TaminPersianService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._numberRange;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._charRange;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._rtlPunctuations;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._ltrPunctuations;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._number;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._letter;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._punctuation;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._text;
    /**
     * @type {?}
     * @private
     */
    TaminPersianService.prototype._rtl;
}
