// tamin-framework/esm5/lib/components/tamin-stepper/tamin-stepper-step/tamin-stepper-step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { BreakpointObserver, Breakpoints } from '@angular/cdk/layout';
var TaminStepperStepComponent = /** @class */ (function () {
    function TaminStepperStepComponent(breakpointObserver) {
        this.breakpointObserver = breakpointObserver;
        this.size = 1;
        this.breakPoints = [Breakpoints.Small, Breakpoints.Handset];
        this.renderMode = 'desktop';
    }
    /**
     * @return {?}
     */
    TaminStepperStepComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.breakpointObserver
            .observe(this.breakPoints)
            .subscribe((/**
         * @param {?} state
         * @return {?}
         */
        function (state) {
            if (state.matches) {
                if (_this.renderMode !== 'mobile') {
                    _this.renderMode = 'mobile';
                }
            }
            else {
                if (_this.renderMode !== 'desktop') {
                    _this.renderMode = 'desktop';
                }
            }
        }));
    };
    TaminStepperStepComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tamin-stepper-step',
                    template: "<ng-container *ngIf=\"renderMode === 'desktop'\">\n  <div [ngClass]=\"'col-' + size.toString()\">\n    <div class=\"stepper-step-desktop\">\n      <div>&nbsp;</div>\n      <div [ngClass]=\"{'stepper-icon-desktop-active' : active, 'stepper-icon-desktop' : !active}\">{{iconText}}</div>\n      <div [ngClass]=\"{'stepper-text-desktop-active' : active, 'stepper-text-desktop' : !active}\">\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      </div>\n    </div>\n  </div>\n</ng-container>\n\n<ng-container *ngIf=\"renderMode === 'mobile'\">\n  <div class=\"stepper-step-mobile-container\">\n    <div style=\"display: inline-block\" [ngClass]=\"{'stepper-icon-mobile-active' : active, 'stepper-icon-mobile' : !active}\">{{iconText}}</div>\n    <div style=\"display: inline-block; vertical-align: middle\"  [ngClass]=\"{'stepper-text-active' : active, 'stepper-mobile-text' : !active}\">\n      <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n    </div>\n  </div>\n</ng-container>\n\n<ng-template #content><ng-content></ng-content></ng-template>\n",
                    styles: [".col-1,.col-10,.col-11,.col-12,.col-2,.col-3,.col-4,.col-5,.col-6,.col-7,.col-8,.col-9{text-align:center;padding:10px}.stepper-step-desktop{position:relative}.stepper-icon-desktop{width:30px;height:30px;border:1px solid #cacaca;border-radius:50%;color:#000;background-color:#fff;margin:0 auto;font-size:1.6em}.stepper-icon-desktop-active{width:30px;height:30px;border:1px solid #00adef;border-radius:50%;color:#fff;background-color:#00adef;margin:0 auto;font-size:1.6em}.stepper-text-desktop{width:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.stepper-text-desktop-active{white-space:nowrap;overflow:hidden;color:#00adef;text-overflow:ellipsis}.stepper-step-mobile-container{width:100%;padding-right:10px}.stepper-icon-mobile{width:30px;height:30px;border:1px solid #cacaca;border-radius:50%;color:#000;background-color:#fff;font-size:1.6em;text-align:center;margin:5px}.stepper-icon-mobile-active{width:30px;height:30px;border:1px solid #00adef;border-radius:50%;color:#fff;background-color:#00adef;font-size:1.6em;text-align:center;margin:5px}.stepper-mobile-text{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:inline-block}.stepper-text-mobile-active{width:100%;white-space:nowrap;overflow:hidden;color:#00adef;text-overflow:ellipsis}"]
                }] }
    ];
    /** @nocollapse */
    TaminStepperStepComponent.ctorParameters = function () { return [
        { type: BreakpointObserver }
    ]; };
    TaminStepperStepComponent.propDecorators = {
        size: [{ type: Input }],
        name: [{ type: Input }]
    };
    return TaminStepperStepComponent;
}());
export { TaminStepperStepComponent };
if (false) {
    /** @type {?} */
    TaminStepperStepComponent.prototype.size;
    /** @type {?} */
    TaminStepperStepComponent.prototype.iconText;
    /** @type {?} */
    TaminStepperStepComponent.prototype.name;
    /** @type {?} */
    TaminStepperStepComponent.prototype.active;
    /** @type {?} */
    TaminStepperStepComponent.prototype.breakPoints;
    /** @type {?} */
    TaminStepperStepComponent.prototype.renderMode;
    /** @type {?} */
    TaminStepperStepComponent.prototype.breakpointObserver;
}
