// tamin-framework/esm5/lib/components/ribbon/ribbon.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, isDevMode } from '@angular/core';
var RibbonComponent = /** @class */ (function () {
    function RibbonComponent() {
        this.color = 'default';
    }
    /**
     * @return {?}
     */
    RibbonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    RibbonComponent.prototype.devMode = /**
     * @return {?}
     */
    function () {
        return isDevMode() || !window.hasOwnProperty('cordova');
    };
    /**
     * @return {?}
     */
    RibbonComponent.prototype.mobileMode = /**
     * @return {?}
     */
    function () {
        return window.hasOwnProperty('cordova');
    };
    RibbonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-ribbon',
                    template: "<div class=\"ribbon\" [ngClass]=\"color\" *ngIf=\"devMode() || !mobileMode()\">\n  <span>\n  <ng-content></ng-content>\n</span>\n</div>\n",
                    styles: [".ribbon{position:absolute;right:-5px;top:-5px;z-index:1001;overflow:hidden;width:75px;height:75px;text-align:right}.ribbon span{font-size:10px;color:#fff;text-transform:uppercase;text-align:center;font-weight:700;line-height:20px;transform:rotate(45deg);-webkit-transform:rotate(45deg);width:100px;display:block;background:#79a70a;background:linear-gradient(#9bc90d 0,#79a70a 100%);box-shadow:0 3px 10px -5px #000;position:absolute;top:19px;right:-21px}.ribbon span::before{content:'';position:absolute;left:0;top:100%;z-index:-1;border-left:3px solid #79a70a;border-right:3px solid transparent;border-bottom:3px solid transparent;border-top:3px solid #79a70a}.ribbon span::after{content:'';position:absolute;right:0;top:100%;z-index:-1;border-right:3px solid #79a70a;border-left:3px solid transparent;border-bottom:3px solid transparent;border-top:3px solid #79a70a}.red span{background:linear-gradient(#f70505 0,#8f0808 100%)}.red span::before{border-left-color:#8f0808;border-top-color:#8f0808}.red span::after{border-right-color:#8f0808;border-top-color:#8f0808}.blue span{background:linear-gradient(#2989d8 0,#1e5799 100%)}.blue span::before{border-left-color:#1e5799;border-top-color:#1e5799}.blue span::after{border-right-color:#1e5799;border-top-color:#1e5799}"]
                }] }
    ];
    /** @nocollapse */
    RibbonComponent.ctorParameters = function () { return []; };
    RibbonComponent.propDecorators = {
        color: [{ type: Input }]
    };
    return RibbonComponent;
}());
export { RibbonComponent };
if (false) {
    /** @type {?} */
    RibbonComponent.prototype.color;
}
