// tamin-framework/esm2015/lib/services/platform.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Platform } from '@angular/cdk/platform';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/platform";
export class PlatformService {
    /**
     * @param {?} platform
     */
    constructor(platform) {
        this.platform = platform;
    }
    /**
     * @return {?}
     */
    isAndroid() {
        return this.platform.ANDROID;
    }
    /**
     * @return {?}
     */
    isIOS() {
        return this.platform.IOS;
    }
    /**
     * @return {?}
     */
    isBlink() {
        return this.platform.BLINK;
    }
    /**
     * @return {?}
     */
    isEdge() {
        return this.platform.EDGE;
    }
    /**
     * @return {?}
     */
    isFirefox() {
        return this.platform.FIREFOX;
    }
    /**
     * @return {?}
     */
    isBrowser() {
        return this.platform.isBrowser;
    }
    /**
     * @return {?}
     */
    isSafari() {
        return this.platform.SAFARI;
    }
    /**
     * @return {?}
     */
    isTrident() {
        return this.platform.TRIDENT;
    }
    /**
     * @return {?}
     */
    isWebkit() {
        return this.platform.WEBKIT;
    }
}
PlatformService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
PlatformService.ctorParameters = () => [
    { type: Platform }
];
/** @nocollapse */ PlatformService.ngInjectableDef = i0.defineInjectable({ factory: function PlatformService_Factory() { return new PlatformService(i0.inject(i1.Platform)); }, token: PlatformService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    PlatformService.prototype.platform;
}
