// tamin-framework/esm2015/lib/services/request-cache.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class RequestCacheService {
    /**
     * @param {?} taminApplicationConfig
     */
    constructor(taminApplicationConfig) {
        this.taminApplicationConfig = taminApplicationConfig;
        this.cache = new Map();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    isCacheable(url) {
        /** @type {?} */
        const tmp = this.taminApplicationConfig.cacheableUrls.find((/**
         * @param {?} c
         * @return {?}
         */
        c => c === url));
        return tmp !== undefined;
    }
    /**
     * @param {?} url
     * @return {?}
     */
    hasCached(url) {
        return this.cache.has(url);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    getCached(url) {
        return this.cache.get(url);
    }
    /**
     * @param {?} url
     * @param {?} data
     * @return {?}
     */
    setCache(url, data) {
        return this.cache.set(url, data);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    deleteCache(url) {
        this.cache.delete(url);
    }
}
RequestCacheService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
RequestCacheService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: ['taminApplicationConfig',] }] }
];
/** @nocollapse */ RequestCacheService.ngInjectableDef = i0.defineInjectable({ factory: function RequestCacheService_Factory() { return new RequestCacheService(i0.inject("taminApplicationConfig")); }, token: RequestCacheService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    RequestCacheService.prototype.cache;
    /**
     * @type {?}
     * @private
     */
    RequestCacheService.prototype.taminApplicationConfig;
}
