// tamin-framework/esm2015/lib/components/fields/tamin-field-checkbox/tamin-field-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { fromEvent } from 'rxjs';
export class TaminFieldCheckboxComponent extends FieldBaseComponent {
    /**
     * @return {?}
     */
    get required() {
        if (this.control.validator === null) {
            return false;
        }
        /** @type {?} */
        const validator = this.control.validator((/** @type {?} */ ({})));
        return (validator && (validator.required));
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        if (typeof (value) !== 'boolean') {
            value = false;
        }
        this.inputElement.nativeElement.checked = value;
        this.onChange(value);
    }
    /**
     * @protected
     * @return {?}
     */
    initValueAccessor() {
        this.control = this.controlContainer.control.get(this.formControlName);
        if (typeof (this.control.value) !== 'boolean') {
            this.control.setValue(false);
        }
        /** @type {?} */
        const change = fromEvent(this.inputElement.nativeElement, 'change');
        this.subscription.add(change.subscribe((/**
         * @param {?} evt
         * @return {?}
         */
        (evt) => {
            this.onChange(evt.target.checked);
        })));
        /** @type {?} */
        const focuses = fromEvent(this.inputElement.nativeElement, 'focus');
        this.subscription.add(focuses.subscribe((/**
         * @param {?} evt
         * @return {?}
         */
        (evt) => {
            this.onTouched();
        })));
    }
}
TaminFieldCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-checkbox',
                template: "<div class=\"control-group\">\n  <label class=\"control control--checkbox\" [ngClass]=\"{'disable-cursor': control.disabled}\"><span [ngClass]=\"{'disable-text': control.disabled}\"><ng-content></ng-content></span>\n    <input type=\"checkbox\" #inputElement [disabled]=\"control.disabled\"/>\n    <div class=\"control__indicator\"></div>\n  </label>\n</div>\n\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldCheckboxComponent)),
                        multi: true
                    }
                ],
                styles: [".disable-text{color:#cacaca}.disable-cursor{cursor:default!important}.control-group{display:inline-block;vertical-align:top;text-align:left}.control{display:block;position:relative;padding-right:25px;cursor:pointer;color:#000;font-size:16px}.control input{position:absolute;z-index:-1;opacity:0}.control__indicator{position:absolute;top:2px;right:0;height:20px;width:20px;background:#e6e6e6}.control input:focus~.control__indicator,.control:hover input~.control__indicator{background:#ccc}.control input:checked~.control__indicator{background:#2aa1c0}.control input:checked:focus~.control__indicator,.control:hover input:not([disabled]):checked~.control__indicator{background:#0e647d}.control input:disabled~.control__indicator{background:#e6e6e6;opacity:.6;pointer-events:none}.control__indicator:after{content:'';position:absolute;display:none}.control input:checked~.control__indicator:after{display:block}.control--checkbox .control__indicator:after{left:8px;top:4px;width:3px;height:8px;border:solid #fff;border-width:0 2px 2px 0;transform:rotate(45deg)}.control--checkbox input:disabled~.control__indicator:after{border-color:#7b7b7b}.control--radio .control__indicator:after{left:7px;top:7px;height:6px;width:6px;border-radius:50%;background:#fff}.control--radio input:disabled~.control__indicator:after{background:#7b7b7b}.select{position:relative;display:inline-block;margin-bottom:15px;width:100%}.select select{display:inline-block;width:100%;cursor:pointer;padding:10px 15px;outline:0;border:0;border-radius:0;background:#e6e6e6;color:#7b7b7b;appearance:none;-webkit-appearance:none;-moz-appearance:none}.select select::-ms-expand{display:none}.select select:focus,.select select:hover{color:#000;background:#ccc}.select select:disabled{opacity:.5;pointer-events:none}.select__arrow{position:absolute;top:16px;right:15px;width:0;height:0;pointer-events:none;border-style:solid;border-width:8px 5px 0;border-color:#7b7b7b transparent transparent}.select select:focus~.select__arrow,.select select:hover~.select__arrow{border-top-color:#000}.select select:disabled~.select__arrow{border-top-color:#ccc}"]
            }] }
];
