// tamin-framework/esm5/lib/components/fields/tamin-field-static-number/tamin-field-static-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldStaticNumberComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldStaticNumberComponent, _super);
    function TaminFieldStaticNumberComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TaminFieldStaticNumberComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var theValue = value == null ? '' : value;
            _this.inputElement.nativeElement.value = _this.commaSeperated(value.toString());
            _this.onChange(value);
        }), 0);
    };
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    TaminFieldStaticNumberComponent.prototype.commaSeperated = /**
     * @private
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var parts = val.split('.');
        parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
        return parts.join('.');
    };
    TaminFieldStaticNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-static-number',
                    template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  readonly=\"readonly\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  #inputElement>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldStaticNumberComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    return TaminFieldStaticNumberComponent;
}(FieldBaseComponent));
export { TaminFieldStaticNumberComponent };
