// tamin-framework/esm2015/lib/components/fields/tamin-field-text/tamin-field-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
export class TaminFieldTextComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.focus = new EventEmitter();
        this.blur = new EventEmitter();
    }
}
TaminFieldTextComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-text',
                template: "<input\n  type=\"text\"\n  class=\"tamin-form-control-input\"\n  (blur)=\"blur.emit()\"\n  (focus)=\"focus.emit()\"\n  [disabled] = \"control?.disabled\"\n  [class.tamin-form-control-valid]=\"control?.touched && control?.valid\"\n  [class.tamin-form-control-invalid]=\"control?.touched && !control?.valid\"\n  [attr.maxLength] = \"maxLength\"\n  [attr.placeholder] =\"placeHolder\"\n  #inputElement>\n<span>{{hint}}</span>\n\n\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldTextComponent)),
                        multi: true
                    }
                ],
                styles: [":host{width:100%}"]
            }] }
];
TaminFieldTextComponent.propDecorators = {
    maxLength: [{ type: Input }],
    focus: [{ type: Output }],
    blur: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaminFieldTextComponent.prototype.maxLength;
    /** @type {?} */
    TaminFieldTextComponent.prototype.focus;
    /** @type {?} */
    TaminFieldTextComponent.prototype.blur;
}
