// tamin-framework/esm2015/lib/components/fields/tamin-field-editable/tamin-field-editable.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldEditableComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.maxLength = Number.MAX_SAFE_INTEGER;
        this.focus = new EventEmitter();
        this.blur = new EventEmitter();
        this.edit = new EventEmitter();
    }
}
TaminFieldEditableComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-editable',
                template: "<div class=\"input-wrapper\">\n  <input\n    type=\"text\"\n    disabled = \"disabled\"\n    [attr.placeholder]=\"placeHolder\"\n    style=\"direction: ltr; padding-left: 35px;\"\n    class=\"tamin-form-control-input\"\n    [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n    [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n    #inputElement>\n  <span class=\"icon\" #icon>\n    <i\n      class=\"icon-pencil\"\n      style=\"font-size: 1.5em\"\n      (click)=\"edit.emit()\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"\n      #iconElement></i>\n  </span>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldEditableComponent)),
                        multi: true
                    }
                ],
                styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
            }] }
];
TaminFieldEditableComponent.propDecorators = {
    maxLength: [{ type: Input }],
    focus: [{ type: Output }],
    blur: [{ type: Output }],
    edit: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TaminFieldEditableComponent.prototype.maxLength;
    /** @type {?} */
    TaminFieldEditableComponent.prototype.focus;
    /** @type {?} */
    TaminFieldEditableComponent.prototype.blur;
    /** @type {?} */
    TaminFieldEditableComponent.prototype.edit;
}
