// tamin-framework/esm2015/lib/services/overlay/overlay.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class OverlayService {
    constructor() {
    }
    /**
     * @param {?=} element
     * @return {?}
     */
    show(element = null) {
        /** @type {?} */
        const overlay = new PlainOverlay(element, { blur: 1, style: { zIndex: 9000 } });
        return overlay.show();
    }
    /**
     * @param {?} message
     * @param {?=} element
     * @return {?}
     */
    showMessage(message, element = null) {
        /** @type {?} */
        const html = '<div></div><div class="tamin-loader-loader4"></div>' + '<hr>' + message;
        /** @type {?} */
        const messageElement = document.createElement('div');
        messageElement.classList.add('overlay-message');
        // messageElement.classList.add('tamin-loader-loader3');
        messageElement.innerHTML = html;
        /** @type {?} */
        const overlay = new PlainOverlay(element, { blur: 1, face: messageElement });
        return overlay.show();
    }
    /**
     * @param {?} overlay
     * @return {?}
     */
    isShown(overlay) {
        return (overlay && (overlay.state === PlainOverlay.STATE_SHOWING || overlay.state === PlainOverlay.STATE_SHOWN));
    }
    /**
     * @param {?} overlay
     * @return {?}
     */
    hide(overlay) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            overlay.hide();
        }), 0);
    }
}
OverlayService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
OverlayService.ctorParameters = () => [];
/** @nocollapse */ OverlayService.ngInjectableDef = i0.defineInjectable({ factory: function OverlayService_Factory() { return new OverlayService(); }, token: OverlayService, providedIn: "root" });
