// tamin-framework/esm2015/lib/components/fields/tamin-field-calculable/tamin-field-calculable.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldCalculableComponent extends FieldBaseComponent {
    constructor() {
        super(...arguments);
        this.min = Number.MIN_SAFE_INTEGER;
        this.max = Number.MAX_SAFE_INTEGER;
        this.negativeNumberColor = 'red';
        this.calculate = new EventEmitter();
    }
    /**
     * @return {?}
     */
    initializeComponent() {
        /** @type {?} */
        const me = this;
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: Number,
            thousandsSeparator: ',',
            min: me.min,
            max: me.max
        });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        /** @type {?} */
        const theValue = value == null ? '' : value;
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.maskRef) {
                this.maskRef.value = value.toString();
            }
            else {
                this.inputElement.nativeElement.value = value.toString();
            }
        }));
        this.onChange(value);
    }
    /**
     * @return {?}
     */
    onCalculate() {
        this.calculate.emit();
    }
}
TaminFieldCalculableComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-tamin-field-calculable',
                template: "<div class=\"input-wrapper\">\n<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  style=\"direction: ltr; padding-left: 35px;\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  [style.color]=\"inputElement?.nativeElement?.value < 0 ? negativeNumberColor : ''\"\n  #inputElement>\n<span class=\"icon\" #icon>\n    <i\n      class=\"icon-calc\"\n      style=\"font-size: 1.5em\"\n      (click)=\"onCalculate()\"\n      [class.tamin-form-control-valid-icon]=\"control.touched && control.valid\"\n      [class.tamin-form-control-invalid-icon]=\"control.touched && !control.valid\"\n      #iconElement></i>\n  </span>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldCalculableComponent)),
                        multi: true
                    }
                ],
                styles: [".input-wrapper{position:relative}.icon{position:absolute;top:23%;left:5px;height:10px;margin-top:-5px;cursor:pointer}.item{padding:5px;cursor:pointer}.item:hover{background-color:silver}"]
            }] }
];
TaminFieldCalculableComponent.propDecorators = {
    min: [{ type: Input }],
    max: [{ type: Input }],
    negativeNumberColor: [{ type: Input }],
    calculate: [{ type: Output, args: ['calculate',] }]
};
if (false) {
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.min;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.max;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.negativeNumberColor;
    /** @type {?} */
    TaminFieldCalculableComponent.prototype.calculate;
}
