// tamin-framework/esm5/lib/components/fields/tamin-field-natioanl-id/tamin-field-natioanl-id.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TaminFieldNatioanlIdComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TaminFieldNatioanlIdComponent, _super);
    function TaminFieldNatioanlIdComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    TaminFieldNatioanlIdComponent.prototype.initializeComponent = /**
     * @return {?}
     */
    function () {
        this._setInputMask();
    };
    /**
     * @private
     * @return {?}
     */
    TaminFieldNatioanlIdComponent.prototype._setInputMask = /**
     * @private
     * @return {?}
     */
    function () {
        this.maskRef = new IMask(this.inputElement.nativeElement, {
            mask: '0000000000'
        });
    };
    TaminFieldNatioanlIdComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-field-natioanl-id',
                    template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  class=\"tamin-form-control-input\"\n  [class.tamin-form-control-valid]=\"control.touched && control.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control.valid\"\n  #inputElement>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TaminFieldNatioanlIdComponent; })),
                            multi: true
                        }
                    ],
                    styles: [""]
                }] }
    ];
    return TaminFieldNatioanlIdComponent;
}(FieldBaseComponent));
export { TaminFieldNatioanlIdComponent };
