// tamin-framework/esm5/lib/components/containers/tamin-tab-panel/tamin-tab-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { TaminTabComponent } from '../tamin-tab/tamin-tab.component';
var TaminTabPanelComponent = /** @class */ (function () {
    function TaminTabPanelComponent() {
        this.currentTab = 0;
    }
    /**
     * @return {?}
     */
    TaminTabPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    TaminTabPanelComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        console.log('ngAfterContentInit');
        this.tabComponents.forEach((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        function (item, index) {
            if (index === _this.currentTab) {
                item.present = true;
            }
        }));
    };
    /**
     * @param {?} tab
     * @param {?} index
     * @return {?}
     */
    TaminTabPanelComponent.prototype.onClick = /**
     * @param {?} tab
     * @param {?} index
     * @return {?}
     */
    function (tab, index) {
        this.currentTab = index;
        this.tabComponents.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.present = item === tab; }));
    };
    TaminTabPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-tab-panel',
                    template: "<div class=\"tab\">\r\n  <ng-container *ngFor=\"let tab of tabComponents, let idx = index\">\r\n    <button style=\"font-family: main-font, tahoma, verdana;\"\r\n      *ngIf=\"tab.visible\" class=\"tablinks\" (click)=\"onClick(tab, idx)\" [class.active] =\"idx === currentTab\">{{tab.title}}\r\n    </button>\r\n  </ng-container>\r\n</div>\r\n<ng-content></ng-content>\r\n",
                    styles: [".tab{overflow:hidden;border:1px solid #ccc;background-color:#f1f1f1;margin:3px 3px 0}.tab button{background-color:inherit;float:right;border:1px solid #aaa;border-bottom:0;border-radius:10px 10px 0 0;outline:0;cursor:pointer;padding:12px 14px;margin:2px 1px 0;transition:.3s}.tab button:hover{background-color:#e4e4e4}.tab button.active{background-color:#dfefff;border-color:#0af}"]
                }] }
    ];
    /** @nocollapse */
    TaminTabPanelComponent.ctorParameters = function () { return []; };
    TaminTabPanelComponent.propDecorators = {
        tabComponents: [{ type: ContentChildren, args: [TaminTabComponent,] }],
        currentTab: [{ type: Input }]
    };
    return TaminTabPanelComponent;
}());
export { TaminTabPanelComponent };
if (false) {
    /** @type {?} */
    TaminTabPanelComponent.prototype.tabComponents;
    /** @type {?} */
    TaminTabPanelComponent.prototype.currentTab;
}
