// tamin-framework/esm5/lib/components/tamin-menu-button/tamin-menu-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var TaminMenuButtonComponent = /** @class */ (function () {
    function TaminMenuButtonComponent() {
        this.menuItems = [];
        this.expanded = 'false';
    }
    /**
     * @return {?}
     */
    TaminMenuButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.menuId = 'id_' + new Date().getTime().toString();
    };
    /**
     * @param {?} name
     * @return {?}
     */
    TaminMenuButtonComponent.prototype.onMenuClicked = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
    };
    /**
     * @return {?}
     */
    TaminMenuButtonComponent.prototype.onButtonClicked = /**
     * @return {?}
     */
    function () {
        this.expanded = this.expanded === 'false' ? 'true' : 'false';
    };
    TaminMenuButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tamin-menu-button',
                    template: "<div data-tamin-menu=\"\">\n  <button\n    [attr.data-tamin-menu-opens]=\"menuId\"\n    aria-haspopup=\"true\"\n    [attr.aria-expanded]=\"expanded\"\n    (click)=\"onButtonClicked()\"\n    class=\"tmb-button\">\n    {{menuTitle}}\n    <span aria-hidden=\"true\">\u25BE</span>\n  </button>\n  <div [attr.id]=\"menuId\" data-tamin-menu-from=\"right\" role=\"menu\" hidden=\"\">\n    <button\n      role=\"menuitem\"\n      [disabled]=\"item.disabled\"\n      tabindex=\"-1\"\n      *ngFor=\"let item of menuItems\"\n      (click)=\"onMenuClicked(item.name)\">\n      {{item.title}}\n    </button>\n  </div>\n</div>\n",
                    styles: ["[data-tamin-menu]{position:relative;display:inline-block}[data-tamin-menu-opens],[data-tamin-menu] [role^=menuitem]{text-align:left;border:0}[data-tamin-menu] [role=menu]{position:absolute;left:0}[data-tamin-menu] [data-tamin-menu-from=right]{left:auto;right:0}[data-tamin-menu] [role^=menuitem]{display:block;min-width:100%;white-space:nowrap}[data-tamin-menu] [role^=menuitem][aria-checked=true]::before{content:'\\2713\\0020'}.tmb-button{font-family:main-font,tahoma,verdana;text-align:center;padding:5px 10px}"]
                }] }
    ];
    /** @nocollapse */
    TaminMenuButtonComponent.ctorParameters = function () { return []; };
    TaminMenuButtonComponent.propDecorators = {
        menuTitle: [{ type: Input }]
    };
    return TaminMenuButtonComponent;
}());
export { TaminMenuButtonComponent };
if (false) {
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuId;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuTitle;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.checkable;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.menuItems;
    /** @type {?} */
    TaminMenuButtonComponent.prototype.expanded;
}
var MenuItem = /** @class */ (function () {
    function MenuItem(title, disabled, name) {
        this.title = title;
        this.disabled = disabled;
        this.name = name;
    }
    return MenuItem;
}());
export { MenuItem };
if (false) {
    /** @type {?} */
    MenuItem.prototype.title;
    /** @type {?} */
    MenuItem.prototype.disabled;
    /** @type {?} */
    MenuItem.prototype.name;
}
