// tamin-framework/esm2015/lib/components/tamin-accordion-group/tamin-accordion-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
  Created by: k_kaviani
*/
import { Component, ContentChildren, QueryList } from '@angular/core';
import { TaminAccordionComponent } from '../tamin-accordion/tamin-accordion.component';
export class TaminAccordionGroupComponent {
    constructor() {
        this.subscriptions = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._accordions = this.accordions;
        this.removeSubscriptions();
        this.addSubscriptions();
        this.accordions.changes.subscribe((/**
         * @param {?} rex
         * @return {?}
         */
        rex => {
            this._accordions = rex;
            this.removeSubscriptions();
            this.addSubscriptions();
        }));
    }
    /**
     * @return {?}
     */
    addSubscriptions() {
        this._accordions.forEach((/**
         * @param {?} a
         * @return {?}
         */
        a => {
            /** @type {?} */
            const subscription = a.toggleAccordion.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            e => {
                this.toogleAccordion(a);
            }));
            this.subscriptions.push(subscription);
        }));
    }
    /**
     * @return {?}
     */
    removeSubscriptions() {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        sub => {
            sub.unsubscribe();
        }));
    }
    /**
     * @param {?} accordion
     * @return {?}
     */
    toogleAccordion(accordion) {
        if (!accordion.active) {
            this.accordions.forEach((/**
             * @param {?} a
             * @return {?}
             */
            a => a.active = false));
        }
        // set active accordion
        accordion.active = !accordion.active;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeSubscriptions();
    }
}
TaminAccordionGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-accordion-group',
                template: "<!--\r\n  Created by: k_kaviani\r\n-->\r\n\r\n<ng-content></ng-content>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
TaminAccordionGroupComponent.ctorParameters = () => [];
TaminAccordionGroupComponent.propDecorators = {
    accordions: [{ type: ContentChildren, args: [TaminAccordionComponent,] }]
};
if (false) {
    /** @type {?} */
    TaminAccordionGroupComponent.prototype.accordions;
    /**
     * @type {?}
     * @private
     */
    TaminAccordionGroupComponent.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    TaminAccordionGroupComponent.prototype._accordions;
}
