// tamin-framework/esm2015/lib/components/fields/tamin-field-static-number/tamin-field-static-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { FieldBaseComponent } from '../../../base/field-base.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class TaminFieldStaticNumberComponent extends FieldBaseComponent {
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const theValue = value == null ? '' : value;
            this.inputElement.nativeElement.value = this.commaSeperated(value.toString());
            this.onChange(value);
        }), 0);
    }
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    commaSeperated(val) {
        /** @type {?} */
        const parts = val.split('.');
        parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
        return parts.join('.');
    }
}
TaminFieldStaticNumberComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-static-number',
                template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  readonly=\"readonly\"\n  class=\"tamin-form-control-input\"\n  style=\"background-color: #efeff0; border-color: #cacacb;\"\n  #inputElement>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldStaticNumberComponent)),
                        multi: true
                    }
                ],
                styles: [""]
            }] }
];
