// tamin-framework/esm5/lib/services/request-cache.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var RequestCacheService = /** @class */ (function () {
    function RequestCacheService(taminApplicationConfig) {
        this.taminApplicationConfig = taminApplicationConfig;
        this.cache = new Map();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    RequestCacheService.prototype.isCacheable = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        /** @type {?} */
        var tmp = this.taminApplicationConfig.cacheableUrls.find((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c === url; }));
        return tmp !== undefined;
    };
    /**
     * @param {?} url
     * @return {?}
     */
    RequestCacheService.prototype.hasCached = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.cache.has(url);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    RequestCacheService.prototype.getCached = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this.cache.get(url);
    };
    /**
     * @param {?} url
     * @param {?} data
     * @return {?}
     */
    RequestCacheService.prototype.setCache = /**
     * @param {?} url
     * @param {?} data
     * @return {?}
     */
    function (url, data) {
        return this.cache.set(url, data);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    RequestCacheService.prototype.deleteCache = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.cache.delete(url);
    };
    RequestCacheService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    RequestCacheService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: ['taminApplicationConfig',] }] }
    ]; };
    /** @nocollapse */ RequestCacheService.ngInjectableDef = i0.defineInjectable({ factory: function RequestCacheService_Factory() { return new RequestCacheService(i0.inject("taminApplicationConfig")); }, token: RequestCacheService, providedIn: "root" });
    return RequestCacheService;
}());
export { RequestCacheService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RequestCacheService.prototype.cache;
    /**
     * @type {?}
     * @private
     */
    RequestCacheService.prototype.taminApplicationConfig;
}
