// tamin-framework/esm2015/lib/components/fields/tamin-field-mask/tamin-field-mask.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FieldBaseComponent } from '../../../base/field-base.component';
export class TaminFieldMaskComponent extends FieldBaseComponent {
    /**
     * @param {?} mask
     * @return {?}
     */
    set mask(mask) {
        /** @type {?} */
        const me = this;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.maskRef = new IMask(this.inputElement.nativeElement, {
                mask: mask,
            });
        }), 0);
    }
    /**
     * @return {?}
     */
    get mask() {
        return this.maskRef.mask;
    }
    /**
     * @param {?} regex
     * @return {?}
     */
    set regex(regex) {
        /** @type {?} */
        const me = this;
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.maskRef = new IMask(this.inputElement.nativeElement, {
                mask: new RegExp(regex),
            });
        }), 0);
    }
    /**
     * @return {?}
     */
    get regex() {
        return this.maskRef.regex;
    }
    /**
     * @return {?}
     */
    initializeComponent() {
    }
}
TaminFieldMaskComponent.decorators = [
    { type: Component, args: [{
                selector: 'tamin-field-mask',
                template: "<input\n  type=\"text\"\n  [disabled] = \"control.disabled\"\n  [attr.placeholder]=\"placeHolder\"\n  class=\"tamin-form-control-input tamin-form-control-mask\"\n  [class.tamin-form-control-valid]=\"control.touched && control?.valid\"\n  [class.tamin-form-control-invalid]=\"control.touched && !control?.valid\"\n  #inputElement>\n<span>{{hint}}</span>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TaminFieldMaskComponent)),
                        multi: true
                    }
                ],
                styles: [".tamin-form-control-mask{direction:ltr;text-align:right}"]
            }] }
];
TaminFieldMaskComponent.propDecorators = {
    mask: [{ type: Input }],
    regex: [{ type: Input }]
};
